<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LCMS data preprocessing and analysis with xcms • xcms</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LCMS data preprocessing and analysis with xcms">
<meta property="og:description" content="xcms">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xcms</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/xcms.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/new_functionality.html">New and modified functionality in xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sneumann/xcms/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LCMS data preprocessing and analysis with xcms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/master/vignettes/xcms.Rmd"><code>vignettes/xcms.Rmd</code></a></small>
      <div class="hidden name"><code>xcms.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.11/xcms">xcms</a></em><br><strong>Authors</strong>: Johannes Rainer<br><strong>Modified</strong>: 2020-05-20 05:27:30<br><strong>Compiled</strong>: Wed May 20 05:41:08 2020</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This documents describes data import, exploration, preprocessing and analysis of LCMS experiments with <code>xcms</code> version &gt;= 3. The examples and basic workflow was adapted from the original <em>LC/MS Preprocessing and Analysis with xcms</em> vignette from Colin A. Smith.</p>
<p>The new user interface and methods use the <code>XCMSnExp</code> object (instead of the <em>old</em> <code>xcmsSet</code> object) as a container for the pre-processing results. To support packages and pipelines relying on the <code>xcmsSet</code> object, it is however possible to convert an <code>XCMSnExp</code> into a <code>xcmsSet</code> object using the <code>as</code> method (i.e.<code>xset &lt;- as(x, "xcmsSet")</code>, with <code>x</code> being an <code>XCMSnExp</code> object.</p>
</div>
<div id="data-import" class="section level1">
<h1 class="hasAnchor">
<a href="#data-import" class="anchor"></a>Data import</h1>
<p><code>xcms</code> supports analysis of LC/MS data from files in (AIA/ANDI) NetCDF, mzML/mzXML and mzData format. For the actual data import Bioconductor’s <em><a href="https://bioconductor.org/packages/3.11/mzR">mzR</a></em> is used. For demonstration purpose we will analyze a subset of the data from <span class="citation">[1]</span> in which the metabolic consequences of knocking out the fatty acid amide hydrolase (FAAH) gene in mice was investigated. The raw data files (in NetCDF format) are provided with the <code>faahKO</code> data package. The data set consists of samples from the spinal cords of 6 knock-out and 6 wild-type mice. Each file contains data in centroid mode acquired in positive ion mode form 200-600 m/z and 2500-4500 seconds.</p>
<p>Below we load all required packages, locate the raw CDF files within the <code>faahKO</code> package and build a <em>phenodata</em> data frame describing the experimental setup.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">xcms</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">faahKO</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RColorBrewer</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pander</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">pheatmap</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SummarizedExperiment</span>)

<span class="co">## Get the full path to the CDF files</span>
<span class="no">cdfs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"cdf"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"faahKO"</span>), <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
            <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">## Create a phenodata data.frame</span>
<span class="no">pd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">sample_name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="no">cdfs</span>), <span class="kw">pattern</span> <span class="kw">=</span> <span class="st">".CDF"</span>,
                                   <span class="kw">replacement</span> <span class="kw">=</span> <span class="st">""</span>, <span class="kw">fixed</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                 <span class="kw">sample_group</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"KO"</span>, <span class="fl">6</span>), <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"WT"</span>, <span class="fl">6</span>)),
                 <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p>Subsequently we load the raw data as an <code>OnDiskMSnExp</code> object using the <code>readMSData</code> method from the <em><a href="https://bioconductor.org/packages/3.11/MSnbase">MSnbase</a></em> package. The <code>MSnbase</code> provides based structures and infrastructure for the processing of mass spectrometry data. Also, <code>MSnbase</code> can be used to <em>centroid</em> profile-mode MS data (see the corresponding vignette in the <code>MSnbase</code> package).</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">raw_data</span> <span class="kw">&lt;-</span> <span class="fu">readMSData</span>(<span class="kw">files</span> <span class="kw">=</span> <span class="no">cdfs</span>, <span class="kw">pdata</span> <span class="kw">=</span> <span class="fu">new</span>(<span class="st">"NAnnotatedDataFrame"</span>, <span class="no">pd</span>),
                       <span class="kw">mode</span> <span class="kw">=</span> <span class="st">"onDisk"</span>)</pre></body></html></div>
<p>The resulting <code>OnDiskMSnExp</code> object contains general information about the number of spectra, retention times, the measured total ion current etc, but does not contain the full raw data (i.e. the m/z and intensity values from each measured spectrum). Its memory footprint is thus rather small making it an ideal object to represent large metabolomics experiments while allowing to perform simple quality controls, data inspection and exploration as well as data sub-setting operations. The m/z and intensity values are imported from the raw data files on demand, hence the location of the raw data files should not be changed after initial data import.</p>
</div>
<div id="initial-data-inspection" class="section level1">
<h1 class="hasAnchor">
<a href="#initial-data-inspection" class="anchor"></a>Initial data inspection</h1>
<p>The <code>OnDiskMSnExp</code> organizes the MS data by spectrum and provides the methods <code>intensity</code>, <code>mz</code> and <code>rtime</code> to access the raw data from the files (the measured intensity values, the corresponding m/z and retention time values). In addition, the <code>spectra</code> method could be used to return all data encapsulated in <code>Spectrum</code> objects. Below we extract the retention time values from the object.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rtime</span>(<span class="no">raw_data</span>))</pre></body></html></div>
<pre><code>## F01.S0001 F01.S0002 F01.S0003 F01.S0004 F01.S0005 F01.S0006 
##  2501.378  2502.943  2504.508  2506.073  2507.638  2509.203</code></pre>
<p>All data is returned as one-dimensional vectors (a numeric vector for <code>rtime</code> and a <code>list</code> of numeric vectors for <code>mz</code> and <code>intensity</code>, each containing the values from one spectrum), even if the experiment consists of multiple files/samples. The <code>fromFile</code> function returns an integer vector providing the mapping of the values to the originating file. Below we use the <code>fromFile</code> indices to organize the <code>mz</code> values by file.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">mzs</span> <span class="kw">&lt;-</span> <span class="fu">mz</span>(<span class="no">raw_data</span>)

<span class="co">## Split the list by file</span>
<span class="no">mzs_by_file</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">mzs</span>, <span class="kw">f</span> <span class="kw">=</span> <span class="fu">fromFile</span>(<span class="no">raw_data</span>))

<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">mzs_by_file</span>)</pre></body></html></div>
<pre><code>## [1] 12</code></pre>
<p>As a first evaluation of the data we plot below the base peak chromatogram (BPC) for each file in our experiment. We use the <code>chromatogram</code> method and set the <code>aggregationFun</code> to <code>"max"</code> to return for each spectrum the maximal intensity and hence create the BPC from the raw data. To create a total ion chromatogram we could set <code>aggregationFun</code> to <code>sum</code>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co">## Get the base peak chromatograms. This reads data from the files.</span>
<span class="no">bpis</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">raw_data</span>, <span class="kw">aggregationFun</span> <span class="kw">=</span> <span class="st">"max"</span>)
<span class="co">## Define colors for the two groups</span>
<span class="no">group_colors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">3</span>, <span class="st">"Set1"</span>)[<span class="fl">1</span>:<span class="fl">2</span>], <span class="st">"60"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">group_colors</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"KO"</span>, <span class="st">"WT"</span>)

<span class="co">## Plot all chromatograms.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">bpis</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">raw_data</span>$<span class="no">sample_group</span>])</pre></body></html></div>
<p><img src="xcms_files/figure-html/data-inspection-bpc-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The <code>chromatogram</code> method returned a <code>Chromatograms</code> object that organizes individual <code>Chromatogram</code> objects (which in fact contain the chromatographic data) in a two-dimensional array: columns represent samples and rows (optionally) m/z and/or retention time ranges. Below we extract the chromatogram of the first sample and access its retention time and intensity values.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">bpi_1</span> <span class="kw">&lt;-</span> <span class="no">bpis</span>[<span class="fl">1</span>, <span class="fl">1</span>]
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rtime</span>(<span class="no">bpi_1</span>))</pre></body></html></div>
<pre><code>## F01.S0001 F01.S0002 F01.S0003 F01.S0004 F01.S0005 F01.S0006 
##  2501.378  2502.943  2504.508  2506.073  2507.638  2509.203</code></pre>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">intensity</span>(<span class="no">bpi_1</span>))</pre></body></html></div>
<pre><code>## F01.S0001 F01.S0002 F01.S0003 F01.S0004 F01.S0005 F01.S0006 
##     43888     43960     43392     42632     42200     42288</code></pre>
<p>The <code>chromatogram</code> method supports also extraction of chromatographic data from a m/z-rt slice of the MS data. In the next section we will use this method to create an extracted ion chromatogram (EIC) for a selected peak.</p>
<p>Note that <code>chromatogram</code> reads the raw data from each file to calculate the chromatogram. The <code>bpi</code> and <code>tic</code> methods on the other hand do not read any data from the raw files but use the respective information that was provided in the header definition of the input files (which might be different from the actual data).</p>
<p>Below we create boxplots representing the distribution of total ion currents per file. Such plots can be very useful to spot problematic or failing MS runs.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## Get the total ion current by file</span>
<span class="no">tc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="fu">tic</span>(<span class="no">raw_data</span>), <span class="kw">f</span> <span class="kw">=</span> <span class="fu">fromFile</span>(<span class="no">raw_data</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">tc</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">raw_data</span>$<span class="no">sample_group</span>],
        <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"intensity"</span>, <span class="kw">main</span> <span class="kw">=</span> <span class="st">"Total ion current"</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-tic-boxplot-1.png" alt="Distribution of total ion currents per file." width="768"><p class="caption">
Distribution of total ion currents per file.
</p>
</div>
<p>Also, we can cluster the samples based on similarity of their base peak chromatogram. This can also be helpful to spot potentially problematic samples in an experiment or generally get an initial overview of the sample grouping in the experiment. Since the retention times between samples are not exactly identical, we use the <code>bin</code> function to group intensities in fixed time ranges (bins) along the retention time axis. In the present example we use a bin size of 1 second, the default is 0.5 seconds. The clustering is performed using complete linkage hierarchical clustering on the pairwise correlations of the binned base peak chromatograms.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co">## Bin the BPC</span>
<span class="no">bpis_bin</span> <span class="kw">&lt;-</span> <span class="fu">bin</span>(<span class="no">bpis</span>, <span class="kw">binSize</span> <span class="kw">=</span> <span class="fl">2</span>)

<span class="co">## Calculate correlation on the log2 transformed base peak intensities</span>
<span class="no">cormat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">bpis_bin</span>, <span class="no">intensity</span>))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cormat</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cormat</span>) <span class="kw">&lt;-</span> <span class="no">raw_data</span>$<span class="no">sample_name</span>

<span class="co">## Define which phenodata columns should be highlighted in the plot</span>
<span class="no">ann</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">raw_data</span>$<span class="no">sample_group</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">ann</span>) <span class="kw">&lt;-</span> <span class="no">raw_data</span>$<span class="no">sample_name</span>

<span class="co">## Perform the cluster analysis</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(<span class="no">cormat</span>, <span class="kw">annotation</span> <span class="kw">=</span> <span class="no">ann</span>,
         <span class="kw">annotation_color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">group</span> <span class="kw">=</span> <span class="no">group_colors</span>))</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-bpc-heatmap-1.png" alt="Grouping of samples based on similarity of their base peak chromatogram." width="672"><p class="caption">
Grouping of samples based on similarity of their base peak chromatogram.
</p>
</div>
<p>The samples cluster in a pairwise manner, the KO and WT samples for the sample index having the most similar BPC. In addition, 3 larger clusters are present grouping KO and WT samples 21 and 22 together as well as 18 and 19. KO and WT samples 15 and 16 separate from these two clusters.</p>
</div>
<div id="chromatographic-peak-detection" class="section level1">
<h1 class="hasAnchor">
<a href="#chromatographic-peak-detection" class="anchor"></a>Chromatographic peak detection</h1>
<p>Next we perform the chromatographic peak detection using the <em>centWave</em> algorithm <span class="citation">[2]</span>. Before running the peak detection it is however strongly suggested to visually inspect e.g. the extracted ion chromatogram of internal standards or known compounds to evaluate and adapt the peak detection settings since the default settings will not be appropriate for most LCMS experiments. The two most critical parameters for <em>centWave</em> are the <code>peakwidth</code> (expected range of chromatographic peak widths) and <code>ppm</code> (maximum expected deviation of m/z values of centroids corresponding to one chromatographic peak; this is usually much larger than the ppm specified by the manufacturer) parameters. To evaluate the typical chromatographic peak width we plot the EIC for one peak.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">## Define the rt and m/z range of the peak area</span>
<span class="no">rtr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2700</span>, <span class="fl">2900</span>)
<span class="no">mzr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">334.9</span>, <span class="fl">335.1</span>)
<span class="co">## extract the chromatogram</span>
<span class="no">chr_raw</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">raw_data</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="no">rtr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_raw</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">chr_raw</span>$<span class="no">sample_group</span>])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-plot-eic-1.png" alt="Extracted ion chromatogram for one peak." width="768"><p class="caption">
Extracted ion chromatogram for one peak.
</p>
</div>
<p>Note that <code>Chromatogram</code> objects extracted by the <code>chromatogram</code> method contain an <code>NA</code> value if in a certain scan (i.e. for a specific retention time) no signal was measured in the respective mz range. This is reflected by the lines not being drawn as continuous lines in the plot above.</p>
<p>The peak above has a width of about 50 seconds. The <code>peakwidth</code> parameter should be set to accommodate the expected widths of peak in the data set. We set it to <code>20,80</code> for the present example data set.</p>
<p>For the <code>ppm</code> parameter we extract the full MS data (intensity, retention time and m/z values) corresponding to the above peak. To this end we first filter the raw object by retention time, then by m/z and finally plot the object with <code>type = "XIC"</code> to produce the plot below. We use the <em>pipe</em> (<code><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></code>) command better illustrate the corresponding workflow. Note also that in this type of plot identified chromatographic peaks would be indicated by default if present.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">raw_data</span> <span class="kw">%&gt;%</span>
    <span class="fu">filterRt</span>(<span class="kw">rt</span> <span class="kw">=</span> <span class="no">rtr</span>) <span class="kw">%&gt;%</span>
    <span class="fu">filterMz</span>(<span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="kw">type</span> <span class="kw">=</span> <span class="st">"XIC"</span>)</pre></body></html></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-detection-plot-ms-data-1.png" alt="Visualization of the raw MS data for one peak. For each plot: upper panel: chromatogram plotting the intensity values against the retention time, lower panel m/z against retention time plot. The individual data points are colored according to the intensity." width="1344"><p class="caption">
Visualization of the raw MS data for one peak. For each plot: upper panel: chromatogram plotting the intensity values against the retention time, lower panel m/z against retention time plot. The individual data points are colored according to the intensity.
</p>
</div>
<p>In the present data there is actually no variation in the m/z values. Usually one would see the m/z values (lower panel) scatter around the <em>real</em> m/z value of the compound. The first step of the <em>centWave</em> algorithm defines so called regions of interest (ROI) based on the difference of m/z values from consecutive scans. In detail, m/z values from consecutive scans are included into a ROI if the difference between the m/z and the mean m/z of the ROI is smaller than the user defined <code>ppm</code> parameter. A reasonable choice for the <code>ppm</code> could thus be the maximal m/z difference of data points from neighboring scans/spectra that are part of the chromatographic peak. It is suggested to inspect the ranges of m/z values for many compounds (either internal standards or compounds known to be present in the sample) and define the <code>ppm</code> parameter for <em>centWave</em> according to these.</p>
<p>Note that we can also perform the peak detection on the extracted ion chromatogram. This can help to evaluate different peak detection settings. Only be aware that peak detection on an extracted ion chromatogram will not consider the <code>ppm</code> parameter and that the estimation of the background signal is different to the peak detection on the full data set; values for the <code>snthresh</code> will hence have different consequences. Below we perform the peak detection with the <code>findChromPeaks</code> function on the extracted ion chromatogram. The submitted <em>parameter</em> object defines which algorithm will be used and allows to define the settings for this algorithm. We use the <em>centWave</em> algorithm with default settings, except for <code>snthresh</code>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">xchr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span>(<span class="no">chr_raw</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span>(<span class="kw">snthresh</span> <span class="kw">=</span> <span class="fl">2</span>))</pre></body></html></div>
<p>We can access the identified chromatographic peaks with the <code>chromPeaks</code> function.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xchr</span>))</pre></body></html></div>
<pre><code>##            rt    rtmin    rtmax       into       intb  maxo sn row column
## [1,] 2781.505 2761.160 2809.674  412134.25  355516.37 16856 13   1      1
## [2,] 2786.199 2764.290 2812.803 1496244.21 1391821.33 58736 20   1      2
## [3,] 2786.201 2764.291 2812.805  211297.25  176419.81  8158  8   1      3
## [4,] 2783.069 2762.725 2812.804  285228.07  144930.91  9154  2   1      4
## [5,] 2734.556 2714.211 2765.855   21579.37   18449.43   899  4   1      5
## [6,] 2797.154 2775.245 2815.933  159058.78  150289.31  6295 12   1      5</code></pre>
<p>Parallel to the <code>chromPeaks</code> matrix there is also a data frame <code>chromPeakData</code> that allows to add arbitrary annotations to each chromatographic peak. Below we extract this data frame that by default contains only the MS level in which the peak was identified.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="no">xchr</span>)</pre></body></html></div>
<pre><code>## DataFrame with 16 rows and 4 columns
##      ms_level is_filled       row    column
##     &lt;integer&gt; &lt;logical&gt; &lt;integer&gt; &lt;integer&gt;
## 1           1     FALSE         1         1
## 2           1     FALSE         1         2
## 3           1     FALSE         1         3
## 4           1     FALSE         1         4
## 5           1     FALSE         1         5
## ...       ...       ...       ...       ...
## 12          1     FALSE         1         8
## 13          1     FALSE         1         9
## 14          1     FALSE         1        10
## 15          1     FALSE         1        11
## 16          1     FALSE         1        12</code></pre>
<p>Next we plot the identified chromatographic peaks in the extracted ion chromatogram. We use the <code>col</code> parameter to color the individual chromatogram lines. Colors can also be specified for the identified peaks, <code>peakCol</code> for the foreground/border color, <code>peakBg</code> for the background/fill color. One color has to be provided for each chromatographic peak listed by <code>chromPeaks</code>. Below we define a color to indicate the sample group from which the sample is and use the sample information in the peaks’ <code>"sample"</code> column to assign the correct color to each chromatographic peak. More peak highlighting options are described further below.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="no">sample_colors</span> <span class="kw">&lt;-</span> <span class="no">group_colors</span>[<span class="no">xchr</span>$<span class="no">sample_group</span>]
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">xchr</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">sample_colors</span>,
     <span class="kw">peakBg</span> <span class="kw">=</span> <span class="no">sample_colors</span>[<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xchr</span>)[, <span class="st">"column"</span>]])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-eic-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. Peak area of identified chromatographic peaks are highlighted in the sample group color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. Peak area of identified chromatographic peaks are highlighted in the sample group color.
</p>
</div>
<p>Finally we perform the chromatographic peak detection on the full data set. Note that we set the argument <code>prefilter</code> to <code><a href="https://rdrr.io/r/base/c.html">c(3, 5000)</a></code> and <code>noise</code> to <code>5000</code> to reduce the run time of this vignette. With this setting we consider only signals with a value larger than 5000 in the peak detection step.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">cwp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span>(<span class="kw">peakwidth</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">20</span>, <span class="fl">80</span>), <span class="kw">noise</span> <span class="kw">=</span> <span class="fl">5000</span>,
                     <span class="kw">prefilter</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">5000</span>))
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span>(<span class="no">raw_data</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">cwp</span>)</pre></body></html></div>
<p>The results are returned as an <code>XCMSnExp</code> object which extends the <code>OnDiskMSnExp</code> object by storing also LC/GC-MS preprocessing results. This means also that all methods to sub-set and filter the data or to access the (raw) data are inherited from the <code>OnDiskMSnExp</code> object and can thus be re-used. Note also that it is possible to perform additional rounds of peak detection (e.g. on MS level &gt; 1 data) on the <code>xdata</code> object by calling <code>findChromPeaks</code> with the parameter <code>add = TRUE</code>.</p>
<p>The results from the chromatographic peak detection can be accessed with the <code>chromPeaks</code> method.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>))</pre></body></html></div>
<pre><code>##           mz mzmin mzmax       rt    rtmin    rtmax      into      intb  maxo
## CP0001 453.2 453.2 453.2 2509.203 2501.378 2527.982 1007409.0 1007380.8 38152
## CP0002 236.1 236.1 236.1 2518.593 2501.378 2537.372  253501.0  226896.3 12957
## CP0003 594.0 594.0 594.0 2601.535 2581.191 2637.529  161042.2  149297.3  7850
## CP0004 577.0 577.0 577.0 2604.665 2581.191 2626.574  136105.2  129195.5  6215
## CP0005 369.2 369.2 369.2 2587.451 2556.151 2631.269  483852.3  483777.1  7215
## CP0006 369.2 369.2 369.2 2568.671 2557.716 2578.061  144624.8  144602.9  7033
##           sn sample
## CP0001 38151      1
## CP0002    11      1
## CP0003    13      1
## CP0004    13      1
## CP0005  7214      1
## CP0006  7032      1</code></pre>
<p>The returned <code>matrix</code> provides the m/z and retention time range for each identified chromatographic peak as well as the integrated signal intensity (“into”) and the maximal peak intensitity (“maxo”). Columns “sample” contains the index of the sample in the object/experiment in which the peak was identified.</p>
<p>Annotations for each individual peak can be extracted with the <code>chromPeakData</code> function. This data frame could also be used to add/store arbitrary annotations for each detected peak.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<pre><code>## DataFrame with 5034 rows and 2 columns
##         ms_level is_filled
##        &lt;integer&gt; &lt;logical&gt;
## CP0001         1     FALSE
## CP0002         1     FALSE
## CP0003         1     FALSE
## CP0004         1     FALSE
## CP0005         1     FALSE
## ...          ...       ...
## CP5030         1     FALSE
## CP5031         1     FALSE
## CP5032         1     FALSE
## CP5033         1     FALSE
## CP5034         1     FALSE</code></pre>
<p>Peak detection will not always work perfectly leading to peak detection artifacts, such as overlapping peaks or artificially split peaks. The <code>refineChromPeaks</code> function allows to <em>refine</em> peak detection results by either or removing identified peaks exceeding a maximal allower rt width or by merging artificially split chromatographic peaks. For the former a <code>CleanPeaksParam</code> parameter object can be submitted to the function while the latter can be configured with the <code>MergeNeighboringPeaksParam</code> object. Below we post-process the peak detection results merging peaks overlapping in a 4 second window per file if the signal between in between them is lower than 75% of the smaller peak’s maximal intensity. See the <code>MergeNeighboringPeaksParam</code> help page for a detailed description of the settings and the approach.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">mpp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-merge.html">MergeNeighboringPeaksParam</a></span>(<span class="kw">expandRt</span> <span class="kw">=</span> <span class="fl">4</span>)
<span class="no">xdata_pp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-clean.html">refineChromPeaks</a></span>(<span class="no">xdata</span>, <span class="no">mpp</span>)</pre></body></html></div>
<p>An example for a merged peak is given below.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">mzr_1</span> <span class="kw">&lt;-</span> <span class="fl">305.1</span> + <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">0.01</span>, <span class="fl">0.01</span>)
<span class="no">chr_1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="fu">filterFile</span>(<span class="no">xdata</span>, <span class="fl">1</span>), <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr_1</span>)
<span class="no">chr_2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="fu">filterFile</span>(<span class="no">xdata_pp</span>, <span class="fl">1</span>), <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr_1</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_1</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_2</span>)</pre></body></html></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The splitted peak was merged into one." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing, right: after refinement. The splitted peak was merged into one.
</p>
</div>
<p>For the first trace in the chromatogram above centWave detected 3 peaks (1 for the full area and two smaller ones, see left panel in the plot above). The peak refinement with <code>MergeNeighboringPeaksParam</code> reduced them to a single peak (right panel in the figure above). Note that this refinement does not merge neighboring peaks for which the signal in between them is lower than a certain proportion (see figure below).</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">mzr_1</span> <span class="kw">&lt;-</span> <span class="fl">496.2</span> + <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">0.01</span>, <span class="fl">0.01</span>)
<span class="no">chr_1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="fu">filterFile</span>(<span class="no">xdata</span>, <span class="fl">1</span>), <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr_1</span>)
<span class="no">chr_2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="fu">filterFile</span>(<span class="no">xdata_pp</span>, <span class="fl">1</span>), <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr_1</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_1</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_2</span>)</pre></body></html></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-not-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The peaks were not merged." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing, right: after refinement. The peaks were not merged.
</p>
</div>
<p>Note also that it is possible to perform the peak refinement on extracted ion chromatograms. This could e.g. be used to fine-tune the settings for the parameter. To illustrate this we perform below a peak refinement on the extracted ion chromatogram <code>chr_1</code> reducing the <code>minProp</code> parameter to force joining the two peaks.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-clean.html">refineChromPeaks</a></span>(<span class="no">chr_1</span>, <span class="fu"><a href="../reference/refineChromPeaks-merge.html">MergeNeighboringPeaksParam</a></span>(<span class="kw">minProp</span> <span class="kw">=</span> <span class="fl">0.05</span>))
<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">res</span>)</pre></body></html></div>
<pre><code>##         mz mzmin mzmax       rt    rtmin    rtmax     into intb    maxo  sn
## CPM1 496.2 496.2 496.2 3384.012 3294.809 3412.181 45940118   NA 1128960 177
##      sample row column
## CPM1      1   1      1</code></pre>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">res</span>)</pre></body></html></div>
<p><img src="xcms_files/figure-html/peak-postprocessing-chr-1.png" width="700"></p>
<p>Before proceeding we replace the <code>xdata</code> object with the results from the peak refinement.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="no">xdata_pp</span></pre></body></html></div>
<p>Below we use the data from the <code>chromPeaks</code> matrix to calculate some per-file summaries.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">summary_fun</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">z</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">peak_count</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">z</span>), <span class="kw">rt</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span>(<span class="no">z</span>[, <span class="st">"rtmax"</span>] - <span class="no">z</span>[, <span class="st">"rtmin"</span>]))

<span class="no">T</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/split.html">split.data.frame</a></span>(
    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>), <span class="kw">f</span> <span class="kw">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>)[, <span class="st">"sample"</span>]),
    <span class="kw">FUN</span> <span class="kw">=</span> <span class="no">summary_fun</span>)
<span class="no">T</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="no">rbind</span>, <span class="no">T</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">T</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="fu">fileNames</span>(<span class="no">xdata</span>))
<span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html">pandoc.table</a></span>(
    <span class="no">T</span>,
    <span class="kw">caption</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Summary statistics on identified chromatographic"</span>,
                     <span class="st">" peaks. Shown are number of identified peaks per"</span>,
                     <span class="st">" sample and widths/duration of chromatographic "</span>,
                     <span class="st">"peaks."</span>))</pre></body></html></div>
<table class="table">
<caption>Summary statistics on identified chromatographic peaks. Shown are number of identified peaks per sample and widths/duration of chromatographic peaks.</caption>
<colgroup>
<col width="20%">
<col width="17%">
<col width="10%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">peak_count</th>
<th align="center">rt.0%</th>
<th align="center">rt.25%</th>
<th align="center">rt.50%</th>
<th align="center">rt.75%</th>
<th align="center">rt.100%</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>ko15.CDF</strong></td>
<td align="center">466</td>
<td align="center">7.824</td>
<td align="center">27</td>
<td align="center">39.12</td>
<td align="center">51.64</td>
<td align="center">319.2</td>
</tr>
<tr class="even">
<td align="center"><strong>ko16.CDF</strong></td>
<td align="center">657</td>
<td align="center">7.824</td>
<td align="center">23.47</td>
<td align="center">39.12</td>
<td align="center">50.08</td>
<td align="center">151.8</td>
</tr>
<tr class="odd">
<td align="center"><strong>ko18.CDF</strong></td>
<td align="center">391</td>
<td align="center">6.26</td>
<td align="center">28.17</td>
<td align="center">42.25</td>
<td align="center">53.99</td>
<td align="center">173.7</td>
</tr>
<tr class="even">
<td align="center"><strong>ko19.CDF</strong></td>
<td align="center">278</td>
<td align="center">7.824</td>
<td align="center">31.3</td>
<td align="center">46.95</td>
<td align="center">60.64</td>
<td align="center">144</td>
</tr>
<tr class="odd">
<td align="center"><strong>ko21.CDF</strong></td>
<td align="center">249</td>
<td align="center">7.824</td>
<td align="center">37.56</td>
<td align="center">48.51</td>
<td align="center">62.6</td>
<td align="center">164.3</td>
</tr>
<tr class="even">
<td align="center"><strong>ko22.CDF</strong></td>
<td align="center">291</td>
<td align="center">6.26</td>
<td align="center">26.6</td>
<td align="center">43.82</td>
<td align="center">57.9</td>
<td align="center">133</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt15.CDF</strong></td>
<td align="center">473</td>
<td align="center">7.824</td>
<td align="center">26.6</td>
<td align="center">40.69</td>
<td align="center">51.64</td>
<td align="center">161.2</td>
</tr>
<tr class="even">
<td align="center"><strong>wt16.CDF</strong></td>
<td align="center">432</td>
<td align="center">7.824</td>
<td align="center">28.17</td>
<td align="center">42.25</td>
<td align="center">54.77</td>
<td align="center">161.2</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt18.CDF</strong></td>
<td align="center">346</td>
<td align="center">7.824</td>
<td align="center">29.73</td>
<td align="center">45.38</td>
<td align="center">59.08</td>
<td align="center">424.1</td>
</tr>
<tr class="even">
<td align="center"><strong>wt19.CDF</strong></td>
<td align="center">299</td>
<td align="center">7.824</td>
<td align="center">31.3</td>
<td align="center">46.95</td>
<td align="center">59.47</td>
<td align="center">593.1</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt21.CDF</strong></td>
<td align="center">283</td>
<td align="center">7.824</td>
<td align="center">28.17</td>
<td align="center">45.38</td>
<td align="center">61.82</td>
<td align="center">172.1</td>
</tr>
<tr class="even">
<td align="center"><strong>wt22.CDF</strong></td>
<td align="center">395</td>
<td align="center">7.824</td>
<td align="center">28.17</td>
<td align="center">42.25</td>
<td align="center">54.77</td>
<td align="center">161.2</td>
</tr>
</tbody>
</table>
<p>We can also plot the location of the identified chromatographic peaks in the m/z - retention time space for one file using the <code>plotChromPeaks</code> function. Below we plot the chromatographic peaks for the 3rd sample.</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeaks</a></span>(<span class="no">xdata</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="fl">3</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peaks-plot-1.png" alt="Identified chromatographic peaks in the m/z by retention time space for one sample." width="768"><p class="caption">
Identified chromatographic peaks in the m/z by retention time space for one sample.
</p>
</div>
<p>To get a global overview of the peak detection we can plot the frequency of identified peaks per file along the retention time axis. This allows to identify time periods along the MS run in which a higher number of peaks was identified and evaluate whether this is consistent across files.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeakImage</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-image-1.png" alt="Frequency of identified chromatographic peaks along the retention time axis. The frequency is color coded with higher frequency being represented by yellow-white. Each line shows the peak frequency for one file." width="960"><p class="caption">
Frequency of identified chromatographic peaks along the retention time axis. The frequency is color coded with higher frequency being represented by yellow-white. Each line shows the peak frequency for one file.
</p>
</div>
<p>Next we highlight the identified chromatographic peaks for the example peak from before. Evaluating such plots on a list of peaks corresponding to known peaks or internal standards helps to ensure that peak detection settings were appropriate and correctly identified the expected peaks. We extract the ion chromatogram from the peak detection result object, which contains then also the identified chromatographic peaks for that ion that we can extract with the <code>chromPeaks</code> function.</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="no">chr_ex</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="no">rtr</span>)
<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">chr_ex</span>)</pre></body></html></div>
<pre><code>##         mz mzmin mzmax       rt    rtmin    rtmax       into       intb  maxo
## CP0057 335   335   335 2781.505 2761.160 2809.674  412134.25  383167.42 16856
## CP0590 335   335   335 2786.199 2764.290 2812.803 1496244.21 1461187.31 58736
## CP1275 335   335   335 2786.201 2764.291 2812.805  211297.25  194283.05  8158
## CP1690 335   335   335 2783.069 2762.725 2812.804  285228.07  264248.60  9154
## CP1990 335   335   335 2797.154 2775.245 2815.933  159058.78  149229.62  6295
## CP3121 335   335   335 2786.199 2764.290 2812.803  932645.21  915333.83 35856
## CP3616 335   335   335 2787.765 2765.855 2820.629 1636669.03 1603950.74 54640
## CP3999 335   335   335 2784.635 2759.595 2809.674  643673.21  631119.07 20672
## CP4347 335   335   335 2792.461 2768.987 2823.760  876585.53  848569.14 27200
## CP4653 335   335   335 2789.329 2773.680 2806.544   89582.57   83926.74  5427
##        sn sample row column
## CP0057 23      1   1      1
## CP0590 72      2   1      2
## CP1275 15      3   1      3
## CP1690 18      4   1      4
## CP1990 13      5   1      5
## CP3121 66      8   1      8
## CP3616 78      9   1      9
## CP3999 54     10   1     10
## CP4347 36     11   1     11
## CP4653 12     12   1     12</code></pre>
<p>We can also plot the extracted ion chromatogram. Identified chromatographic peaks will be automatically highlighted in the plot. Below we highlight chromatographic peaks with a rectangle from the peak’s minimal to maximal rt and from an intensity of 0 to the maximal signal of the peak.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">sample_colors</span> <span class="kw">&lt;-</span> <span class="no">group_colors</span>[<span class="no">chr_ex</span>$<span class="no">sample_group</span>]
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_ex</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">sample_colors</span>, <span class="kw">peakType</span> <span class="kw">=</span> <span class="st">"rectangle"</span>,
     <span class="kw">peakCol</span> <span class="kw">=</span> <span class="no">sample_colors</span>[<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">chr_ex</span>)[, <span class="st">"sample"</span>]],
     <span class="kw">peakBg</span> <span class="kw">=</span> <span class="fl">NA</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The rectangles indicate the identified chromatographic peaks per sample." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The rectangles indicate the identified chromatographic peaks per sample.
</p>
</div>
<p>Alternatively to the rectangle visualization above, it is possible to represent the apex position of each peak with a single point (passing argument <code>type = "point"</code> to the function), or draw the actually identified peak by specifying <code>type = "polygon"</code>. To completely omit highlighting the identified peaks (e.g. to plot base peak chromatograms or similar) <code>type = "none"</code> can be used. Below we use <code>type = "polygon"</code> to fill the peak area for each identified chromatographic peak in each sample. Whether individual peaks can be still identified in such a plot depends however on the number of samples from which peaks are drawn.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_ex</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">chr_raw</span>$<span class="no">sample_group</span>], <span class="kw">lwd</span> <span class="kw">=</span> <span class="fl">2</span>,
     <span class="kw">peakBg</span> <span class="kw">=</span> <span class="no">sample_colors</span>[<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">chr_ex</span>)[, <span class="st">"sample"</span>]])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-polygon-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The signal area of identified chromatographic peaks are filled with a color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The signal area of identified chromatographic peaks are filled with a color.
</p>
</div>
<p>Note that we can also specifically extract identified chromatographic peaks for a selected region by providing the respective m/z and retention time ranges with the <code>mz</code> and <code>rt</code> arguments in the <code>chromPeaks</code> method.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html">pander</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="no">rtr</span>),
       <span class="kw">caption</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"Identified chromatographic peaks in a selected "</span>,
                       <span class="st">"m/z and retention time range."</span>))</pre></body></html></div>
<table class="table">
<caption>Identified chromatographic peaks in a selected m/z and retention time range. (continued below)</caption>
<colgroup>
<col width="16%">
<col width="7%">
<col width="10%">
<col width="10%">
<col width="8%">
<col width="10%">
<col width="10%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">mz</th>
<th align="center">mzmin</th>
<th align="center">mzmax</th>
<th align="center">rt</th>
<th align="center">rtmin</th>
<th align="center">rtmax</th>
<th align="center">into</th>
<th align="center">intb</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>CP0057</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2782</td>
<td align="center">2761</td>
<td align="center">2810</td>
<td align="center">412134</td>
<td align="center">383167</td>
</tr>
<tr class="even">
<td align="center"><strong>CP0590</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2786</td>
<td align="center">2764</td>
<td align="center">2813</td>
<td align="center">1496244</td>
<td align="center">1461187</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1275</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2786</td>
<td align="center">2764</td>
<td align="center">2813</td>
<td align="center">211297</td>
<td align="center">194283</td>
</tr>
<tr class="even">
<td align="center"><strong>CP1690</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2783</td>
<td align="center">2763</td>
<td align="center">2813</td>
<td align="center">285228</td>
<td align="center">264249</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1990</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2797</td>
<td align="center">2775</td>
<td align="center">2816</td>
<td align="center">159059</td>
<td align="center">149230</td>
</tr>
<tr class="even">
<td align="center"><strong>CP3121</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2786</td>
<td align="center">2764</td>
<td align="center">2813</td>
<td align="center">932645</td>
<td align="center">915334</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP3616</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2788</td>
<td align="center">2766</td>
<td align="center">2821</td>
<td align="center">1636669</td>
<td align="center">1603951</td>
</tr>
<tr class="even">
<td align="center"><strong>CP3999</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2785</td>
<td align="center">2760</td>
<td align="center">2810</td>
<td align="center">643673</td>
<td align="center">631119</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP4347</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2792</td>
<td align="center">2769</td>
<td align="center">2824</td>
<td align="center">876586</td>
<td align="center">848569</td>
</tr>
<tr class="even">
<td align="center"><strong>CP4653</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2789</td>
<td align="center">2774</td>
<td align="center">2807</td>
<td align="center">89583</td>
<td align="center">83927</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="18%">
<col width="11%">
<col width="6%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">maxo</th>
<th align="center">sn</th>
<th align="center">sample</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>CP0057</strong></td>
<td align="center">16856</td>
<td align="center">23</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="center"><strong>CP0590</strong></td>
<td align="center">58736</td>
<td align="center">72</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1275</strong></td>
<td align="center">8158</td>
<td align="center">15</td>
<td align="center">3</td>
</tr>
<tr class="even">
<td align="center"><strong>CP1690</strong></td>
<td align="center">9154</td>
<td align="center">18</td>
<td align="center">4</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1990</strong></td>
<td align="center">6295</td>
<td align="center">13</td>
<td align="center">5</td>
</tr>
<tr class="even">
<td align="center"><strong>CP3121</strong></td>
<td align="center">35856</td>
<td align="center">66</td>
<td align="center">8</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP3616</strong></td>
<td align="center">54640</td>
<td align="center">78</td>
<td align="center">9</td>
</tr>
<tr class="even">
<td align="center"><strong>CP3999</strong></td>
<td align="center">20672</td>
<td align="center">54</td>
<td align="center">10</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP4347</strong></td>
<td align="center">27200</td>
<td align="center">36</td>
<td align="center">11</td>
</tr>
<tr class="even">
<td align="center"><strong>CP4653</strong></td>
<td align="center">5427</td>
<td align="center">12</td>
<td align="center">12</td>
</tr>
</tbody>
</table>
<p>Finally we plot also the distribution of peak intensity per sample. This allows to investigate whether systematic differences in peak signals between samples are present.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="co">## Extract a list of per-sample peak intensities (in log2 scale)</span>
<span class="no">ints</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>)[, <span class="st">"into"</span>]),
              <span class="kw">f</span> <span class="kw">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">xdata</span>)[, <span class="st">"sample"</span>])
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span>(<span class="no">ints</span>, <span class="kw">varwidth</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">xdata</span>$<span class="no">sample_group</span>],
        <span class="kw">ylab</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span>(<span class="no">log</span>[<span class="fl">2</span>]~<span class="no">intensity</span>), <span class="kw">main</span> <span class="kw">=</span> <span class="st">"Peak intensities"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/grid.html">grid</a></span>(<span class="kw">nx</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="kw">ny</span> <span class="kw">=</span> <span class="kw">NULL</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-intensity-boxplot-1.png" alt="Peak intensity distribution per sample." width="960"><p class="caption">
Peak intensity distribution per sample.
</p>
</div>
</div>
<div id="alignment" class="section level1">
<h1 class="hasAnchor">
<a href="#alignment" class="anchor"></a>Alignment</h1>
<p>The time at which analytes elute in the chromatography can vary between samples (and even compounds). Such a difference was already observable in the extracted ion chromatogram plot shown as an example in the previous section. The alignment step, also referred to as retention time correction, aims at adjusting this by shifting signals along the retention time axis to align the signals between different samples within an experiment.</p>
<p>A plethora of alignment algorithms exist (see <span class="citation">[3]</span>), with some of them being implemented also in <code>xcms</code>. The method to perform the alignment/retention time correction in <code>xcms</code> is <code>adjustRtime</code> which uses different alignment algorithms depending on the provided parameter class.</p>
<p>In the example below we use the <em>obiwarp</em> method <span class="citation">[4]</span> to align the samples. We use a <code>binSize = 0.6</code> which creates warping functions in mz bins of 0.6. Also here it is advisable to modify the settings for each experiment and evaluate if retention time correction did align internal controls or known compounds properly.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span>(<span class="no">xdata</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="fu"><a href="../reference/adjustRtime-obiwarp.html">ObiwarpParam</a></span>(<span class="kw">binSize</span> <span class="kw">=</span> <span class="fl">0.6</span>))</pre></body></html></div>
<p><code>adjustRtime</code>, besides calculating adjusted retention times for each spectrum, does also adjust the reported retention times of the identified chromatographic peaks.</p>
<p>To extract the adjusted retention times we can use the <code>adjustedRtime</code> method, or simply the <code>rtime</code> method that, if present, returns by default adjusted retention times from an <code>XCMSnExp</code> object.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="co">## Extract adjusted retention times</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">adjustedRtime</a></span>(<span class="no">xdata</span>))</pre></body></html></div>
<pre><code>## F01.S0001 F01.S0002 F01.S0003 F01.S0004 F01.S0005 F01.S0006 
##  2501.378  2502.958  2504.538  2506.118  2507.699  2509.280</code></pre>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="co">## Or simply use the rtime method</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">rtime</span>(<span class="no">xdata</span>))</pre></body></html></div>
<pre><code>## F01.S0001 F01.S0002 F01.S0003 F01.S0004 F01.S0005 F01.S0006 
##  2501.378  2502.958  2504.538  2506.118  2507.699  2509.280</code></pre>
<p><em>Raw</em> retention times can be extracted from an <code>XCMSnExp</code> containing aligned data with <code>rtime(xdata, adjusted = FALSE)</code>.</p>
<p>To evaluate the impact of the alignment we plot the BPC on the adjusted data. In addition we plot the differences of the adjusted- to the raw retention times per sample using the <code>plotAdjustedRtime</code> function. Note that we <strong>don’t</strong> want to highlight any chromatographic peaks (because there would simply be too many) and set thus <code>peakType = "none"</code> in the <code>plot</code> call.</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="co">## Get the base peak chromatograms.</span>
<span class="no">bpis_adj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata</span>, <span class="kw">aggregationFun</span> <span class="kw">=</span> <span class="st">"max"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">1</span>), <span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4.5</span>, <span class="fl">4.2</span>, <span class="fl">1</span>, <span class="fl">0.5</span>))
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">bpis_adj</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">bpis_adj</span>$<span class="no">sample_group</span>], <span class="kw">peakType</span> <span class="kw">=</span> <span class="st">"none"</span>)
<span class="co">## Plot also the difference of adjusted to raw retention time.</span>
<span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span>(<span class="no">xdata</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">xdata</span>$<span class="no">sample_group</span>])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-obiwarp-plot-1.png" alt="Obiwarp aligned data. Base peak chromatogram after alignment (top) and difference between adjusted and raw retention times along the retention time axis (bottom)." width="960"><p class="caption">
Obiwarp aligned data. Base peak chromatogram after alignment (top) and difference between adjusted and raw retention times along the retention time axis (bottom).
</p>
</div>
<p>Too large differences between adjusted and raw retention times could indicate poorly performing samples or alignment.</p>
<p>Alternatively to <em>obiwarp</em> we could use the <em>peak groups</em> alignment method that adjusts the retention time by aligning previously identified <em>hook peaks</em> (chromatographic peaks present in most/all samples). Ideally, these hook peaks should span most part of the retention time range. Below we first restore the raw retention times (also of the identified peaks) using the <code>dropAdjustedRtime</code> methods. Note that a <code>drop*</code> method is available for each preprocessing step allowing to remove the respective results from the <code>XCMSnExp</code> object.</p>
<div class="sourceCode" id="cb49"><html><body><pre class="r"><span class="co">## Does the object have adjusted retention times?</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="co">## Drop the alignment results.</span>
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">dropAdjustedRtime</a></span>(<span class="no">xdata</span>)

<span class="co">## Does the object have adjusted retention times?</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<p><strong>Note</strong>: <code>XCMSnExp</code> objects hold the raw along with the adjusted retention times and subsetting will in most cases drop the adjusted retention times. Sometimes it might thus be useful to <strong>replace</strong> the raw retention times with the adjusted retention times. This can be done with the <code>applyAdjustedRtime</code>.</p>
<p>As noted above the <em>peak groups</em> method requires peak groups (features) present in most samples to perform the alignment. We thus have to perform a first correspondence run to identify such peaks (details about the algorithm used are presented in the next section). We use here again default settings, but it is strongly advised to adapt the parameters for each data set. The definition of the sample groups (i.e. assignment of individual samples to the sample groups in the experiment) is mandatory for the <code>PeakDensityParam</code>. If there are no sample groups in the experiment <code>sampleGroups</code> should be set to a single value for each file (e.g. <code>rep(1, length(fileNames(xdata))</code>).</p>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="co">## Correspondence: group peaks across samples.</span>
<span class="no">pdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span>(<span class="kw">sampleGroups</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_group</span>,
                        <span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.8</span>)
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span>(<span class="no">xdata</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pdp</span>)

<span class="co">## Now the retention time correction.</span>
<span class="no">pgp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime-peakGroups.html">PeakGroupsParam</a></span>(<span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.85</span>)

<span class="co">## Get the peak groups that would be used for alignment.</span>
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span>(<span class="no">xdata</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pgp</span>)</pre></body></html></div>
<p>Note also that we could use the <code>adjustRtimePeakGroups</code> method on the object before alignment to evaluate on which features (peak groups) the alignment would be performed. This can be useful to test different settings for the peak groups algorithm. Also, it is possible to manually select or define certain peak groups (i.e. their retention times per sample) and provide this matrix to the <code>PeakGroupsParam</code> class with the <code>peakGroupsMatrix</code> argument.</p>
<p>Below plot the difference between raw and adjusted retention times using the <code>plotAdjustedRtime</code> function, which, if the <em>peak groups</em> method is used for alignment, also highlights the peak groups used in the adjustment.</p>
<div class="sourceCode" id="cb54"><html><body><pre class="r"><span class="co">## Plot the difference of adjusted to raw retention time.</span>
<span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span>(<span class="no">xdata</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">xdata</span>$<span class="no">sample_group</span>],
                  <span class="kw">peakGroupsCol</span> <span class="kw">=</span> <span class="st">"grey"</span>, <span class="kw">peakGroupsPch</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-peak-groups-plot-1.png" alt="Peak groups aligned data." width="960"><p class="caption">
Peak groups aligned data.
</p>
</div>
<p>At last we evaluate the impact of the alignment on the test peak.</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">1</span>))
<span class="co">## Plot the raw data</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_raw</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">chr_raw</span>$<span class="no">sample_group</span>])

<span class="co">## Extract the chromatogram from the adjusted object</span>
<span class="no">chr_adj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="no">rtr</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_adj</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">chr_raw</span>$<span class="no">sample_group</span>], <span class="kw">peakType</span> <span class="kw">=</span> <span class="st">"none"</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-peak-groups-example-peak-1.png" alt="Example extracted ion chromatogram before (top) and after alignment (bottom)." width="960"><p class="caption">
Example extracted ion chromatogram before (top) and after alignment (bottom).
</p>
</div>
<div id="subset-based-alignment" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-based-alignment" class="anchor"></a>Subset-based alignment</h2>
<p>In some experiments it might be helpful to perform the alignment based on only a subset of the samples, e.g. if QC samples were injected at regular intervals or if the experiment contains blanks. Alignment method in <code>xcms</code> allow to estimate retention time drifts on a subset of samples (either all samples excluding blanks or QC samples injected at regular intervals during a measurement run) and use these to adjust the full data set.</p>
<p>Parameters <code>subset</code> (of the <code>PeakGroupsParam</code> or <code>ObiwarpParam</code> object) can be used to define the subset of samples on which the alignment of the full data set will be based (e.g. <code>subset</code> being the index of QC samples), and parameter <code>subsetAdjust</code> allows to specify the method by which the <em>left-out</em> samples will be adjusted. There are currently two options available:</p>
<ul>
<li><p><code>subsetAdjust = "previous"</code>: adjust the retention times of a non-subset sample based on the alignment results of the previous subset sample (e.g. a QC sample). If samples are e.g. in the order <em>A1</em>, <em>B1</em>, <em>B2</em>, <em>A2</em>, <em>B3</em>, <em>B4</em> with <em>A</em> representing QC samples and <em>B</em> study samples, using <code>subset = c(1, 4)</code> and <code>subsetAdjust = "previous"</code> would result in all <em>A</em> samples to be aligned with each other and non-subset samples <em>B1</em> and <em>B2</em> being adjusted based on the alignment result of subset samples <em>A1</em> and <em>B3</em> and <em>B4</em> on those of <em>A2</em>.</p></li>
<li><p><code>subsetAdjust = "average"</code>: adjust retention times of non-subset samples based on an interpolation of the alignment results of the previous and subsequent subset sample. In the example above, <em>B1</em> would be adjusted based on the average of adjusted retention times between subset (QC) samples <em>A1</em> and <em>A2</em>. Since there is no subset sample after non-subset samples <em>B3</em> and <em>B4</em> these will be adjusted based on the alignment results of <em>A2</em> alone. Note that a weighted average is used to calculate the adjusted retention time averages, which uses the inverse of the difference of the index of the non-subset sample to the subset samples as weights. Thus, if we have a setup like <em>A1</em>, <em>B1</em>, <em>B2</em>, <em>A2</em> the adjusted retention times of <em>A1</em> would get a larger weight than those of <em>A2</em> in the adjustment of non-subset sample <em>B1</em> causing it’s adjusted retention times to be closer to those of <em>A1</em> than to <em>A2</em>. See below for examples.</p></li>
</ul>
<p>Both cases require a meaningful/correct ordering of the samples within the object (e.g. ordering by injection index).</p>
<p>The examples below aim to illustrate the effect of these alignment options. First we assume samples 1, 6 and 11 in the <em>faahKO</em> data set to be QC samples (sample pools). We thus want to perform the alignment based on these samples and subsequently adjust the retention times of the left-out samples (2, 3, 4, 5, 7, 8, 9, 10 and 12) based on interpolation of the results from the neighboring <em>subset</em> samples. After initial peak grouping we perform below the alignment with the <em>peak groups</em> method passing the indices of the samples on which we want the alignment to be based with the <code>subset</code> argument and specify <code>subsetAdjust = "average"</code> to adjust the study samples based on interpolation of the alignment results from neighboring subset/QC samples.</p>
<p>Note that for any subset-alignment all parameters such as <code>minFraction</code> are relative to the <code>subset</code>, not the full experiment!</p>
<div class="sourceCode" id="cb56"><html><body><pre class="r"><span class="no">xdata_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">dropAdjustedRtime</a></span>(<span class="no">xdata</span>)
<span class="co">## Define the experimental layout</span>
<span class="no">xdata_subs</span>$<span class="no">sample_type</span> <span class="kw">&lt;-</span> <span class="st">"study"</span>
<span class="no">xdata_subs</span>$<span class="no">sample_type</span>[<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">6</span>, <span class="fl">11</span>)] <span class="kw">&lt;-</span> <span class="st">"QC"</span>

<span class="co">## Initial peak grouping. Use sample_type as grouping variable</span>
<span class="no">pdp_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span>(<span class="kw">sampleGroups</span> <span class="kw">=</span> <span class="no">xdata_subs</span>$<span class="no">sample_type</span>,
                             <span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.9</span>)
<span class="no">xdata_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span>(<span class="no">xdata_subs</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pdp_subs</span>)

<span class="co">## Define subset-alignment options and perform the alignment</span>
<span class="no">pgp_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime-peakGroups.html">PeakGroupsParam</a></span>(<span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.85</span>,
                            <span class="kw">subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">xdata_subs</span>$<span class="no">sample_type</span> <span class="kw">==</span> <span class="st">"QC"</span>),
                            <span class="kw">subsetAdjust</span> <span class="kw">=</span> <span class="st">"average"</span>, <span class="kw">span</span> <span class="kw">=</span> <span class="fl">0.4</span>)
<span class="no">xdata_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span>(<span class="no">xdata_subs</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pgp_subs</span>)</pre></body></html></div>
<p>To illustrate the alignment results we plot the results only for samples 1 to 6 (1 and 6 being QC samples and 2 to 5 study samples that were adjusted based on the QC samples). This nicely shows how the interpolation of the <code>subsetAdjust = "average"</code> works: retention times of sample 2 are adjusted based on those from subset sample 1 and 6, giving however more weight to the closer subset sample 1 which results in the adjusted retention times of 2 being more similar to those of sample 1. Sample 5 on the other hand gets adjusted giving more weight to the second subset sample (6).</p>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="no">clrs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">NA</span>, <span class="fl">12</span>)
<span class="no">clrs</span>[<span class="fl">1</span>:<span class="fl">6</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">6</span>, <span class="st">"Dark2"</span>), <span class="fl">80</span>)

<span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span>(<span class="no">xdata_subs</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">clrs</span>, <span class="kw">lwd</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">peakGroupsCol</span> <span class="kw">=</span> <span class="fl">NA</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topleft"</span>, <span class="kw">legend</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">6</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">6</span>, <span class="st">"Dark2"</span>), <span class="kw">lwd</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-subset-plot-1-1.png" alt="Alignment results for samples 1 to 6." width="960"><p class="caption">
Alignment results for samples 1 to 6.
</p>
</div>
<p>We also plot the full alignment results, drawing the <code>subset</code> samples as green and study samples as grey lines.</p>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="no">clrs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"#00000040"</span>, <span class="fl">12</span>)
<span class="no">clrs</span>[<span class="no">xdata_subs</span>$<span class="no">sample_type</span> <span class="kw">==</span> <span class="st">"QC"</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#00ce0080"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">1</span>), <span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">1</span>, <span class="fl">0.5</span>))
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata_subs</span>, <span class="kw">aggregationFun</span> <span class="kw">=</span> <span class="st">"sum"</span>),
     <span class="kw">col</span> <span class="kw">=</span> <span class="no">clrs</span>, <span class="kw">peakType</span> <span class="kw">=</span> <span class="st">"none"</span>)
<span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span>(<span class="no">xdata_subs</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">clrs</span>, <span class="kw">peakGroupsPch</span> <span class="kw">=</span> <span class="fl">1</span>,
                  <span class="kw">peakGroupsCol</span> <span class="kw">=</span> <span class="st">"#00ce0040"</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-subset-plot-2-1.png" alt="Subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey." width="960"><p class="caption">
Subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey.
</p>
</div>
<p>Option <code>subsetAdjust = "previous"</code> adjusts the retention times of a non-subset sample based on a single subset sample (the previous), which results in most cases in the adjusted retention times of the non-subset sample being highly similar to those of the subset sample which was used for adjustment.</p>
<p>Also obiwarp alignment supports the same types of subset-alignment as shown below.</p>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="no">xdata_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">dropAdjustedRtime</a></span>(<span class="no">xdata_subs</span>)</pre></body></html></div>
<pre><code>## Reverting retention times of identified peaks to original values ... OK</code></pre>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="no">prm</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime-obiwarp.html">ObiwarpParam</a></span>(<span class="kw">subset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">xdata_subs</span>$<span class="no">sample_type</span> <span class="kw">==</span> <span class="st">"QC"</span>),
                    <span class="kw">subsetAdjust</span> <span class="kw">=</span> <span class="st">"average"</span>)
<span class="no">xdata_subs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span>(<span class="no">xdata_subs</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">prm</span>)</pre></body></html></div>
<pre><code>## Sample number 2 used as center sample.
## Aligning ko15.CDF against ko22.CDF ... OK
## Aligning wt21.CDF against ko22.CDF ... OK
## Aligning sample number 2 against subset ... OK
## Aligning sample number 3 against subset ... OK
## Aligning sample number 4 against subset ... OK
## Aligning sample number 5 against subset ... OK
## Aligning sample number 7 against subset ... OK
## Aligning sample number 8 against subset ... OK
## Aligning sample number 9 against subset ... OK
## Aligning sample number 10 against subset ... OK
## Aligning sample number 12 against subset ... OK
## Applying retention time adjustment to the identified chromatographic peaks ... OK</code></pre>
<p>We again plot the results.</p>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="no">clrs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"#00000040"</span>, <span class="fl">12</span>)
<span class="no">clrs</span>[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">xdata_subs</span>$<span class="no">sample_type</span> <span class="kw">==</span> <span class="st">"QC"</span>)] <span class="kw">&lt;-</span> <span class="st">"#00ce0080"</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">1</span>), <span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">1</span>, <span class="fl">0.5</span>))
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata_subs</span>, <span class="kw">aggregationFun</span> <span class="kw">=</span> <span class="st">"sum"</span>),
     <span class="kw">col</span> <span class="kw">=</span> <span class="no">clrs</span>, <span class="kw">peakType</span> <span class="kw">=</span> <span class="st">"none"</span>)
<span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span>(<span class="no">xdata_subs</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">clrs</span>, <span class="kw">peakGroupsPch</span> <span class="kw">=</span> <span class="fl">1</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-subset-obiwarp-plot-1.png" alt="Obiwarp subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey." width="960"><p class="caption">
Obiwarp subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey.
</p>
</div>
</div>
</div>
<div id="correspondence" class="section level1">
<h1 class="hasAnchor">
<a href="#correspondence" class="anchor"></a>Correspondence</h1>
<p>The final step in the metabolomics preprocessing is the correspondence that matches detected chromatographic peaks between samples (and depending on the settings, also within samples if they are adjacent). The method to perform the correspondence in <code>xcms</code> is <code>groupChromPeaks</code>. We will use the <em>peak density</em> method <span class="citation">[5]</span> to group chromatographic peaks. The algorithm combines chromatographic peaks depending on the density of peaks along the retention time axis within small slices along the mz dimension. To illustrate this we plot below the chromatogram for an mz slice with multiple chromatographic peaks within each sample. We use below a value of 0.4 for the <code>minFraction</code> parameter hence only chromatographic peaks present in at least 40% of the samples per sample group are grouped into a feature. The sample group assignment is specified with the <code>sampleGroups</code> argument.</p>
<div class="sourceCode" id="cb64"><html><body><pre class="r"><span class="co">## Define the mz slice.</span>
<span class="no">mzr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">305.05</span>, <span class="fl">305.15</span>)

<span class="co">## Extract and plot the chromatograms</span>
<span class="no">chr_mzr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span>(<span class="no">xdata</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2500</span>, <span class="fl">4000</span>))
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">1</span>), <span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">1</span>, <span class="fl">0.5</span>))
<span class="no">cols</span> <span class="kw">&lt;-</span> <span class="no">group_colors</span>[<span class="no">chr_mzr</span>$<span class="no">sample_group</span>]
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">chr_mzr</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">cols</span>, <span class="kw">xaxt</span> <span class="kw">=</span> <span class="st">"n"</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">""</span>,
     <span class="kw">peakBg</span> <span class="kw">=</span> <span class="no">sample_colors</span>[<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">chr_mzr</span>)[, <span class="st">"sample"</span>]])
<span class="co">## Define the parameters for the peak density method</span>
<span class="no">pdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span>(<span class="kw">sampleGroups</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_group</span>,
                        <span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.4</span>, <span class="kw">bw</span> <span class="kw">=</span> <span class="fl">30</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">1</span>, <span class="fl">0.5</span>))
<span class="fu"><a href="../reference/plotChromPeakDensity.html">plotChromPeakDensity</a></span>(<span class="no">xdata</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">cols</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pdp</span>,
                     <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">16</span>, <span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2500</span>, <span class="fl">4000</span>))</pre></body></html></div>
<pre><code>## Warning: Use of 'plotChromPeakDensity' on 'XCMSnExp' isdiscouraged. Please
## extract chromatographic data first and call 'plotChromPeakDensity' directly
## on the 'XChromatograms' object. See ?XChromatograms, section 'Correspondence
## analysis' for more details.</code></pre>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="co">## Use a different bw</span>
<span class="no">pdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span>(<span class="kw">sampleGroups</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_group</span>,
                        <span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.4</span>, <span class="kw">bw</span> <span class="kw">=</span> <span class="fl">20</span>)
<span class="fu"><a href="../reference/plotChromPeakDensity.html">plotChromPeakDensity</a></span>(<span class="no">xdata</span>, <span class="kw">mz</span> <span class="kw">=</span> <span class="no">mzr</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">cols</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pdp</span>,
                     <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">16</span>, <span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2500</span>, <span class="fl">4000</span>))</pre></body></html></div>
<pre><code>## Warning: Use of 'plotChromPeakDensity' on 'XCMSnExp' isdiscouraged. Please
## extract chromatographic data first and call 'plotChromPeakDensity' directly
## on the 'XChromatograms' object. See ?XChromatograms, section 'Correspondence
## analysis' for more details.</code></pre>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/correspondence-example-slice-1.png" alt="Example for peak density correspondence. Upper panel: chromatogram for an mz slice with multiple chromatographic peaks. Middle and lower panel: identified chromatographic peaks at their retention time (x-axis) and index within samples of the experiments (y-axis) for different values of the bw parameter. The black line represents the peak density estimate. Grouping of peaks (based on the provided settings) is indicated by grey rectangles." width="960"><p class="caption">
Example for peak density correspondence. Upper panel: chromatogram for an mz slice with multiple chromatographic peaks. Middle and lower panel: identified chromatographic peaks at their retention time (x-axis) and index within samples of the experiments (y-axis) for different values of the bw parameter. The black line represents the peak density estimate. Grouping of peaks (based on the provided settings) is indicated by grey rectangles.
</p>
</div>
<p>The upper panel in the plot above shows the extracted ion chromatogram for each sample with the detected peaks highlighted. The middle and lower plot shows the retention time for each detected peak within the different samples. The black solid line represents the density distribution of detected peaks along the retention times. Peaks combined into <em>features</em> (peak groups) are indicated with grey rectangles. Different values for the <code>bw</code> parameter of the <code>PeakDensityParam</code> were used:<code>bw = 30</code> in the middle and <code>bw = 20</code> in the lower panel. With the default value for the parameter <code>bw</code> the two neighboring chromatographic peaks would be grouped into the same feature, while with a <code>bw</code> of 20 they would be grouped into separate features. This grouping depends on the parameters for the density function and other parameters passed to the algorithm with the <code>PeakDensityParam</code>.</p>
<div class="sourceCode" id="cb68"><html><body><pre class="r"><span class="co">## Perform the correspondence</span>
<span class="no">pdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span>(<span class="kw">sampleGroups</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_group</span>,
                        <span class="kw">minFraction</span> <span class="kw">=</span> <span class="fl">0.4</span>, <span class="kw">bw</span> <span class="kw">=</span> <span class="fl">20</span>)
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span>(<span class="no">xdata</span>, <span class="kw">param</span> <span class="kw">=</span> <span class="no">pdp</span>)</pre></body></html></div>
<p>Results from the xcms-based preprocessing can be summarized into a <code>SummarizedExperiment</code> object from the <em><a href="https://bioconductor.org/packages/3.11/SummarizedExperiment">SummarizedExperiment</a></em> package with the <code>quantify</code> method. This object will contain the feature abundances as the <em>assay matrix</em>, the feature definition (their m/z, retention time and other metadata) as <code>rowData</code> (i.e. row annotations) and the sample/phenotype information as <code>colData</code> (i.e. column annotations). All the processing history will be put into the object’s <code>metadata</code>. This object can then be used for any further (<code>xcms</code>-independent) processing and analysis.</p>
<p>Below we use <code>quantify</code> to generate the result object for the present analysis. The parameters <code>value</code> and any other additional parameters are passed along to the <code>featureValues</code> method that is used internally to create the feature abundance matrix.</p>
<div class="sourceCode" id="cb69"><html><body><pre class="r"><span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu">quantify</span>(<span class="no">xdata</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="st">"into"</span>)</pre></body></html></div>
<p>Sample annotations can be accessed with the <code>colData</code> method.</p>
<div class="sourceCode" id="cb70"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span>(<span class="no">res</span>)</pre></body></html></div>
<pre><code>## DataFrame with 12 rows and 2 columns
##          sample_name sample_group
##          &lt;character&gt;  &lt;character&gt;
## ko15.CDF        ko15           KO
## ko16.CDF        ko16           KO
## ko18.CDF        ko18           KO
## ko19.CDF        ko19           KO
## ko21.CDF        ko21           KO
## ...              ...          ...
## wt16.CDF        wt16           WT
## wt18.CDF        wt18           WT
## wt19.CDF        wt19           WT
## wt21.CDF        wt21           WT
## wt22.CDF        wt22           WT</code></pre>
<p>Feature annotations with <code>rowData</code>:</p>
<div class="sourceCode" id="cb72"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">rowData</a></span>(<span class="no">res</span>)</pre></body></html></div>
<pre><code>## DataFrame with 333 rows and 11 columns
##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks
##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## FT001     205.0     205.0     205.0   2789.05   2787.04   2790.55        12
## FT002     206.0     206.0     206.0   2788.69   2785.64   2791.90        11
## FT003     207.1     207.1     207.1   2718.71   2711.13   2721.93         9
## FT004     233.0     233.0     233.0   3022.91   3014.33   3061.48        10
## FT005     233.1     233.1     233.1   3023.59   3017.84   3031.48         5
## ...         ...       ...       ...       ...       ...       ...       ...
## FT329     594.2     594.2     594.2   3403.38   3376.60   3410.87         6
## FT330     594.3     594.3     594.4   3615.82   3608.82   3662.80        11
## FT331     595.2     595.2     595.3   3005.89   2998.65   3017.55         8
## FT332     596.3     596.3     596.4   3817.96   3803.02   3826.67        13
## FT333     597.4     597.3     597.4   3817.93   3812.15   3828.09         6
##              KO        WT            peakidx  ms_level
##       &lt;numeric&gt; &lt;numeric&gt;             &lt;list&gt; &lt;integer&gt;
## FT001         6         6   58, 526,1171,...         1
## FT002         5         6   41, 512,1157,...         1
## FT003         5         4   25, 499,1146,...         1
## FT004         4         5   92,1205,1844,...         1
## FT005         4         1   88,1201,1847,...         1
## ...         ...       ...                ...       ...
## FT329         1         5 1910,2574,3392,...         1
## FT330         5         3    316,893,902,...         1
## FT331         2         4     81, 86,639,...         1
## FT332         6         4  419,1053,1463,...         1
## FT333         2         3 1049,1466,3537,...         1</code></pre>
<p>The feature abundances can be accessed with the <code>assay</code> method. Note also that a <code>SummarizedExperiment</code> supports multiple such assay matrices.</p>
<div class="sourceCode" id="cb74"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span>(<span class="no">res</span>))</pre></body></html></div>
<pre><code>##        ko15.CDF  ko16.CDF  ko18.CDF   ko19.CDF  ko21.CDF  ko22.CDF  wt15.CDF
## FT001 1924712.0 1757151.0 1714581.8 1220358.09 1383416.7 1180288.2 2129885.1
## FT002  213659.3  289500.7  194603.7   92590.27        NA  178285.7  253825.6
## FT003  349011.5  451863.7  337473.3         NA  343897.8  208002.8  364609.8
## FT004  286221.4        NA  364299.5         NA  137261.0  149097.6  255697.7
## FT005  162905.3        NA  209999.6         NA  164009.0  111157.9        NA
## FT006 1160580.5        NA 1345515.1  608016.51  380970.3  588986.4 1286883.0
##        wt16.CDF  wt18.CDF  wt19.CDF  wt21.CDF  wt22.CDF
## FT001 1634342.0 1810112.3 1507943.1 1623589.2 1354004.9
## FT002  241844.4  228501.1  216393.1  240606.0  185399.5
## FT003  360908.9  182162.0        NA        NA  221937.5
## FT004  311296.8  272267.9        NA  181640.2  271128.0
## FT005        NA        NA        NA  366441.5        NA
## FT006 1739516.6  515676.9  621437.9  639755.3  508546.4</code></pre>
<p>In addition it is possible to extract the results from the correspondence analysis individually using the <code>featureDefinitions</code> and <code>featureValues</code> methods, that former returning a <code>DataFrame</code> with the definition of the features (i.e. the mz and rt ranges and, in column <code>peakidx</code>, the index of the chromatographic peaks in the <code>chromPeaks</code> matrix for each feature), the latter the feature abundances.</p>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="co">## Extract the feature definitions</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">featureDefinitions</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<pre><code>## DataFrame with 333 rows and 11 columns
##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks
##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## FT001     205.0     205.0     205.0   2789.05   2787.04   2790.55        12
## FT002     206.0     206.0     206.0   2788.69   2785.64   2791.90        11
## FT003     207.1     207.1     207.1   2718.71   2711.13   2721.93         9
## FT004     233.0     233.0     233.0   3022.91   3014.33   3061.48        10
## FT005     233.1     233.1     233.1   3023.59   3017.84   3031.48         5
## ...         ...       ...       ...       ...       ...       ...       ...
## FT329     594.2     594.2     594.2   3403.38   3376.60   3410.87         6
## FT330     594.3     594.3     594.4   3615.82   3608.82   3662.80        11
## FT331     595.2     595.2     595.3   3005.89   2998.65   3017.55         8
## FT332     596.3     596.3     596.4   3817.96   3803.02   3826.67        13
## FT333     597.4     597.3     597.4   3817.93   3812.15   3828.09         6
##              KO        WT            peakidx  ms_level
##       &lt;numeric&gt; &lt;numeric&gt;             &lt;list&gt; &lt;integer&gt;
## FT001         6         6   58, 526,1171,...         1
## FT002         5         6   41, 512,1157,...         1
## FT003         5         4   25, 499,1146,...         1
## FT004         4         5   92,1205,1844,...         1
## FT005         4         1   88,1201,1847,...         1
## ...         ...       ...                ...       ...
## FT329         1         5 1910,2574,3392,...         1
## FT330         5         3    316,893,902,...         1
## FT331         2         4     81, 86,639,...         1
## FT332         6         4  419,1053,1463,...         1
## FT333         2         3 1049,1466,3537,...         1</code></pre>
<p>The <code>featureValues</code> method returns a <code>matrix</code> with rows being features and columns samples. The content of this matrix can be defined using the <code>value</code> argument. The default is<code>value = "index"</code> which simply returns the index of the peak (in the <code>chromPeaks</code> matrix) assigned to a feature in a sample. Setting <code>value = "into"</code> returns a matrix with the integrated signal of the peaks corresponding to a feature in a sample. Any column name of the <code>chromPeaks</code> matrix can be passed to the argument <code>value</code>. Below we extract the integrated peak intensity per feature/sample.</p>
<div class="sourceCode" id="cb78"><html><body><pre class="r"><span class="co">## Extract the into column for each feature.</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span>(<span class="no">xdata</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="st">"into"</span>))</pre></body></html></div>
<pre><code>##        ko15.CDF  ko16.CDF  ko18.CDF   ko19.CDF  ko21.CDF  ko22.CDF  wt15.CDF
## FT001 1924712.0 1757151.0 1714581.8 1220358.09 1383416.7 1180288.2 2129885.1
## FT002  213659.3  289500.7  194603.7   92590.27        NA  178285.7  253825.6
## FT003  349011.5  451863.7  337473.3         NA  343897.8  208002.8  364609.8
## FT004  286221.4        NA  364299.5         NA  137261.0  149097.6  255697.7
## FT005  162905.3        NA  209999.6         NA  164009.0  111157.9        NA
## FT006 1160580.5        NA 1345515.1  608016.51  380970.3  588986.4 1286883.0
##        wt16.CDF  wt18.CDF  wt19.CDF  wt21.CDF  wt22.CDF
## FT001 1634342.0 1810112.3 1507943.1 1623589.2 1354004.9
## FT002  241844.4  228501.1  216393.1  240606.0  185399.5
## FT003  360908.9  182162.0        NA        NA  221937.5
## FT004  311296.8  272267.9        NA  181640.2  271128.0
## FT005        NA        NA        NA  366441.5        NA
## FT006 1739516.6  515676.9  621437.9  639755.3  508546.4</code></pre>
<p>This feature matrix contains <code>NA</code> for samples in which no chromatographic peak was detected in the feature’s m/z-rt region. While in many cases there might indeed be no peak signal in the respective region, it might also be that there is signal, but the peak detection algorithm failed to detect a chromatographic peak. <code>xcms</code> provides the <code>fillChromPeaks</code> method to <em>fill in</em> intensity data for such missing values from the original files. The <em>filled in</em> peaks are added to the <code>chromPeaks</code> matrix and get a <code>TRUE</code> value in the <code>"is_filled"</code> column of the <code>chromPeakData</code> data frame. Below we perform such a filling-in of missing peaks.</p>
<div class="sourceCode" id="cb80"><html><body><pre class="r"><span class="co">## Filling missing peaks using default settings. Alternatively we could</span>
<span class="co">## pass a FillChromPeaksParam object to the method.</span>
<span class="no">xdata</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fillChromPeaks.html">fillChromPeaks</a></span>(<span class="no">xdata</span>)

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span>(<span class="no">xdata</span>))</pre></body></html></div>
<pre><code>##        ko15.CDF  ko16.CDF  ko18.CDF   ko19.CDF  ko21.CDF  ko22.CDF  wt15.CDF
## FT001 1924712.0 1757151.0 1714581.8 1220358.09 1383416.7 1180288.2 2129885.1
## FT002  213659.3  289500.7  194603.7   92590.27  156117.9  178285.7  253825.6
## FT003  349011.5  451863.7  337473.3   96651.64  343897.8  208002.8  364609.8
## FT004  286221.4  292059.8  364299.5  205221.99  137261.0  149097.6  255697.7
## FT005  162905.3  142448.4  209999.6   53040.36  164009.0  111157.9  213633.9
## FT006 1160580.5 1077370.9 1345515.1  608016.51  380970.3  588986.4 1286883.0
##         wt16.CDF  wt18.CDF  wt19.CDF  wt21.CDF   wt22.CDF
## FT001 1634341.99 1810112.3 1507943.1 1623589.2 1354004.93
## FT002  241844.44  228501.1  216393.1  240606.0  185399.47
## FT003  360908.93  182162.0  151129.0  233471.4  221937.53
## FT004  311296.82  272267.9  120833.4  181640.2  271128.02
## FT005   40742.77  172659.7  111011.3  366441.5   68987.67
## FT006 1739516.56  515676.9  621437.9  639755.3  508546.42</code></pre>
<p>For features without detected peaks in a sample, the method extracts all intensities in the mz-rt region of the feature, integrates the signal and adds a <em>filled-in</em> peak to the <code>chromPeaks</code> matrix. No peak is added if no signal is measured/available for the mz-rt region of the feature. For these, even after filling in missing peak data, a <code>NA</code> is reported in the <code>featureValues</code> matrix.</p>
<p>It should be mentioned that <code>fillChromPeaks</code> uses columns <code>"rtmin"</code>, <code>"rtmax"</code>, <code>"mzmin"</code> and <code>"mzmax"</code> from the <code>featureDefinitions</code> table to define the region from which the signal should be integrated to generate the filled-in peak signal. These values correspond however to the positions of the peak apex not the peak boundaries of all chromatographic peaks assigned to a feature. It might be advisable to increase this area in retention time dimension by a constant value appropriate to the average peak width in the experiment. Such a value can be specified with <code>fixedRt</code> of the <code>FillChromPeaksParam</code>. If the average peak width in the experiment is 20 seconds, specifying<code>fixedRt = 10</code> ensures that the area from which peaks are integrated is at least 20 seconds wide.</p>
<p>Below we compare the number of missing values before and after filling in missing values. We can use the parameter <code>filled</code> of the <code>featureValues</code> method to define whether or not filled-in peak values should be returned too.</p>
<div class="sourceCode" id="cb82"><html><body><pre class="r"><span class="co">## Missing values before filling in peaks</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span>(<span class="no">xdata</span>, <span class="kw">filled</span> <span class="kw">=</span> <span class="fl">FALSE</span>), <span class="kw">MARGIN</span> <span class="kw">=</span> <span class="fl">2</span>,
      <span class="kw">FUN</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">z</span>) <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">z</span>)))</pre></body></html></div>
<pre><code>## ko15.CDF ko16.CDF ko18.CDF ko19.CDF ko21.CDF ko22.CDF wt15.CDF wt16.CDF 
##       68       70       65      109      130      105       95       88 
## wt18.CDF wt19.CDF wt21.CDF wt22.CDF 
##       90      114      123      106</code></pre>
<div class="sourceCode" id="cb84"><html><body><pre class="r"><span class="co">## Missing values after filling in peaks</span>
<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span>(<span class="no">xdata</span>), <span class="kw">MARGIN</span> <span class="kw">=</span> <span class="fl">2</span>,
      <span class="kw">FUN</span> <span class="kw">=</span> <span class="kw">function</span>(<span class="no">z</span>) <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">z</span>)))</pre></body></html></div>
<pre><code>## ko15.CDF ko16.CDF ko18.CDF ko19.CDF ko21.CDF ko22.CDF wt15.CDF wt16.CDF 
##        9        5        3        8       12        6        8        8 
## wt18.CDF wt19.CDF wt21.CDF wt22.CDF 
##        7       13       14        6</code></pre>
<p>Next we use the <code>featureSummary</code> function to get a general per-feature summary that includes the number of samples in which a peak was found or the number of samples in which more than one peak was assigned to the feature. Specifying also sample groups breaks down these summary statistics for each individual sample group.</p>
<div class="sourceCode" id="cb86"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/featureSummary.html">featureSummary</a></span>(<span class="no">xdata</span>, <span class="kw">group</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_group</span>))</pre></body></html></div>
<pre><code>##       count      perc multi_count multi_perc       rsd KO_count   KO_perc
## FT001    12 100.00000           0    0.00000 0.1789898        6 100.00000
## FT002    11  91.66667           0    0.00000 0.2409441        5  83.33333
## FT003     9  75.00000           0    0.00000 0.2844117        5  83.33333
## FT004     9  75.00000           1   11.11111 0.3082528        4  66.66667
## FT005     5  41.66667           0    0.00000 0.4824173        4  66.66667
## FT006    11  91.66667           2   18.18182 0.5229604        5  83.33333
##       KO_multi_count KO_multi_perc    KO_rsd WT_count   WT_perc WT_multi_count
## FT001              0             0 0.2027357        6 100.00000              0
## FT002              0             0 0.3653441        6 100.00000              0
## FT003              0             0 0.2562703        4  66.66667              0
## FT004              0             0 0.4694612        5  83.33333              1
## FT005              0             0 0.2492851        1  16.66667              0
## FT006              1            20 0.5059513        6 100.00000              1
##       WT_multi_perc    WT_rsd
## FT001       0.00000 0.1601279
## FT002       0.00000 0.1069526
## FT003       0.00000 0.3335925
## FT004      20.00000 0.1840914
## FT005       0.00000        NA
## FT006      16.66667 0.5758382</code></pre>
<p>The performance of peak detection, alignment and correspondence should always be evaluated by inspecting extracted ion chromatograms e.g. of known compounds, internal standards or identified features in general. The <code>featureChromatograms</code> function allows to extract chromatograms for each feature present in <code>featureDefinitions</code>. The returned <code>Chromatograms</code> object contains an ion chromatogram for each feature (each row containing the data for one feature) and sample (each column representing containing data for one sample). Below we extract the chromatograms for the first 4 features.</p>
<div class="sourceCode" id="cb88"><html><body><pre class="r"><span class="no">feature_chroms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/featureChromatograms.html">featureChromatograms</a></span>(<span class="no">xdata</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">4</span>)

<span class="no">feature_chroms</span></pre></body></html></div>
<pre><code>## XChromatograms with 4 rows and 12 columns
##                    1               2               3               4
##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;
## [1,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [3,]        peaks: 1        peaks: 1        peaks: 1        peaks: 0
## [4,]        peaks: 1        peaks: 0        peaks: 1        peaks: 0
##                    5               6               7               8
##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;
## [1,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [2,]        peaks: 0        peaks: 1        peaks: 1        peaks: 1
## [3,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [4,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
##                    9              10              11              12
##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;
## [1,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1
## [3,]        peaks: 1        peaks: 0        peaks: 0        peaks: 1
## [4,]        peaks: 1        peaks: 0        peaks: 2        peaks: 1
## phenoData with 2 variables
## featureData with 5 variables
## - - - xcms preprocessing - - -
## Chromatographic peak detection:
##  method: centWave 
## Correspondence:
##  method: chromatographic peak density 
##  4 feature(s) identified.</code></pre>
<p>And plot the extracted ion chromatograms. We again use the group color for each identified peak to fill the area.</p>
<div class="sourceCode" id="cb90"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">feature_chroms</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">sample_colors</span>,
     <span class="kw">peakBg</span> <span class="kw">=</span> <span class="no">sample_colors</span>[<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">feature_chroms</span>)[, <span class="st">"sample"</span>]])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/feature-eic-1.png" alt="Extracted ion chromatograms for features 1 to 4." width="768"><p class="caption">
Extracted ion chromatograms for features 1 to 4.
</p>
</div>
<div class="sourceCode" id="cb91"><html><body><pre class="r"><span class="no">third</span> <span class="kw">&lt;-</span> <span class="no">feature_chroms</span>[<span class="fl">3</span>, ]
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">third</span>)
<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">third</span>)</pre></body></html></div>
<pre><code>##           mz mzmin mzmax       rt    rtmin    rtmax     into     intb  maxo sn
## CP0030 207.1 207.1 207.1 2717.287 2700.120 2740.550 349011.5 332229.1 18800 17
## CP0561 207.1 207.1 207.1 2718.925 2704.969 2736.062 451863.7 436882.9 22024 27
## CP1253 207.1 207.1 207.1 2717.047 2703.711 2747.103 337473.3 335177.5 14035 47
## CP1980 207.1 207.1 207.1 2719.651 2696.360 2750.715 343897.8 314663.2 10746 12
## CP2246 207.1 207.1 207.1 2721.932 2706.187 2740.914 208002.8 193780.3 10176 12
## CP2575 207.1 207.1 207.1 2721.641 2695.165 2746.272 364609.8 343754.9 18280 19
## CP3095 207.1 207.1 207.1 2718.712 2696.417 2740.634 360908.9 331504.1 16119 19
## CP3575 207.1 207.1 207.1 2711.131 2692.806 2725.409 182162.0 165026.9 10354 13
## CP4638 207.1 207.1 207.1 2717.360 2702.484 2737.485 221937.5 211243.0 10195 13
##        sample row column
## CP0030      1   1      1
## CP0561      2   1      2
## CP1253      3   1      3
## CP1980      5   1      5
## CP2246      6   1      6
## CP2575      7   1      7
## CP3095      8   1      8
## CP3575      9   1      9
## CP4638     12   1     12</code></pre>
<div class="sourceCode" id="cb93"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">feature_chroms</span>[<span class="fl">3</span>, ], <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">feature_chroms</span>$<span class="no">sample_group</span>])
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">feature_chroms</span>[<span class="fl">4</span>, ], <span class="kw">col</span> <span class="kw">=</span> <span class="no">group_colors</span>[<span class="no">feature_chroms</span>$<span class="no">sample_group</span>])</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/feature-eic-2.png" alt="Extracted ion chromatograms for features 1 to 4." width="768"><p class="caption">
Extracted ion chromatograms for features 1 to 4.
</p>
</div>
<p>At last we perform a principal component analysis to evaluate the grouping of the samples in this experiment. Note that we did not perform any data normalization hence the grouping might (and will) also be influenced by technical biases.</p>
<div class="sourceCode" id="cb94"><html><body><pre class="r"><span class="co">## Extract the features and log2 transform them</span>
<span class="no">ft_ints</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span>(<span class="no">xdata</span>, <span class="kw">value</span> <span class="kw">=</span> <span class="st">"into"</span>))

<span class="co">## Perform the PCA omitting all features with an NA in any of the</span>
<span class="co">## samples. Also, the intensities are mean centered.</span>
<span class="no">pc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html">na.omit</a></span>(<span class="no">ft_ints</span>)), <span class="kw">center</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="co">## Plot the PCA</span>
<span class="no">cols</span> <span class="kw">&lt;-</span> <span class="no">group_colors</span>[<span class="no">xdata</span>$<span class="no">sample_group</span>]
<span class="no">pcSummary</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">pc</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pc</span>$<span class="no">x</span>[, <span class="fl">1</span>], <span class="no">pc</span>$<span class="no">x</span>[,<span class="fl">2</span>], <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">21</span>, <span class="kw">main</span> <span class="kw">=</span> <span class="st">""</span>,
     <span class="kw">xlab</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"PC1: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span>(<span class="no">pcSummary</span>$<span class="no">importance</span>[<span class="fl">2</span>, <span class="fl">1</span>] * <span class="fl">100</span>,
                                   <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">3</span>), <span class="st">" % variance"</span>),
     <span class="kw">ylab</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"PC2: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/format.html">format</a></span>(<span class="no">pcSummary</span>$<span class="no">importance</span>[<span class="fl">2</span>, <span class="fl">2</span>] * <span class="fl">100</span>,
                                   <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">3</span>), <span class="st">" % variance"</span>),
     <span class="kw">col</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>, <span class="kw">bg</span> <span class="kw">=</span> <span class="no">cols</span>, <span class="kw">cex</span> <span class="kw">=</span> <span class="fl">2</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/grid.html">grid</a></span>()
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="no">pc</span>$<span class="no">x</span>[, <span class="fl">1</span>], <span class="no">pc</span>$<span class="no">x</span>[,<span class="fl">2</span>], <span class="kw">labels</span> <span class="kw">=</span> <span class="no">xdata</span>$<span class="no">sample_name</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="st">"darkgrey"</span>,
     <span class="kw">pos</span> <span class="kw">=</span> <span class="fl">3</span>, <span class="kw">cex</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/final-pca-1.png" alt="PCA for the faahKO data set, un-normalized intensities." width="768"><p class="caption">
PCA for the faahKO data set, un-normalized intensities.
</p>
</div>
<p>We can see the expected separation between the KO and WT samples on PC2. On PC1 samples separate based on their ID, samples with an ID &lt;= 18 from samples with an ID &gt; 18. This separation might be caused by a technical bias (e.g. measurements performed on different days/weeks) or due to biological properties of the mice analyzed (sex, age, litter mates etc).</p>
</div>
<div id="further-data-processing-and-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#further-data-processing-and-analysis" class="anchor"></a>Further data processing and analysis</h1>
<p>Normalizing features’ signal intensities is required, but at present not (yet) supported in <code>xcms</code> (some methods might be added in near future). It is advised to use the <code>SummarizedExperiment</code> returned by the <code>quantify</code> method for any further data processing, as this type of object stores feature definitions, sample annotations as well as feature abundances in the same object. For the identification of e.g. features with significant different intensities/abundances it is suggested to use functionality provided in other R packages, such as Bioconductor’s excellent <code>limma</code> package. To enable support also for other packages that rely on the <em>old</em> <code>xcmsSet</code> result object, it is possible to coerce the new <code>XCMSnExp</code> object to an <code>xcmsSet</code> object using <code>xset &lt;- as(x, "xcmsSet")</code>, with <code>x</code> being an <code>XCMSnExp</code> object.</p>
</div>
<div id="additional-details-and-notes" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-details-and-notes" class="anchor"></a>Additional details and notes</h1>
<p>For a detailed description of the new data objects and changes/improvements compared to the original user interface see the <em>new_functionality</em> vignette.</p>
<div id="evaluating-the-process-history" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluating-the-process-history" class="anchor"></a>Evaluating the process history</h2>
<p><code>XCMSnExp</code> objects allow to capture all performed pre-processing steps along with the used parameter class within the <code>@processHistory</code> slot. Storing also the parameter class ensures the highest possible degree of analysis documentation and in future might enable to <em>replay</em> analyses or parts of it. The list of all performed preprocessings can be extracted using the <code>processHistory</code> method.</p>
<div class="sourceCode" id="cb95"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span>(<span class="no">xdata</span>)</pre></body></html></div>
<pre><code>## [[1]]
## Object of class "XProcessHistory"
##  type: Peak detection 
##  date: Wed May 20 05:41:37 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: CentWaveParam 
##  MS level(s) 1 
## 
## [[2]]
## Object of class "XProcessHistory"
##  type: Peak refinement 
##  date: Wed May 20 05:42:02 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: MergeNeighboringPeaksParam 
##  MS level(s) 1 
## 
## [[3]]
## Object of class "XProcessHistory"
##  type: Peak grouping 
##  date: Wed May 20 05:42:15 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: PeakDensityParam 
##  MS level(s) 1 
## 
## [[4]]
## Object of class "XProcessHistory"
##  type: Retention time correction 
##  date: Wed May 20 05:42:16 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: PeakGroupsParam 
##  MS level(s) 1 
## 
## [[5]]
## Object of class "XProcessHistory"
##  type: Peak grouping 
##  date: Wed May 20 05:42:33 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: PeakDensityParam 
##  MS level(s) 1 
## 
## [[6]]
## Object of class "XProcessHistory"
##  type: Missing peak filling 
##  date: Wed May 20 05:42:34 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: FillChromPeaksParam 
##  MS level(s) 1</code></pre>
<p>It is also possible to extract specific processing steps by specifying its type. Available <em>types</em> can be listed with the <code>processHistoryTypes</code> function. Below we extract the parameter class for the alignment/retention time adjustment step.</p>
<div class="sourceCode" id="cb97"><html><body><pre class="r"><span class="no">ph</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span>(<span class="no">xdata</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"Retention time correction"</span>)

<span class="no">ph</span></pre></body></html></div>
<pre><code>## [[1]]
## Object of class "XProcessHistory"
##  type: Retention time correction 
##  date: Wed May 20 05:42:16 2020 
##  info:  
##  fileIndex: 1,2,3,4,5,6,7,8,9,10,11,12 
##  Parameter class: PeakGroupsParam 
##  MS level(s) 1</code></pre>
<p>And we can also extract the parameter class used in this preprocessing step.</p>
<div class="sourceCode" id="cb99"><html><body><pre class="r"><span class="co">## Access the parameter</span>
<span class="fu"><a href="../reference/ProcessHistory-class.html">processParam</a></span>(<span class="no">ph</span><span class="kw">[[</span><span class="fl">1</span>]])</pre></body></html></div>
<pre><code>## Object of class:  PeakGroupsParam 
## Parameters:
##  minFraction: 0.85 
##  extraPeaks: 1 
##  smooth: loess 
##  span: 0.2 
##  family: gaussian 
##  subset:  
##  number of peak groups: 55</code></pre>
</div>
<div id="subsetting-and-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-and-filtering" class="anchor"></a>Subsetting and filtering</h2>
<p><code>XCMSnEx</code> objects can be subsetted/filtered using the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> method, or one of the many <code>filter*</code> methods. All these methods aim to ensure that the data in the returned object is consistent. This means for example that if the object is subsetted by selecting specific spectra (by using the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> method) all identified chromatographic peaks are removed. Correspondence results (i.e. identified features) are removed if the object is subsetted to contain only data from selected files (using the <code>filterFile</code> method). This is because the correspondence results depend on the files on which the analysis was performed - running a correspondence on a subset of the files would lead to different results.</p>
<p>As an exception, it is possible to force keeping adjusted retention times in the subsetted object setting the <code>keepAdjustedRtime</code> argument to <code>TRUE</code> in any of the subsetting methods.</p>
<p>Below we subset our results object the data for the files 2 and 4.</p>
<div class="sourceCode" id="cb101"><html><body><pre class="r"><span class="no">subs</span> <span class="kw">&lt;-</span> <span class="fu">filterFile</span>(<span class="no">xdata</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2</span>, <span class="fl">4</span>))

<span class="co">## Do we have identified chromatographic peaks?</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">hasChromPeaks</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>Peak detection is performed separately on each file, thus the subsetted object contains all identified chromatographic peaks from the two files. However, we used a retention time adjustment (alignment) that was based on available features. All features have however been removed and also the adjusted retention times (since the alignment based on features that were identified on chromatographic peaks on all files).</p>
<div class="sourceCode" id="cb103"><html><body><pre class="r"><span class="co">## Do we still have features?</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">hasFeatures</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb105"><html><body><pre class="r"><span class="co">## Do we still have adjusted retention times?</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] FALSE</code></pre>
<p>We can however use the <code>keepAdjustedRtime</code> argument to force keeping the adjusted retention times.</p>
<div class="sourceCode" id="cb107"><html><body><pre class="r"><span class="no">subs</span> <span class="kw">&lt;-</span> <span class="fu">filterFile</span>(<span class="no">xdata</span>, <span class="kw">keepAdjustedRtime</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>The <code>filterRt</code> method can be used to subset the object to spectra within a certain retention time range.</p>
<div class="sourceCode" id="cb109"><html><body><pre class="r"><span class="no">subs</span> <span class="kw">&lt;-</span> <span class="fu">filterRt</span>(<span class="no">xdata</span>, <span class="kw">rt</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3000</span>, <span class="fl">3500</span>))

<span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="fu">rtime</span>(<span class="no">subs</span>))</pre></body></html></div>
<pre><code>## [1] 3000.203 3499.910</code></pre>
<p>Filtering by retention time does not change/affect adjusted retention times (also, if adjusted retention times are present, the filtering is performed <strong>on</strong> the adjusted retention times).</p>
<div class="sourceCode" id="cb111"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>Also, we have all identified chromatographic peaks within the specified retention time range:</p>
<div class="sourceCode" id="cb113"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">hasChromPeaks</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb115"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">subs</span>)[, <span class="st">"rt"</span>])</pre></body></html></div>
<pre><code>## [1] 3000.203 3499.910</code></pre>
<p>The most natural way to subset any object in R is with <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code>. Using <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> on an <code>XCMSnExp</code> object subsets it keeping only the selected spectra. The index <code>i</code> used in <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> has thus to be an integer between 1 and the total number of spectra (across all files). Below we subset <code>xdata</code> using both <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> and <code>filterFile</code> to keep all spectra from one file.</p>
<div class="sourceCode" id="cb117"><html><body><pre class="r"><span class="co">## Extract all data from the 3rd file.</span>
<span class="no">one_file</span> <span class="kw">&lt;-</span> <span class="fu">filterFile</span>(<span class="no">xdata</span>, <span class="kw">file</span> <span class="kw">=</span> <span class="fl">3</span>)

<span class="no">one_file_2</span> <span class="kw">&lt;-</span> <span class="no">xdata</span>[<span class="fu">fromFile</span>(<span class="no">xdata</span>) <span class="kw">==</span> <span class="fl">3</span>]

<span class="co">## Is the content the same?</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.equal.html">all.equal</a></span>(<span class="fu">spectra</span>(<span class="no">one_file</span>), <span class="fu">spectra</span>(<span class="no">one_file_2</span>))</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<p>While the spectra-content is the same in both objects, <code>one_file</code> contains also the identified chromatographic peaks while <code>one_file_2</code> does not. Thus, in most situations subsetting using one of the filter functions is preferred over the use of <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code>.</p>
<div class="sourceCode" id="cb119"><html><body><pre class="r"><span class="co">## Subsetting with filterFile preserves chromatographic peaks</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">one_file</span>))</pre></body></html></div>
<pre><code>##         mz mzmin mzmax       rt    rtmin    rtmax      into      intb   maxo sn
## CP1229 316   316   316 2517.029 2501.379 2535.808  741708.6  727319.6  27816 47
## CP1230 333   333   333 2515.464 2501.379 2540.503  960445.7  942106.5  33992 42
## CP1231 338   338   338 2517.029 2501.379 2540.503  788766.4  758962.1  29288 35
## CP1232 332   332   332 2517.029 2501.379 2545.198 4870388.0 4768520.5 169280 92
## CP1233 315   315   315 2515.464 2501.379 2540.503 3714288.6 3634444.9 131136 95
## CP1234 337   337   337 2517.029 2501.379 2549.893 4356732.5 4216751.1 142720 83
##        sample
## CP1229      1
## CP1230      1
## CP1231      1
## CP1232      1
## CP1233      1
## CP1234      1</code></pre>
<div class="sourceCode" id="cb121"><html><body><pre class="r"><span class="co">## Subsetting with [ not</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="no">one_file_2</span>))</pre></body></html></div>
<pre><code>## Warning in .local(object, ...): No chromatographic peaks available.</code></pre>
<pre><code>## NULL</code></pre>
<p>Note however that also <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> does support the <code>keepAdjustedRtime</code> argument. Below we subset the object to spectra 20:30.</p>
<div class="sourceCode" id="cb124"><html><body><pre class="r"><span class="no">subs</span> <span class="kw">&lt;-</span> <span class="no">xdata</span>[<span class="fl">20</span>:<span class="fl">30</span>, <span class="kw">keepAdjustedRtime</span> <span class="kw">=</span> <span class="fl">TRUE</span>]</pre></body></html></div>
<pre><code>## Warning in xdata[20:30, keepAdjustedRtime = TRUE]: Removed preprocessing results</code></pre>
<div class="sourceCode" id="cb126"><html><body><pre class="r"><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb128"><html><body><pre class="r"><span class="co">## Access adjusted retention times:</span>
<span class="fu">rtime</span>(<span class="no">subs</span>)</pre></body></html></div>
<pre><code>## F01.S0020 F01.S0021 F01.S0022 F01.S0023 F01.S0024 F01.S0025 F01.S0026 F01.S0027 
##  2535.998  2537.563  2539.128  2540.693  2542.258  2543.823  2545.388  2546.953 
## F01.S0028 F01.S0029 F01.S0030 
##  2548.518  2550.083  2551.648</code></pre>
<div class="sourceCode" id="cb130"><html><body><pre class="r"><span class="co">## Access raw retention times:</span>
<span class="fu">rtime</span>(<span class="no">subs</span>, <span class="kw">adjusted</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## F01.S0020 F01.S0021 F01.S0022 F01.S0023 F01.S0024 F01.S0025 F01.S0026 F01.S0027 
##  2531.112  2532.677  2534.242  2535.807  2537.372  2538.937  2540.502  2542.067 
## F01.S0028 F01.S0029 F01.S0030 
##  2543.632  2545.197  2546.762</code></pre>
<p>As with <code>MSnExp</code> and <code>OnDiskMSnExp</code> objects, <code><a href="https://rdrr.io/r/base/Extract.html">[[</a></code> can be used to extract a single spectrum object from an <code>XCMSnExp</code> object. The retention time of the spectrum corresponds to the adjusted retention time if present.</p>
<div class="sourceCode" id="cb132"><html><body><pre class="r"><span class="co">## Extract a single spectrum</span>
<span class="no">xdata</span><span class="kw">[[</span><span class="fl">14</span>]]</pre></body></html></div>
<pre><code>## Object of class "Spectrum1"
##  Retention time: 42:7 
##  MSn level: 1 
##  Total ion count: 445 
##  Polarity: NA</code></pre>
<p>At last we can also use the <code>split</code> method that allows to split an <code>XCMSnExp</code> based on a provided factor <code>f</code>. Below we split <code>xdata</code> per file. Using <code>keepAdjustedRtime = TRUE</code> ensures that adjusted retention times are not removed.</p>
<div class="sourceCode" id="cb134"><html><body><pre class="r"><span class="no">x_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="no">xdata</span>, <span class="kw">f</span> <span class="kw">=</span> <span class="fu">fromFile</span>(<span class="no">xdata</span>), <span class="kw">keepAdjustedRtime</span> <span class="kw">=</span> <span class="fl">TRUE</span>)

<span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span>(<span class="no">x_list</span>)</pre></body></html></div>
<pre><code>##   1   2   3   4   5   6   7   8   9  10  11  12 
## 958 958 958 958 958 958 958 958 958 958 958 958</code></pre>
<div class="sourceCode" id="cb136"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">x_list</span>, <span class="no">hasAdjustedRtime</span>)</pre></body></html></div>
<pre><code>## $`1`
## [1] TRUE
## 
## $`2`
## [1] TRUE
## 
## $`3`
## [1] TRUE
## 
## $`4`
## [1] TRUE
## 
## $`5`
## [1] TRUE
## 
## $`6`
## [1] TRUE
## 
## $`7`
## [1] TRUE
## 
## $`8`
## [1] TRUE
## 
## $`9`
## [1] TRUE
## 
## $`10`
## [1] TRUE
## 
## $`11`
## [1] TRUE
## 
## $`12`
## [1] TRUE</code></pre>
<p>Note however that there is also a dedicated <code>splitByFile</code> method instead for that operation, that internally uses <code>filterFile</code> and hence does e.g. not remove identified chromatographic peaks. The method does not yet support the <code>keepAdjustedRtime</code> parameter and thus removes by default adjusted retention times.</p>
<div class="sourceCode" id="cb138"><html><body><pre class="r"><span class="no">xdata_by_file</span> <span class="kw">&lt;-</span> <span class="fu">splitByFile</span>(<span class="no">xdata</span>, <span class="kw">f</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu">fileNames</span>(<span class="no">xdata</span>))))

<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">xdata_by_file</span>, <span class="no">hasChromPeaks</span>)</pre></body></html></div>
<pre><code>## $`1`
## [1] TRUE
## 
## $`2`
## [1] TRUE
## 
## $`3`
## [1] TRUE
## 
## $`4`
## [1] TRUE
## 
## $`5`
## [1] TRUE
## 
## $`6`
## [1] TRUE
## 
## $`7`
## [1] TRUE
## 
## $`8`
## [1] TRUE
## 
## $`9`
## [1] TRUE
## 
## $`10`
## [1] TRUE
## 
## $`11`
## [1] TRUE
## 
## $`12`
## [1] TRUE</code></pre>
</div>
<div id="parallel-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-processing" class="anchor"></a>Parallel processing</h2>
<p>Most methods in <code>xcms</code> support parallel processing. Parallel processing is handled and configured by the <code>BiocParallel</code> Bioconductor package and can be globally defined for an R session.</p>
<p>Unix-based systems (Linux, macOS) support <code>multicore</code>-based parallel processing. To configure it globally we <code>register</code> the parameter class. Note also that <code>bpstart</code> is used below to initialize the parallel processes.</p>
<div class="sourceCode" id="cb140"><html><body><pre class="r"><span class="fu">register</span>(<span class="fu">bpstart</span>(<span class="fu">MulticoreParam</span>(<span class="fl">2</span>)))</pre></body></html></div>
<p>Windows supports only socket-based parallel processing:</p>
<div class="sourceCode" id="cb141"><html><body><pre class="r"><span class="fu">register</span>(<span class="fu">bpstart</span>(<span class="fu">SnowParam</span>(<span class="fl">2</span>)))</pre></body></html></div>
<p>Note that <code>multicore</code>-based parallel processing might be buggy or failing on macOS. If so, the <code>DoparParam</code> could be used instead (requiring the <code>foreach</code> package).</p>
<p>For other options and details see the vignettes from the <code>BiocParallel</code> package.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Saghatelian04">
<p>1. Saghatelian A, Trauger SA, Want EJ, Hawkins EG, Siuzdak G, Cravatt BF: <strong>Assignment of endogenous substrates to enzymes by global metabolite profiling</strong>. <em>Biochemistry</em> 2004, <strong>43</strong>:14332–9.</p>
</div>
<div id="ref-Tautenhahn:2008fx">
<p>2. Tautenhahn R, Böttcher C, Neumann S: <strong>Highly sensitive feature detection for high resolution LC/MS.</strong> <em>BMC Bioinformatics</em> 2008, <strong>9</strong>:504.</p>
</div>
<div id="ref-Smith:2013gr">
<p>3. Smith R, Ventura D, Prince JT: <strong>LC-MS alignment in theory and practice: a comprehensive algorithmic review.</strong> <em>Briefings in bioinformatics</em> 2013, <strong>16</strong>:bbt080–117.</p>
</div>
<div id="ref-Prince:2006jj">
<p>4. Prince JT, Marcotte EM: <strong>Chromatographic alignment of ESI-LC-MS proteomics data sets by ordered bijective interpolated warping.</strong> <em>Analytical chemistry</em> 2006, <strong>78</strong>:6140–6152.</p>
</div>
<div id="ref-Smith:2006ic">
<p>5. Smith CA, Want EJ, O’Maille G, Abagyan R, Siuzdak G: <strong>XCMS: processing mass spectrometry data for metabolite profiling using nonlinear peak alignment, matching, and identification.</strong> <em>Analytical chemistry</em> 2006, <strong>78</strong>:779–787.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Steffen Neumann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
