<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LC-MS/MS data analysis with xcms • xcms</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LC-MS/MS data analysis with xcms">
<meta property="og:description" content="xcms">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xcms</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.11.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/xcms.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sneumann/xcms/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="xcms-lcms-ms_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LC-MS/MS data analysis with xcms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/master/vignettes/xcms-lcms-ms.Rmd"><code>vignettes/xcms-lcms-ms.Rmd</code></a></small>
      <div class="hidden name"><code>xcms-lcms-ms.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.12/xcms">xcms</a></em><br><strong>Authors</strong>: Johannes Rainer, Michael Witting<br><strong>Modified</strong>: 2020-06-25 14:40:01<br><strong>Compiled</strong>: Thu Jun 25 15:13:50 2020</p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Metabolite identification is an important step in non-targeted metabolomics and requires different steps. One involves the use of tandem mass spectrometry to generate fragmentation spectra of detected metabolites (LC-MS/MS), which are then compared to fragmentation spectra of known metabolites. Different approaches exist for the generation of these fragmentation spectra, whereas the most used is data dependent acquisition (DDA) also known as the top-n method. In this method the top N most intense m/z values from a MS1 scan are selected for fragmentation in the next N scans before the cycle starts again. This method allows to generate clean MS2 fragmentation spectra on the fly during acquisition without the need for further experiments, but suffers from poor coverage of the detected metabolites (since only a limited number of ions are fragmented).</p>
<p>Data independent approaches (DIA) like Bruker bbCID, Agilent AllIons or Waters MSe don’t use such a preselection, but rather fragment all detected molecules at once. They are using alternating schemes with scan of low and high collision energy to collect MS1 and MS2 data. Using this approach, there is no problem in coverage, but the relation between the precursor and fragment masses is lost leading to chimeric spectra. Sequential Window Acquisition of all Theoretical Mass Spectra (or SWATH <span class="citation">[1]</span>) combines both approaches through a middle-way approach. There is no precursor selection and acquisition is independent of acquired data, but rather than isolating all precusors at once, defined windows (i.e. ranges of m/z values) are used and scanned. This reduces the overlap of fragment spectra while still keeping a high coverage.</p>
<p>This document showcases the analysis of two small LC-MS/MS data sets using <code>r Biocpkg("xcms")</code>. The data files used are reversed-phase LC-MS/MS runs from the Agilent Pesticide mix obtained from a Sciex 6600 Triple ToF operated in SWATH acquisition mode. For comparison a DDA file from the same sample is included.</p>
</div>
<div id="analysis-of-dda-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-of-dda-data" class="anchor"></a>Analysis of DDA data</h1>
<p>Below we load the example DDA data set using the <code>readMSData</code> function from the <em><a href="https://bioconductor.org/packages/3.12/MSnbase">MSnbase</a></em> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/sneumann/xcms">xcms</a></span>)

<span class="kw">dda_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"TripleTOF-SWATH/PestMix1_DDA.mzML"</span>,
                        package = <span class="st">"msdata"</span>)
<span class="kw">dda_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/readMSData.html">readMSData</a></span>(<span class="kw">dda_file</span>, mode = <span class="st">"onDisk"</span>)
</pre></div>
<p>The variable <code>dda_data</code> contains now all MS1 and MS2 spectra from the specified mzML file. The number of spectra for each MS level is listed below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">msLevel</a></span>(<span class="kw">dda_data</span>))
</pre></div>
<pre><code>## 
##    1    2 
## 1504 2238</code></pre>
<p>Next we perform the chromatographic peak detection on the MS level 1 data with the <code>findChromPeaks</code> method. Below we define the settings for a <em>centWave</em>-based peak detection and perform the analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span>(snthresh = <span class="fl">5</span>, noise = <span class="fl">100</span>, ppm = <span class="fl">10</span>,
                     peakwidth = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">30</span>))
<span class="kw">dda_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span>(<span class="kw">dda_data</span>, param = <span class="kw">cwp</span>)
</pre></div>
<p>In total 112 peaks were identified in the present data set.</p>
<p>The advantage of LC-MS/MS data is that (MS1) ions are fragmented and the corresponding MS2 spectra measured. Thus, for some of the ions (identified as MS1 chromatographic peaks) MS2 spectra are available. These can facilitate the annotation of the respective MS1 chromatographic peaks (or MS1 features after a correspondence analysis). MS2 spectra for identified chromatographic peaks can be extracted with the <code>chromPeakSpectra</code> method which returns a <code>MSpectra</code> object of MS2 spectra associated with a chromatographic peak (i.e. with their retention time within the retention time window of a chromatographic peak and their precursor m/z within the m/z range of the same chromatographic peak.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="kw">dda_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromPeakSpectra.html">chromPeakSpectra</a></span>(<span class="kw">dda_data</span>)
<span class="kw">dda_spectra</span>
</pre></div>
<pre><code>## MSpectra with 150 spectra and 1 metadata column(s):
##                    msLevel     rtime peaksCount |     peak_id
##                  &lt;integer&gt; &lt;numeric&gt;  &lt;integer&gt; | &lt;character&gt;
##   CP004.F1.S1812         2   237.869       1489 |       CP004
##   CP004.F1.S1846         2   241.299       1807 |       CP004
##   CP005.F1.S2446         2   326.583         21 |       CP005
##   CP006.F1.S2450         2   327.113         85 |       CP006
##   CP006.F1.S2502         2   330.273        168 |       CP006
##              ...       ...       ...        ... .         ...
##   CP100.F1.S5110         2   574.725        890 |       CP100
##   CP100.F1.S5115         2   575.255       1042 |       CP100
##   CP101.F1.S5272         2   596.584        196 |       CP101
##   CP102.F1.S5236         2   592.424         12 |       CP102
##   CP102.F1.S5266         2   596.054         88 |       CP102</code></pre>
<p>The metadata column <code>"peak_id"</code> contains the identifiers of the chromatographic peaks the MS2 spectrum is associated with.</p>
<p>We next use the MS2 information to aid in the annotation of a chromatographic peak. As an example we use a chromatographic peak of an ion with an m/z of 304.1131 which we extract in the code block below.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="kw">ex_mz</span> <span class="op">&lt;-</span> <span class="fl">304.1131</span>
<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">dda_data</span>, mz = <span class="kw">ex_mz</span>, ppm = <span class="fl">20</span>)
</pre></div>
<pre><code>##             mz    mzmin    mzmax      rt   rtmin   rtmax    into     intb
## CP057 304.1133 304.1126 304.1143 425.024 417.985 441.773 13040.8 12884.14
##           maxo sn sample
## CP057 3978.987 79      1</code></pre>
<p>A search of potential ions with a similar m/z in a reference database (e.g. <a href="https://metlin.scripps.edu">Metlin</a>) returned a large list of potential hits, most with a very small ppm. For two of the hits, <a href="https://en.wikipedia.org/wiki/Flumazenil">Flumazenil</a> (Metlin ID 2724) and <a href="https://en.wikipedia.org/wiki/Fenamiphos">Fenamiphos</a> (Metlin ID 72445) experimental MS2 spectra are available. Thus, we could match the MS2 spectrum for the identified chromatographic peak against these to annotate our ion. Below we extract all MS2 spectra that were associated with the candidate chromatographic peak using the ID of the peak in the present data set.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="kw">ex_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">dda_data</span>, mz = <span class="kw">ex_mz</span>, ppm = <span class="fl">20</span>))
<span class="kw">ex_spectra</span> <span class="op">&lt;-</span> <span class="kw">dda_spectra</span>[<span class="fu">mcols</span>(<span class="kw">dda_spectra</span>)<span class="op">$</span><span class="kw">peak_id</span> <span class="op">==</span> <span class="kw">ex_id</span>]
<span class="kw">ex_spectra</span>
</pre></div>
<pre><code>## MSpectra with 5 spectra and 1 metadata column(s):
##                    msLevel     rtime peaksCount |     peak_id
##                  &lt;integer&gt; &lt;numeric&gt;  &lt;integer&gt; | &lt;character&gt;
##   CP057.F1.S3505         2   418.926         10 |       CP057
##   CP057.F1.S3510         2   419.306         30 |       CP057
##   CP057.F1.S3582         2   423.036        694 |       CP057
##   CP057.F1.S3603         2   423.966        783 |       CP057
##   CP057.F1.S3609         2   424.296        753 |       CP057</code></pre>
<p>There are 5 MS2 spectra representing fragmentation of the ion(s) measured in our candidate chromatographic peak. We next reduce this to a single MS2 spectrum using the <code>combineSpectra</code> method employing the <code>consensusSpectrum</code> function to determine which peaks to keep in the resulting spectrum.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="kw">ex_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/combineSpectra.html">combineSpectra</a></span>(<span class="kw">ex_spectra</span>, method = <span class="kw">consensusSpectrum</span>, mzd = <span class="fl">0</span>,
                              ppm = <span class="fl">20</span>, minProp = <span class="fl">0.8</span>, weighted = <span class="fl">FALSE</span>,
                              intensityFun = <span class="kw">median</span>, mzFun = <span class="kw">median</span>)
<span class="kw">ex_spectrum</span>
</pre></div>
<pre><code>## MSpectra with 1 spectra and 1 metadata column(s):
##                    msLevel     rtime peaksCount |     peak_id
##                  &lt;integer&gt; &lt;numeric&gt;  &lt;integer&gt; | &lt;character&gt;
##   CP057.F1.S3505         2   418.926         24 |       CP057</code></pre>
<p>Mass peaks from all input spectra with a difference in m/z smaller 20 ppm (parameter <code>ppm</code>) were combined into one peak and the median m/z and intensity is reported for these. Due to parameter <code>minProp = 0.8</code>, the resulting MS2 spectrum contains only peaks that were present in 80% of the input spectra.</p>
<p>A plot of this <em>consensus</em> spectrum is shown below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]])
</pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/dda-ms2-consensus-plot-1.png" alt="Consensus MS2 spectrum created from all measured MS2 spectra for ions of chromatographic peak CP53." width="768"><p class="caption">
Consensus MS2 spectrum created from all measured MS2 spectra for ions of chromatographic peak CP53.
</p>
</div>
<p>We could now match the consensus spectrum against a database of MS2 spectra. In our example we simply load MS2 spectra for the two compounds with matching m/z exported from Metlin. For each of the compounds MS2 spectra created with collision energies of 0V, 10V, 20V and 40V are available. Below we import the respective data and plot our candidate spectrum against the MS2 spectra of Flumanezil and Fenamiphos (from a collision energy of 20V).</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="kw">flumanezil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/pSet-class.html">spectra</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/readMgfData.html">readMgfData</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"mgf/metlin-2724.mgf"</span>, package = <span class="st">"xcms"</span>)))
<span class="kw">fenamiphos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/pSet-class.html">spectra</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/readMgfData.html">readMgfData</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"mgf/metlin-72445.mgf"</span>, package = <span class="st">"xcms"</span>)))

<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(mfrow = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">flumanezil</span>[[<span class="fl">3</span>]], main = <span class="st">"against Flumanezil"</span>,
     tolerance = <span class="fl">40e-6</span>)
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">fenamiphos</span>[[<span class="fl">3</span>]], main = <span class="st">"against Fenamiphos"</span>,
     tolerance = <span class="fl">40e-6</span>)
</pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/dda-ms2-metlin-match-1.png" alt="Mirror plots for the candidate MS2 spectrum against Flumanezil (left) and Fenamiphos (right). The upper panel represents the candidate MS2 spectrum, the lower the target MS2 spectrum. Matching peaks are indicated with a dot." width="1152"><p class="caption">
Mirror plots for the candidate MS2 spectrum against Flumanezil (left) and Fenamiphos (right). The upper panel represents the candidate MS2 spectrum, the lower the target MS2 spectrum. Matching peaks are indicated with a dot.
</p>
</div>
<p>Our candidate spectrum matches Fenamiphos, thus, our example chromatographic peak represents signal measured for this compound. In addition to plotting the spectra, we can also calculate similarities between them with the <code>compareSpectra</code> method (e.g. using the <em>dotproduct</em> method as shown below).</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/compareSpectra-methods.html">compareSpectra</a></span>(<span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">flumanezil</span>[[<span class="fl">3</span>]], binSize = <span class="fl">0.02</span>,
               fun = <span class="st">"dotproduct"</span>)
</pre></div>
<pre><code>## [1] 0.005350294</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/compareSpectra-methods.html">compareSpectra</a></span>(<span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">fenamiphos</span>[[<span class="fl">3</span>]], binSize = <span class="fl">0.02</span>,
               fun = <span class="st">"dotproduct"</span>)
</pre></div>
<pre><code>## [1] 0.9092189</code></pre>
<p>Clearly, the candidate spectrum does not match Flumanezil, while it has a high similarity to Fenamiphos. While we performed here the MS2-based annotation on a single chromatographic peak, this could be easily extended to the full list of MS2 spectra (returned by <code>chromPeakSpectra</code>) for all chromatographic peaks in an experiment. Respective code facilitating this will be implemented in future.</p>
<p>In the present example we used only a single data file and we did thus not need to perform a sample alignment and correspondence analysis. These tasks could however be performed similarly to <em>plain</em> LC-MS data, retention times of recorded MS2 spectra would however also be adjusted during alignment based on the MS1 data. After correspondence analysis (peak grouping) MS2 spectra for <em>features</em> can be extracted with the <code>featureSpectra</code> function which returns all MS2 spectra associated with any chromatographic peak of a feature.</p>
<p>Note also that this workflow can be included into the <em>Feature-Based Molecular Networking</em> <a href="https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking/">FBMN</a> to match MS2 spectra against <a href="https://gnps.ucsd.edu/">GNPS</a>. See <a href="https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking-with-xcms3/">here</a> for more details and examples.</p>
</div>
<div id="swath-data-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#swath-data-analysis" class="anchor"></a>SWATH data analysis</h1>
<p>In this section we analyze a small SWATH data set consisting of a single mzML file with data from the same sample analyzed in the previous section but recorded in SWATH mode. We again read the data with the <code>readMSData</code> function. The resulting object will contain all recorded MS1 and MS2 spectra in the specified file.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="kw">swath_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"TripleTOF-SWATH"</span>,
                          <span class="st">"PestMix1_SWATH.mzML"</span>,
                          package = <span class="st">"msdata"</span>)

<span class="kw">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/readMSData.html">readMSData</a></span>(<span class="kw">swath_file</span>, mode = <span class="st">"onDisk"</span>)
</pre></div>
<p>Below we determine the number of MS level 1 and 2 spectra in the present data set.</p>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">msLevel</a></span>(<span class="kw">swath_data</span>))
</pre></div>
<pre><code>## 
##    1    2 
##  444 3556</code></pre>
<p>As described in the introduction, in SWATH mode all ions within pre-defined isolation windows are fragmented and MS2 spectra measured. The definition of these isolation windows (SWATH pockets) is imported from the mzML files and stored in the object’s <code>fData</code> (which provides additional annotations for each individual spectrum). Below we inspect the respective information for the first few spectra. The upper and lower isolation window m/z can be extracted with the <code>isolationWindowLowerMz</code> and <code>isolationWindowUpperMz</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu">fData</span>(<span class="kw">swath_data</span>)[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"isolationWindowTargetMZ"</span>,
                           <span class="st">"isolationWindowLowerOffset"</span>,
                           <span class="st">"isolationWindowUpperOffset"</span>,
                           <span class="st">"msLevel"</span>, <span class="st">"retentionTime"</span>)])
</pre></div>
<pre><code>##          isolationWindowTargetMZ isolationWindowLowerOffset
## F1.S2000                  208.95                      21.95
## F1.S2001                  244.05                      14.15
## F1.S2002                  270.85                      13.65
## F1.S2003                  299.10                      15.60
## F1.S2004                  329.80                      16.10
## F1.S2005                  367.35                      22.45
##          isolationWindowUpperOffset msLevel retentionTime
## F1.S2000                      21.95       2       200.084
## F1.S2001                      14.15       2       200.181
## F1.S2002                      13.65       2       200.278
## F1.S2003                      15.60       2       200.375
## F1.S2004                      16.10       2       200.472
## F1.S2005                      22.45       2       200.569</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/pSet-class.html">isolationWindowLowerMz</a></span>(<span class="kw">swath_data</span>))
</pre></div>
<pre><code>## [1] 187.0 229.9 257.2 283.5 313.7 344.9</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/pSet-class.html">isolationWindowUpperMz</a></span>(<span class="kw">swath_data</span>))
</pre></div>
<pre><code>## [1] 230.9 258.2 284.5 314.7 345.9 389.8</code></pre>
<p>In the present data set we use the value of the <em>isolation window target m/z</em> to define the individual SWATH pockets. Below we list the number of spectra that are recorded in each pocket/isolation window.</p>
<div class="sourceCode" id="cb28"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="../reference/isolationWindowTargetMz,OnDiskMSnExp-method.html">isolationWindowTargetMz</a></span>(<span class="kw">swath_data</span>))
</pre></div>
<pre><code>## 
## 163.75 208.95 244.05 270.85  299.1  329.8 367.35 601.85 
##    444    445    445    445    445    444    444    444</code></pre>
<p>We have thus 1,000 MS2 spectra measured in each isolation window.</p>
<div id="chromatographic-peak-detection-in-ms1-and-ms2-data" class="section level2">
<h2 class="hasAnchor">
<a href="#chromatographic-peak-detection-in-ms1-and-ms2-data" class="anchor"></a>Chromatographic peak detection in MS1 and MS2 data</h2>
<p>Similar to a <em>conventional</em> LC-MS analysis, we perform first a chromatographic peak detection (on the MS level 1 data) with the <code>findChromPeaks</code> method. Below we define the settings for a <em>centWave</em>-based peak detection and perform the analysis.</p>
<div class="sourceCode" id="cb30"><pre class="downlit">
<span class="kw">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span>(snthresh = <span class="fl">5</span>, noise = <span class="fl">100</span>, ppm = <span class="fl">10</span>,
                     peakwidth = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">30</span>))
<span class="kw">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span>(<span class="kw">swath_data</span>, param = <span class="kw">cwp</span>)
</pre></div>
<p>Next we perform a chromatographic peak detection in the MS level 2 data of each isolation window. We use the <code>findChromPeaksIsolationWindow</code> function employing the same peak detection algorithm reducing however the required signal-to-noise ratio. The <code>isolationWindow</code> parameter allows to specify which MS2 spectra belong to which isolation window and hence defines in which set of MS2 spectra chromatographic peak detection should be performed. While the default value for this parameter uses isolation windows provided by calling <code>isolationWindowTargetMz</code> on the object, it would also be possible to manually define the isolation windows, e.g. if the corresponding information is not available in the input mzML files.</p>
<div class="sourceCode" id="cb31"><pre class="downlit">
<span class="kw">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span>(snthresh = <span class="fl">3</span>, noise = <span class="fl">10</span>, ppm = <span class="fl">10</span>,
                     peakwidth = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">30</span>))
<span class="kw">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaksIsolationWindow.html">findChromPeaksIsolationWindow</a></span>(<span class="kw">swath_data</span>, param = <span class="kw">cwp</span>)
</pre></div>
<p>The <code>findChromPeaksIsolationWindow</code> function added all peaks identified in the individual isolation windows to the <code>chromPeaks</code> matrix containing already the MS1 chromatographic peaks. These newly added peaks can be identified by the value of the <code>"isolationWindow"</code> column in the corresponding row in <code>chromPeakData</code>, which lists also the MS level in which the peak was identified.</p>
<div class="sourceCode" id="cb32"><pre class="downlit">
<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="kw">swath_data</span>)
</pre></div>
<pre><code>## DataFrame with 368 rows and 6 columns
##        ms_level is_filled isolationWindow isolationWindowTargetMZ
##       &lt;integer&gt; &lt;logical&gt;        &lt;factor&gt;               &lt;numeric&gt;
## CP01          1     FALSE              NA                      NA
## CP02          1     FALSE              NA                      NA
## CP03          1     FALSE              NA                      NA
## CP04          1     FALSE              NA                      NA
## CP05          1     FALSE              NA                      NA
## ...         ...       ...             ...                     ...
## CP364         2     FALSE          601.85                  601.85
## CP365         2     FALSE          601.85                  601.85
## CP366         2     FALSE          601.85                  601.85
## CP367         2     FALSE          601.85                  601.85
## CP368         2     FALSE          601.85                  601.85
##       isolationWindowLowerMz isolationWindowUpperMz
##                    &lt;numeric&gt;              &lt;numeric&gt;
## CP01                      NA                     NA
## CP02                      NA                     NA
## CP03                      NA                     NA
## CP04                      NA                     NA
## CP05                      NA                     NA
## ...                      ...                    ...
## CP364                  388.8                  814.9
## CP365                  388.8                  814.9
## CP366                  388.8                  814.9
## CP367                  388.8                  814.9
## CP368                  388.8                  814.9</code></pre>
<p>Below we count the number of chromatographic peaks identified within each isolation window (the number of chromatographic peaks identified in MS1 is 62).</p>
<div class="sourceCode" id="cb34"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="kw">swath_data</span>)<span class="op">$</span><span class="kw">isolationWindow</span>)
</pre></div>
<pre><code>## 
## 163.75 208.95 244.05 270.85  299.1  329.8 367.35 601.85 
##      2     38     32     14    105     23     61     31</code></pre>
<p>We thus successfully identified chromatographic peaks in the different MS levels and isolation windows, but don’t have any actual MS2 <em>spectra</em> yet. These have to be reconstructed from the available chromatographic peak data which we will done in the next section.</p>
</div>
<div id="reconstruction-of-ms2-spectra" class="section level2">
<h2 class="hasAnchor">
<a href="#reconstruction-of-ms2-spectra" class="anchor"></a>Reconstruction of MS2 spectra</h2>
<p>Identifying the signal of the fragment ions for the precursor measured by each MS1 chromatographic peak is a non-trivial task. The MS2 spectrum of the fragment ion for each MS1 chromatographic peak has to be reconstructed from the available MS2 signal (i.e. the chromatographic peaks identified in MS level 2). For SWATH data, fragment ion signal should be present in the isolation window that contains the m/z of the precursor ion and the chromatographic peak shape of the MS2 chromatographic peaks of fragment ions of a specific precursor should have a similar retention time and peak shape than the precursor’s MS1 chromatographic peak.</p>
<p>After detection of MS1 and MS2 chromatographic peaks has been performed, we can reconstruct the MS2 spectra using the <code>reconstructChromPeakSpectra</code> function. This function defines an MS2 spectrum for each MS1 chromatographic peak based on the following approach:</p>
<ul>
<li>Identify MS2 chromatographic peaks in the isolation window containing the m/z of the ion (the MS1 chromatographic peak) that have approximately the same retention time than the MS1 chromatographic peak (the accepted difference in retention time can be defined with the <code>diffRt</code> parameter).</li>
<li>Extract the MS1 chromatographic peak and all MS2 chromatographic peaks identified by the previous step and correlate the peak shapes of the candidate MS2 chromatographic peaks with the shape of the MS1 peak. MS2 chromatographic peaks with a correlation coefficient larger than <code>minCor</code> are retained.</li>
<li>Reconstruct the MS2 spectrum using the m/z of all above selected MS2 chromatographic peaks and their intensity; each MS2 chromatographic peak selected for an MS1 peak will thus represent one <strong>mass peak</strong> in the reconstructed spectrum.</li>
</ul>
<p>To illustrate this process we perform the individual steps on the example of Fenamiphos (exact mass 303.105800777 and m/z of [M+H]+ adduct 304.113077). As a first step we extract the chromatographic peak for this ion.</p>
<div class="sourceCode" id="cb36"><pre class="downlit">
<span class="kw">fenamiphos_mz</span> <span class="op">&lt;-</span> <span class="fl">304.113077</span>
<span class="kw">fenamiphos_ms1_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>, mz = <span class="kw">fenamiphos_mz</span>, ppm = <span class="fl">2</span>)
<span class="kw">fenamiphos_ms1_peak</span>
</pre></div>
<pre><code>##            mz    mzmin    mzmax      rt   rtmin   rtmax     into     intb
## CP34 304.1124 304.1121 304.1126 423.945 419.445 428.444 10697.34 10688.34
##          maxo  sn sample
## CP34 2401.849 618      1</code></pre>
<p>Next we identify all MS2 chromatographic peaks that were identified in the isolation window containing the m/z of Fenamiphos. The information on the isolation window in which a chromatographic peak was identified is available in the <code>chromPeakData</code> (which contains arbitrary additional annotations to each individual chromatographic peak).</p>
<div class="sourceCode" id="cb38"><pre class="downlit">
<span class="kw">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="kw">swath_data</span>)<span class="op">$</span><span class="kw">isolationWindowLowerMz</span> <span class="op">&lt;</span> <span class="kw">fenamiphos_mz</span> <span class="op">&amp;</span>
        <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span>(<span class="kw">swath_data</span>)<span class="op">$</span><span class="kw">isolationWindowUpperMz</span> <span class="op">&gt;</span> <span class="kw">fenamiphos_mz</span>
</pre></div>
<p>We also require the retention time of the MS2 chromatographic peaks to be similar to the retention time of the MS1 peak and extract the corresponding peak information.</p>
<div class="sourceCode" id="cb39"><pre class="downlit">
<span class="kw">keep</span> <span class="op">&lt;-</span> <span class="kw">keep</span> <span class="op">&amp;</span>
    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>)[, <span class="st">"rtmin"</span>] <span class="op">&lt;</span> <span class="kw">fenamiphos_ms1_peak</span>[, <span class="st">"rt"</span>] <span class="op">&amp;</span>
    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>)[, <span class="st">"rtmax"</span>] <span class="op">&gt;</span> <span class="kw">fenamiphos_ms1_peak</span>[, <span class="st">"rt"</span>]

<span class="kw">fenamiphos_ms2_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>)[<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw">keep</span>), ]
</pre></div>
<p>In total 24 MS2 chromatographic peaks match all the above condition. Next we extract their corresponding ion chromatograms, as well as the ion chromatogram of the MS1 peak. In addition we have to filter the object first by isolation window, keeping only spectra that were measured in that specific window and to specify to extract the chromatographic data from MS2 spectra (with <code>msLevel = 2L</code>).</p>
<div class="sourceCode" id="cb40"><pre class="downlit">
<span class="kw">rtr</span> <span class="op">&lt;-</span> <span class="kw">fenamiphos_ms1_peak</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rtmin"</span>, <span class="st">"rtmax"</span>)]
<span class="kw">mzr</span> <span class="op">&lt;-</span> <span class="kw">fenamiphos_ms1_peak</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"mzmin"</span>, <span class="st">"mzmax"</span>)]
<span class="kw">fenamiphos_ms1_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram,XCMSnExp-method.html">chromatogram</a></span>(<span class="kw">swath_data</span>, rt = <span class="kw">rtr</span>, mz = <span class="kw">mzr</span>)

<span class="kw">rtr</span> <span class="op">&lt;-</span> <span class="kw">fenamiphos_ms2_peak</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rtmin"</span>, <span class="st">"rtmax"</span>)]
<span class="kw">mzr</span> <span class="op">&lt;-</span> <span class="kw">fenamiphos_ms2_peak</span>[, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"mzmin"</span>, <span class="st">"mzmax"</span>)]
<span class="kw">fenamiphos_ms2_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram,XCMSnExp-method.html">chromatogram</a></span>(
    <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/MSnExp-class.html">filterIsolationWindow</a></span>(<span class="kw">swath_data</span>, mz = <span class="kw">fenamiphos_mz</span>),
    rt = <span class="kw">rtr</span>, mz = <span class="kw">mzr</span>, msLevel = <span class="fl">2L</span>)
</pre></div>
<p>We can now plot the extracted ion chromatogram of the MS1 and the extracted MS2 data.</p>
<div class="sourceCode" id="cb41"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">rtime</a></span>(<span class="kw">fenamiphos_ms1_chr</span>[<span class="fl">1</span>, <span class="fl">1</span>]),
     <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">intensity</a></span>(<span class="kw">fenamiphos_ms1_chr</span>[<span class="fl">1</span>, <span class="fl">1</span>]),
     xlab = <span class="st">"retention time [s]"</span>, ylab = <span class="st">"intensity"</span>, pch = <span class="fl">16</span>,
     ylim = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">5000</span>), col = <span class="st">"blue"</span>, type = <span class="st">"b"</span>, lwd = <span class="fl">2</span>)
<span class="co">#' Add data from all MS2 peaks</span>
<span class="kw">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw">fenamiphos_ms2_chr</span><span class="op">@</span>.Data,
              <span class="fu">function</span>(<span class="kw">z</span>) <span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">rtime</a></span>(<span class="kw">z</span>), <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum-class.html">intensity</a></span>(<span class="kw">z</span>),
                                 col = <span class="st">"#00000080"</span>,
                                 type = <span class="st">"b"</span>, pch = <span class="fl">16</span>))
</pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/fena-eic-plot-1.png" alt="Extracted ion chromatograms for Fenamiphos from MS1 (blue) and potentially related signal in MS2 (grey)." width="960"><p class="caption">
Extracted ion chromatograms for Fenamiphos from MS1 (blue) and potentially related signal in MS2 (grey).
</p>
</div>
<p>Next we can calculate correlations between the peak shapes of each MS2 chromatogram with the MS1 peak. We perform the correlation below for one of the MS2 chromatographic peaks. Note that, because spectra are recorded consecutively, the retention times of the individual data points will differ for the MS2 and MS1 chromatographic data and data points have thus to be matched (aligned) before performing the correlation analysis. This is done automatically by the <code>correlate</code> function. See the help for the <code>align</code> method for more information on alignment options.</p>
<div class="sourceCode" id="cb42"><pre class="downlit">
<span class="fu"><a href="../reference/correlate,Chromatogram,Chromatogram-method.html">correlate</a></span>(<span class="kw">fenamiphos_ms2_chr</span>[<span class="fl">1</span>, <span class="fl">1</span>],
          <span class="kw">fenamiphos_ms1_chr</span>[<span class="fl">1</span>, <span class="fl">1</span>], align = <span class="st">"approx"</span>)
</pre></div>
<pre><code>## [1] 0.9997871</code></pre>
<p>After identifying the MS2 chromatographic peaks with shapes of enough high similarity to the MS1 chromatographic peaks, an MS2 spectrum could be <em>reconstructed</em> based on the m/z and intensities of the MS2 chromatographic peaks.</p>
<p>The <code>reconstructChromPeakSpectra</code> function performs the above analysis for each individual MS1 chromatographic peak in a SWATH data set. Below we reconstruct MS2 spectra for our example data requiring a peak shape correlation higher than <code>0.9</code> between the candidate MS2 chromatographic peak and the target MS1 chromatographic peak.</p>
<div class="sourceCode" id="cb44"><pre class="downlit">
<span class="kw">swath_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reconstructChromPeakSpectra.html">reconstructChromPeakSpectra</a></span>(<span class="kw">swath_data</span>, minCor = <span class="fl">0.9</span>)
<span class="kw">swath_spectra</span>
</pre></div>
<pre><code>## MSpectra with 62 spectra and 3 metadata column(s):
##       msLevel     rtime peaksCount |     ms2_peak_id      ms2_peak_cor
##     &lt;integer&gt; &lt;numeric&gt;  &lt;integer&gt; | &lt;CharacterList&gt;     &lt;NumericList&gt;
##   1         2   239.458          0 |                                  
##   1         2   240.358          0 |                                  
##   1         2   329.577          1 |           CP063          0.950582
##   1         2   329.771          1 |           CP105           0.95157
##   1         2   346.164          1 |           CP153          0.924545
##   .       ...       ...        ... .             ...               ...
##   1         2   551.735          0 |                                  
##   1         2   551.735          0 |                                  
##   1         2   575.134          0 |                                  
##   1         2   575.134          0 |                                  
##   1         2   574.942          2 |     CP334,CP329 0.918915,0.911944
##         peak_id
##     &lt;character&gt;
##   1        CP01
##   1        CP02
##   1        CP03
##   1        CP04
##   1        CP05
##   .         ...
##   1        CP58
##   1        CP59
##   1        CP60
##   1        CP61
##   1        CP62</code></pre>
<p>As a result we got a <code>Spectra</code> object of length equal to the number of MS1 peaks in our data. A <code>peaksCount</code> of <code>0</code> indicates that no MS2 spectrum could be defined based on the used settings. For reconstructed spectra additional annotations are available such as the IDs of the MS2 chromatographic peaks from which the spectrum was reconstructed (<code>"ms2_peak_id"</code>) as well as the correlation coefficient of their chromatographic peak shape with the precursor’s shape (<code>"ms2_peak_cor"</code>). Metadata column <code>"peak_id"</code> contains the ID of the MS1 chromatographic peak.</p>
<p>We next extract the MS2 spectrum for our example peak most likely representing [M+H]+ ions of Fenamiphos using its chromatographic peak ID:</p>
<div class="sourceCode" id="cb46"><pre class="downlit">
<span class="kw">fenamiphos_swath_spectrum</span> <span class="op">&lt;-</span> <span class="kw">swath_spectra</span>[
    <span class="fu">mcols</span>(<span class="kw">swath_spectra</span>)<span class="op">$</span><span class="kw">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">fenamiphos_ms1_peak</span>)]
</pre></div>
<p>We can now compare the reconstructed spectrum to the example consensus spectrum from the DDA experiment in the previous section (variable <code>ex_spectrum</code>) as well as to the MS2 spectrum for Fenamiphos from Metlin (with a collision energy of 10V).</p>
<div class="sourceCode" id="cb47"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(mfrow = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">fenamiphos_swath_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">ex_spectrum</span>[[<span class="fl">1</span>]],
     tolerance = <span class="fl">50e-6</span>, main = <span class="st">"against DDA"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">fenamiphos_swath_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">fenamiphos</span>[[<span class="fl">2</span>]],
     tolerance = <span class="fl">50e-6</span>, main = <span class="st">"against Metlin"</span>)
</pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/fena-swath-plot-1.png" alt="Mirror plot comparing the reconstructed MS2 spectrum for Fenamiphos (upper panel) against the measured spectrum from the DDA data and the Fenamiphhos spectrum from Metlin." width="1152"><p class="caption">
Mirror plot comparing the reconstructed MS2 spectrum for Fenamiphos (upper panel) against the measured spectrum from the DDA data and the Fenamiphhos spectrum from Metlin.
</p>
</div>
<p>If we wanted to get the EICs for the MS2 chromatographic peaks used to generate this MS2 spectrum we can use the IDs of these peaks which are provided with <code>$ms2_peak_id</code> of the result spectrum.</p>
<div class="sourceCode" id="cb48"><pre class="downlit">
<span class="kw">pk_ids</span> <span class="op">&lt;-</span> <span class="fu">mcols</span>(<span class="kw">fenamiphos_swath_spectrum</span>)<span class="op">$</span><span class="kw">ms2_peak_id</span>[[<span class="fl">1</span>]]
<span class="kw">pk_ids</span>
</pre></div>
<pre><code>##  [1] "CP199" "CP201" "CP211" "CP208" "CP200" "CP202" "CP217" "CP215" "CP205"
## [10] "CP212" "CP221" "CP223" "CP213" "CP207" "CP220"</code></pre>
<p>With these peak IDs available we can extract their retention time window and m/z ranges from the <code>chromPeaks</code> matrix and use the <code>chromatogram</code> function to extract their EIC. Note however that for SWATH data we have MS2 signal from different isolation windows. Thus we have to first filter the <code>swath_data</code> object by the isolation window containing the precursor m/z with the <code>filterIsolationWindow</code> to subset the data to MS2 spectra related to the ion of interest. In addition, we have to use <code>msLevel = 2L</code> in the <code>chromatogram</code> call because <code>chromatogram</code> extracts by default only data from MS1 spectra.</p>
<div class="sourceCode" id="cb50"><pre class="downlit">
<span class="kw">rt_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>)[<span class="kw">pk_ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rtmin"</span>, <span class="st">"rtmax"</span>)]
<span class="kw">mz_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>)[<span class="kw">pk_ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"mzmin"</span>, <span class="st">"mzmax"</span>)]

<span class="kw">pmz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/Spectrum2-class.html">precursorMz</a></span>(<span class="kw">fenamiphos_swath_spectrum</span>)[<span class="fl">1</span>]
<span class="kw">swath_data_iwindow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/MSnExp-class.html">filterIsolationWindow</a></span>(<span class="kw">swath_data</span>, mz = <span class="kw">pmz</span>)
</pre></div>
<pre><code>## Warning in object[keep]: Removed preprocessing results</code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit">
<span class="kw">ms2_eics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram,XCMSnExp-method.html">chromatogram</a></span>(<span class="kw">swath_data_iwindow</span>, rt = <span class="kw">rt_range</span>,
                         mz = <span class="kw">mz_range</span>, msLevel = <span class="fl">2L</span>)
</pre></div>
<p>Each row of this <code>ms2_eics</code> contains now the EIC of one of the MS2 chromatographic peaks.</p>
<p>As a second example we analyze the signal from an [M+H]+ ion with an m/z of 376.0381 (which would match <a href="https://en.wikipedia.org/wiki/Prochloraz">Prochloraz</a>). We first identify the MS1 chromatographic peak for that m/z and retrieve the reconstructed MS2 spectrum for that peak.</p>
<div class="sourceCode" id="cb53"><pre class="downlit">
<span class="kw">prochloraz_mz</span> <span class="op">&lt;-</span> <span class="fl">376.0381</span>

<span class="kw">prochloraz_ms1_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">swath_data</span>, msLevel = <span class="fl">1L</span>,
                                  mz = <span class="kw">prochloraz_mz</span>, ppm = <span class="fl">5</span>)
<span class="kw">prochloraz_ms1_peak</span>
</pre></div>
<pre><code>##            mz   mzmin    mzmax      rt   rtmin   rtmax     into     intb
## CP22 376.0373 376.037 376.0374 405.046 401.446 409.546 3664.051 3655.951
##          maxo  sn sample
## CP22 897.3923 278      1</code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit">
<span class="kw">prochloraz_swath_spectrum</span> <span class="op">&lt;-</span> <span class="kw">swath_spectra</span>[
    <span class="fu">mcols</span>(<span class="kw">swath_spectra</span>)<span class="op">$</span><span class="kw">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">prochloraz_ms1_peak</span>)]
</pre></div>
<p>In addition we identify the corresponding MS1 peak in the DDA data set, extract all measured MS2 chromatographic peaks and build the consensus spectrum from these.</p>
<div class="sourceCode" id="cb56"><pre class="downlit">
<span class="kw">prochloraz_dda_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span>(<span class="kw">dda_data</span>, msLevel = <span class="fl">1L</span>,
                                  mz = <span class="kw">prochloraz_mz</span>, ppm = <span class="fl">5</span>)
<span class="kw">prochloraz_dda_peak</span>
</pre></div>
<pre><code>##             mz    mzmin    mzmax      rt   rtmin   rtmax     into     intb
## CP034 376.0385 376.0378 376.0391 405.715 400.346 410.555 5088.528 5078.727
##           maxo  sn sample
## CP034 1350.633 436      1</code></pre>
<p>The retention times for the chromatographic peaks from the DDA and SWATH data match almost perfectly. Next we get the MS2 spectra for this peak.</p>
<div class="sourceCode" id="cb58"><pre class="downlit">
<span class="kw">prochloraz_dda_spectra</span> <span class="op">&lt;-</span> <span class="kw">dda_spectra</span>[
    <span class="fu">mcols</span>(<span class="kw">dda_spectra</span>)<span class="op">$</span><span class="kw">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="kw">prochloraz_dda_peak</span>)]
<span class="kw">prochloraz_dda_spectra</span>
</pre></div>
<pre><code>## MSpectra with 5 spectra and 1 metadata column(s):
##                    msLevel     rtime peaksCount |     peak_id
##                  &lt;integer&gt; &lt;numeric&gt;  &lt;integer&gt; | &lt;character&gt;
##   CP034.F1.S3253         2   401.438         84 |       CP034
##   CP034.F1.S3259         2   402.198         67 |       CP034
##   CP034.F1.S3306         2   404.677        277 |       CP034
##   CP034.F1.S3316         2   405.127        311 |       CP034
##   CP034.F1.S3325         2   405.877        383 |       CP034</code></pre>
<p>In total 5 spectra were measured, some with a relatively high number of peaks. Next we combine them into a consensus spectrum.</p>
<div class="sourceCode" id="cb60"><pre class="downlit">
<span class="kw">prochloraz_dda_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/combineSpectra.html">combineSpectra</a></span>(
    <span class="kw">prochloraz_dda_spectra</span>, method = <span class="kw">consensusSpectrum</span>, msd = <span class="fl">0</span>, ppm = <span class="fl">20</span>,
    minProp = <span class="fl">0.8</span>, weighted = <span class="fl">FALSE</span>, intensityFun = <span class="kw">median</span>, mzFun = <span class="kw">median</span>)
</pre></div>
<p>At last we load also the Prochloraz MS2 spectra (for different collision energies) from Metlin.</p>
<div class="sourceCode" id="cb61"><pre class="downlit">
<span class="kw">prochloraz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/pSet-class.html">spectra</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/readMgfData.html">readMgfData</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"mgf/metlin-68898.mgf"</span>, package = <span class="st">"xcms"</span>)))
</pre></div>
<p>To validate the reconstructed spectrum we plot it against the corresponding DDA spectrum and the MS2 spectrum for Prochloraz (for a collision energy of 10V) from Metlin.</p>
<div class="sourceCode" id="cb62"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(mfrow = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">2</span>))
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">prochloraz_swath_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">prochloraz_dda_spectrum</span>[[<span class="fl">1</span>]],
     tolerance = <span class="fl">50e-6</span>, main = <span class="st">"against DDA"</span>)
<span class="fu"><a href="https://rdrr.io/pkg/MSnbase/man/plot-methods.html">plot</a></span>(<span class="kw">prochloraz_swath_spectrum</span>[[<span class="fl">1</span>]], <span class="kw">prochloraz</span>[[<span class="fl">2</span>]],
     tolerance = <span class="fl">50e-6</span>, main = <span class="st">"against Metlin"</span>)
</pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/pro-swath-plot-1.png" alt="Mirror plot comparing the reconstructed MS2 spectrum for Prochloraz (upper panel) against the measured spectrum from the DDA data and the Prochloraz spectrum from Metlin." width="1152"><p class="caption">
Mirror plot comparing the reconstructed MS2 spectrum for Prochloraz (upper panel) against the measured spectrum from the DDA data and the Prochloraz spectrum from Metlin.
</p>
</div>
<p>The spectra fit relatively well. Interestingly, the peak representing the precursor (the right-most peak) seems to have a slightly shifted m/z value in the reconstructed spectrum.</p>
<p>Similar to the DDA data, the reconstructed MS2 spectra from SWATH data could be used in the annotation of the MS1 chromatographic peaks.</p>
<!-- Other compounds with some issues to check: -->
<!-- - 306.162857, Buprofezin, Metlin 68681, consensus matches OKish with -->
<!--   SWATH. Metlin matches the DDA nicely, but not the SWATH. -->
<!-- - 300.0301, Azaconazole, Metlin 72479. Intensities match nicely, but m/z are -->
<!--   shifted. Metlin 20V matches DDA perfectly. SWATH m/z shifted. -->
<!-- - 224.083201. Spectra match relatively well, some almost perfect matches, but -->
<!--   consensus does not match that nicely. Nothing on Metlin. -->
<!-- - 256.1095, Dimethachlor, Metlin 72494. Intensities OK, m/z shifted. Metlin 20V -->
<!--   matches DDA perfectly. SWATH m/z shifted. -->
</div>
</div>
<div id="outlook" class="section level1">
<h1 class="hasAnchor">
<a href="#outlook" class="anchor"></a>Outlook</h1>
<p>Currently, spectra data representation, handling and processing is being re-implemented as part of the <a href="https://rformassspectrometry.org">RforMassSpectrometry</a> initiative aiming at increasing the performance of methods and simplifying their use. Thus, parts of the workflow described here will be changed (improved) in future.</p>
<p>Along with these developments, improved matching strategies for larger data sets will be implemented as well as functionality to compare <code>Spectra</code> directly to reference MS2 spectra from public annotation resources (e.g. Massbank or HMDB).</p>
<p>Regarding SWATH data analysis, future development will involve improved selection of the correct MS2 chromatographic peaks considering also correlation with intensity values across several samples.</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb63"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
</pre></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.4 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] pander_0.6.3        xcms_3.11.5         MSnbase_2.15.3     
##  [4] ProtGenerics_1.21.0 S4Vectors_0.27.12   mzR_2.23.0         
##  [7] Rcpp_1.0.4.6        BiocParallel_1.23.0 Biobase_2.49.0     
## [10] BiocGenerics_0.35.4 BiocStyle_2.17.0   
## 
## loaded via a namespace (and not attached):
##  [1] vsn_3.57.0                  foreach_1.5.0              
##  [3] assertthat_0.2.1            highr_0.8                  
##  [5] BiocManager_1.30.10         affy_1.67.0                
##  [7] GenomeInfoDbData_1.2.3      yaml_2.2.1                 
##  [9] robustbase_0.93-6           impute_1.63.0              
## [11] pillar_1.4.4                backports_1.1.8            
## [13] lattice_0.20-41             glue_1.4.1                 
## [15] limma_3.45.7                digest_0.6.25              
## [17] GenomicRanges_1.41.5        RColorBrewer_1.1-2         
## [19] XVector_0.29.2              colorspace_1.4-1           
## [21] htmltools_0.5.0             preprocessCore_1.51.0      
## [23] Matrix_1.2-18               plyr_1.8.6                 
## [25] MALDIquant_1.19.3           XML_3.99-0.3               
## [27] pkgconfig_2.0.3             bookdown_0.19              
## [29] zlibbioc_1.35.0             scales_1.1.1               
## [31] RANN_2.6.1                  affyio_1.59.0              
## [33] tibble_3.0.1                farver_2.0.3               
## [35] IRanges_2.23.10             ggplot2_3.3.2              
## [37] ellipsis_0.3.1              SummarizedExperiment_1.19.5
## [39] MassSpecWavelet_1.55.0      magrittr_1.5               
## [41] crayon_1.3.4                memoise_1.1.0              
## [43] evaluate_0.14               fs_1.4.1                   
## [45] ncdf4_1.17                  doParallel_1.0.15          
## [47] MASS_7.3-51.5               tools_4.0.0                
## [49] lifecycle_0.2.0             matrixStats_0.56.0         
## [51] stringr_1.4.0               munsell_0.5.0              
## [53] DelayedArray_0.15.5         pcaMethods_1.81.0          
## [55] compiler_4.0.0              pkgdown_1.5.1.9000         
## [57] GenomeInfoDb_1.25.3         mzID_1.27.0                
## [59] rlang_0.4.6                 grid_4.0.0                 
## [61] RCurl_1.98-1.2              iterators_1.0.12           
## [63] MsCoreUtils_1.1.2           labeling_0.3               
## [65] bitops_1.0-6                rmarkdown_2.3.1            
## [67] gtable_0.3.0                codetools_0.2-16           
## [69] R6_2.4.1                    knitr_1.28                 
## [71] rprojroot_1.3-2             desc_1.2.0                 
## [73] stringi_1.4.6               vctrs_0.3.1                
## [75] DEoptimR_1.0-8              xfun_0.15</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Ludwig:2018hv">
<p>1. Ludwig C, Gillet L, Rosenberger G, Amon S, Collins BC, Aebersold R: <strong>Data-independent acquisition-based SWATH-MS for quantitative proteomics: a tutorial.</strong> <em>Molecular systems biology</em> 2018, <strong>14</strong>:e8126.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Steffen Neumann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
