<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LCMS data preprocessing and analysis with xcms • xcms</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LCMS data preprocessing and analysis with xcms">
<meta property="og:description" content="xcms">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xcms</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.21.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/xcms.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/LC-MS-feature-grouping.html">Compounding (grouping) of LC-MS features</a>
    </li>
    <li>
      <a href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sneumann/xcms/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LCMS data preprocessing and analysis with
xcms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/HEAD/vignettes/xcms.Rmd" class="external-link"><code>vignettes/xcms.Rmd</code></a></small>
      <div class="hidden name"><code>xcms.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.17/xcms" class="external-link">xcms</a></em><br><strong>Authors</strong>: Johannes Rainer<br><strong>Modified</strong>: 2023-01-16 11:30:08.912539<br><strong>Compiled</strong>: Mon Jan 16 12:16:57 2023</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This documents describes data import, exploration, preprocessing and
analysis of LCMS experiments with <code>xcms</code> version &gt;= 3. The
examples and basic workflow was adapted from the original <em>LC/MS
Preprocessing and Analysis with xcms</em> vignette from Colin A.
Smith.</p>
<p>The new user interface and methods use the <code>XCMSnExp</code>
object (instead of the <em>old</em> <code>xcmsSet</code> object) as a
container for the pre-processing results. To support packages and
pipelines relying on the <code>xcmsSet</code> object, it is however
possible to convert an <code>XCMSnExp</code> into a <code>xcmsSet</code>
object using the <code>as</code> method
(i.e.<code>xset &lt;- as(x, "xcmsSet")</code>, with <code>x</code> being
an <code>XCMSnExp</code> object.</p>
</div>
<div class="section level2">
<h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h2>
<p><code>xcms</code> supports analysis of LC/MS data from files in
(AIA/ANDI) NetCDF, mzXML and mzML format. For the actual data import
Bioconductor’s <em><a href="https://bioconductor.org/packages/3.17/mzR" class="external-link">mzR</a></em> is used.
For demonstration purpose we will analyze a subset of the data from
<span class="citation">[1]</span> in which the metabolic consequences of
knocking out the fatty acid amide hydrolase (FAAH) gene in mice was
investigated. The raw data files (in NetCDF format) are provided with
the <code>faahKO</code> data package. The data set consists of samples
from the spinal cords of 6 knock-out and 6 wild-type mice. Each file
contains data in centroid mode acquired in positive ion mode form
200-600 m/z and 2500-4500 seconds. To speed up processing of this
vignette we will restrict the analysis to only 8 files and to the
retention time range from 2500 to 3500 seconds.</p>
<p>Below we load all required packages, locate the raw CDF files within
the <code>faahKO</code> package and build a <em>phenodata</em> data
frame describing the experimental setup. Note that for <em>real</em>
experiments it is suggested to define a file (table) that contains the
file names of the raw data files along with descriptions of the samples
for each file as additional columns. Such a file could then be imported
with e.g. <code>read.table</code> as variable <code>pd</code> (instead
of being defined within R as in the example below) and the file names
could be passed along to the <code>readMSData</code> function below with
e.g. <code>files = paste0(MZML_PATH, "/", pd$mzML_file)</code> where
<code>MZML_PATH</code> would be the path to directory in which the files
are located and <code>"mzML_file"</code> the name of the column in the
phenodata file that contains the file names.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/xcms" class="external-link">xcms</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.doi.org/10.1021/bi0480335" class="external-link">faahKO</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Get the full path to the CDF files</span></span>
<span><span class="va">cdfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"cdf"</span>, package <span class="op">=</span> <span class="st">"faahKO"</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Create a phenodata data.frame</span></span>
<span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">cdfs</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">".CDF"</span>,</span>
<span>                                   replacement <span class="op">=</span> <span class="st">""</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                 sample_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"KO"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Subsequently we load the raw data as an <code>OnDiskMSnExp</code>
object using the <code>readMSData</code> method from the <em><a href="https://bioconductor.org/packages/3.17/MSnbase" class="external-link">MSnbase</a></em>
package. The <code>MSnbase</code> provides based structures and
infrastructure for the processing of mass spectrometry data. Also,
<code>MSnbase</code> can be used to <em>centroid</em> profile-mode MS
data (see the corresponding vignette in the <code>MSnbase</code>
package).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/readMSData.html" class="external-link">readMSData</a></span><span class="op">(</span>files <span class="op">=</span> <span class="va">cdfs</span>, pdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">"NAnnotatedDataFrame"</span>, <span class="va">pd</span><span class="op">)</span>,</span>
<span>                       mode <span class="op">=</span> <span class="st">"onDisk"</span><span class="op">)</span></span></code></pre></div>
<p>We next restrict the data set to the retention time range from 2500
to 3500 seconds. This is merely to reduce the processing time of this
vignette.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">raw_data</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2500</span>, <span class="fl">3500</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The resulting <code>OnDiskMSnExp</code> object contains general
information about the number of spectra, retention times, the measured
total ion current etc, but does not contain the full raw data (i.e. the
m/z and intensity values from each measured spectrum). Its memory
footprint is thus rather small making it an ideal object to represent
large metabolomics experiments while allowing to perform simple quality
controls, data inspection and exploration as well as data sub-setting
operations. The m/z and intensity values are imported from the raw data
files on demand, hence the location of the raw data files should not be
changed after initial data import.</p>
</div>
<div class="section level2">
<h2 id="initial-data-inspection">Initial data inspection<a class="anchor" aria-label="anchor" href="#initial-data-inspection"></a>
</h2>
<p>The <code>OnDiskMSnExp</code> organizes the MS data by spectrum and
provides the methods <code>intensity</code>, <code>mz</code> and
<code>rtime</code> to access the raw data from the files (the measured
intensity values, the corresponding m/z and retention time values). In
addition, the <code>spectra</code> method could be used to return all
data encapsulated in <code>Spectrum</code> objects. Below we extract the
retention time values from the object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">raw_data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0001 F1.S0002 F1.S0003 F1.S0004 F1.S0005 F1.S0006 </span></span>
<span><span class="co">## 2501.378 2502.943 2504.508 2506.073 2507.638 2509.203</span></span></code></pre>
<p>All data is returned as one-dimensional vectors (a numeric vector for
<code>rtime</code> and a <code>list</code> of numeric vectors for
<code>mz</code> and <code>intensity</code>, each containing the values
from one spectrum), even if the experiment consists of multiple
files/samples. The <code>fromFile</code> function returns an integer
vector providing the mapping of the values to the originating file.
Below we use the <code>fromFile</code> indices to organize the
<code>mz</code> values by file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">mz</a></span><span class="op">(</span><span class="va">raw_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Split the list by file</span></span>
<span><span class="va">mzs_by_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">mzs</span>, f <span class="op">=</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">raw_data</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mzs_by_file</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 8</span></span></code></pre>
<p>As a first evaluation of the data we plot below the base peak
chromatogram (BPC) for each file in our experiment. We use the
<code>chromatogram</code> method and set the <code>aggregationFun</code>
to <code>"max"</code> to return for each spectrum the maximal intensity
and hence create the BPC from the raw data. To create a total ion
chromatogram we could set <code>aggregationFun</code> to
<code>sum</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the base peak chromatograms. This reads data from the files.</span></span>
<span><span class="va">bpis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">raw_data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="co">## Define colors for the two groups</span></span>
<span><span class="va">group_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"Set1"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"60"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">group_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot all chromatograms.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpis</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">raw_data</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms_files/figure-html/data-inspection-bpc-1.png" width="1152" style="display: block; margin: auto;"></p>
<p>The <code>chromatogram</code> method returned a
<code>MChromatograms</code> object that organizes individual
<code>Chromatogram</code> objects (which in fact contain the
chromatographic data) in a two-dimensional array: columns represent
samples and rows (optionally) m/z and/or retention time ranges. Below we
extract the chromatogram of the first sample and access its retention
time and intensity values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bpi_1</span> <span class="op">&lt;-</span> <span class="va">bpis</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">bpi_1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0001 F1.S0002 F1.S0003 F1.S0004 F1.S0005 F1.S0006 </span></span>
<span><span class="co">## 2501.378 2502.943 2504.508 2506.073 2507.638 2509.203</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">bpi_1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0001 F1.S0002 F1.S0003 F1.S0004 F1.S0005 F1.S0006 </span></span>
<span><span class="co">##    43888    43960    43392    42632    42200    42288</span></span></code></pre>
<p>The <code>chromatogram</code> method supports also extraction of
chromatographic data from a m/z-rt slice of the MS data. In the next
section we will use this method to create an extracted ion chromatogram
(EIC) for a selected peak.</p>
<p>Note that <code>chromatogram</code> reads the raw data from each file
to calculate the chromatogram. The <code>bpi</code> and <code>tic</code>
methods on the other hand do not read any data from the raw files but
use the respective information that was provided in the header
definition of the input files (which might be different from the actual
data).</p>
<p>Below we create boxplots representing the distribution of total ion
currents per file. Such plots can be very useful to spot problematic or
failing MS runs.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the total ion current by file</span></span>
<span><span class="va">tc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">tic</a></span><span class="op">(</span><span class="va">raw_data</span><span class="op">)</span>, f <span class="op">=</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">raw_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">tc</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">raw_data</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"intensity"</span>, main <span class="op">=</span> <span class="st">"Total ion current"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-tic-boxplot-1.png" alt="Distribution of total ion currents per file." width="768"><p class="caption">
Distribution of total ion currents per file.
</p>
</div>
<p>Also, we can cluster the samples based on similarity of their base
peak chromatogram. This can also be helpful to spot potentially
problematic samples in an experiment or generally get an initial
overview of the sample grouping in the experiment. Since the retention
times between samples are not exactly identical, we use the
<code>bin</code> function to group intensities in fixed time ranges
(bins) along the retention time axis. In the present example we use a
bin size of 1 second, the default is 0.5 seconds. The clustering is
performed using complete linkage hierarchical clustering on the pairwise
correlations of the binned base peak chromatograms.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Bin the BPC</span></span>
<span><span class="va">bpis_bin</span> <span class="op">&lt;-</span> <span class="fu">MSnbase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">bin</a></span><span class="op">(</span><span class="va">bpis</span>, binSize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate correlation on the log2 transformed base peak intensities</span></span>
<span><span class="va">cormat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bpis_bin</span>, <span class="va">intensity</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cormat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cormat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">raw_data</span><span class="op">$</span><span class="va">sample_name</span></span>
<span></span>
<span><span class="co">## Define which phenodata columns should be highlighted in the plot</span></span>
<span><span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">raw_data</span><span class="op">$</span><span class="va">sample_group</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ann</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">raw_data</span><span class="op">$</span><span class="va">sample_name</span></span>
<span></span>
<span><span class="co">## Perform the cluster analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">cormat</span>, annotation <span class="op">=</span> <span class="va">ann</span>,</span>
<span>         annotation_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">group_colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-bpc-heatmap-1.png" alt="Grouping of samples based on similarity of their base peak chromatogram." width="672"><p class="caption">
Grouping of samples based on similarity of their base peak chromatogram.
</p>
</div>
<p>The samples cluster in a pairwise manner, the KO and WT samples for
the sample index having the most similar BPC.</p>
</div>
<div class="section level2">
<h2 id="chromatographic-peak-detection">Chromatographic peak detection<a class="anchor" aria-label="anchor" href="#chromatographic-peak-detection"></a>
</h2>
<p>Next we perform the chromatographic peak detection using the
<em>centWave</em> algorithm <span class="citation">[2]</span>. Before
running the peak detection it is however strongly suggested to visually
inspect e.g. the extracted ion chromatogram of internal standards or
known compounds to evaluate and adapt the peak detection settings since
the default settings will not be appropriate for most LCMS experiments.
The two most critical parameters for <em>centWave</em> are the
<code>peakwidth</code> (expected range of chromatographic peak widths)
and <code>ppm</code> (maximum expected deviation of m/z values of
centroids corresponding to one chromatographic peak; this is usually
much larger than the ppm specified by the manufacturer) parameters. To
evaluate the typical chromatographic peak width we plot the EIC for one
peak.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define the rt and m/z range of the peak area</span></span>
<span><span class="va">rtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2700</span>, <span class="fl">2900</span><span class="op">)</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">334.9</span>, <span class="fl">335.1</span><span class="op">)</span></span>
<span><span class="co">## extract the chromatogram</span></span>
<span><span class="va">chr_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">raw_data</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_raw</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-plot-eic-1.png" alt="Extracted ion chromatogram for one peak." width="768"><p class="caption">
Extracted ion chromatogram for one peak.
</p>
</div>
<p>Note that <code>Chromatogram</code> objects extracted by the
<code>chromatogram</code> method contain an <code>NA</code> value if in
a certain scan (i.e. for a specific retention time) no signal was
measured in the respective mz range. This is reflected by the lines not
being drawn as continuous lines in the plot above.</p>
<p>The peak above has a width of about 50 seconds. The
<code>peakwidth</code> parameter should be set to accommodate the
expected widths of peak in the data set. We set it to <code>20,80</code>
for the present example data set.</p>
<p>For the <code>ppm</code> parameter we extract the full MS data
(intensity, retention time and m/z values) corresponding to the above
peak. To this end we first filter the raw object by retention time, then
by m/z and finally plot the object with <code>type = "XIC"</code> to
produce the plot below. We use the <em>pipe</em> (<code>%&gt;%</code>)
command better illustrate the corresponding workflow. Note also that in
this type of plot identified chromatographic peaks would be indicated by
default if present.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"XIC"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-detection-plot-ms-data-1.png" alt="Visualization of the raw MS data for one peak. For each plot: upper panel: chromatogram plotting the intensity values against the retention time, lower panel m/z against retention time plot. The individual data points are colored according to the intensity." width="1344"><p class="caption">
Visualization of the raw MS data for one peak. For each plot: upper
panel: chromatogram plotting the intensity values against the retention
time, lower panel m/z against retention time plot. The individual data
points are colored according to the intensity.
</p>
</div>
<p>In the present data there is actually no variation in the m/z values.
Usually one would see the m/z values (lower panel) scatter around the
<em>real</em> m/z value of the compound. The first step of the
<em>centWave</em> algorithm defines so called regions of interest (ROI)
based on the difference of m/z values from consecutive scans. In detail,
m/z values from consecutive scans are included into a ROI if the
difference between the m/z and the mean m/z of the ROI is smaller than
the user defined <code>ppm</code> parameter. A reasonable choice for the
<code>ppm</code> could thus be the maximal m/z difference of data points
from neighboring scans/spectra that are part of the chromatographic
peak. It is suggested to inspect the ranges of m/z values for many
compounds (either internal standards or compounds known to be present in
the sample) and define the <code>ppm</code> parameter for
<em>centWave</em> according to these.</p>
<p>Note that we can also perform the peak detection on the extracted ion
chromatogram. This can help to evaluate different peak detection
settings. Only be aware that peak detection on an extracted ion
chromatogram will not consider the <code>ppm</code> parameter and that
the estimation of the background signal is different to the peak
detection on the full data set; values for the <code>snthresh</code>
will hence have different consequences. Below we perform the peak
detection with the <code>findChromPeaks</code> function on the extracted
ion chromatogram. The submitted <em>parameter</em> object defines which
algorithm will be used and allows to define the settings for this
algorithm. We use the <em>centWave</em> algorithm with default settings,
except for <code>snthresh</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xchr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span><span class="op">(</span><span class="va">chr_raw</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>snthresh <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can access the identified chromatographic peaks with the
<code>chromPeaks</code> function.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            rt    rtmin    rtmax       into       intb  maxo  sn row column</span></span>
<span><span class="co">## [1,] 2781.505 2761.160 2809.674  412134.25  355516.37 16856  13   1      1</span></span>
<span><span class="co">## [2,] 2786.199 2764.290 2812.803 1496244.21 1391821.33 58736  20   1      2</span></span>
<span><span class="co">## [3,] 2734.556 2714.211 2765.855   21579.37   18449.43   899   4   1      3</span></span>
<span><span class="co">## [4,] 2797.154 2775.245 2815.933  159058.78  150289.31  6295  12   1      3</span></span>
<span><span class="co">## [5,] 2784.635 2761.160 2808.109   54947.54   37923.53  2715   2   1      4</span></span>
<span><span class="co">## [6,] 2859.752 2845.668 2878.532   13895.21   13874.87   905 904   1      4</span></span></code></pre>
<p>Parallel to the <code>chromPeaks</code> matrix there is also a data
frame <code>chromPeakData</code> that allows to add arbitrary
annotations to each chromatographic peak. Below we extract this data
frame that by default contains only the MS level in which the peak was
identified.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 12 rows and 4 columns</span></span>
<span><span class="co">##      ms_level is_filled       row    column</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;logical&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1     FALSE         1         1</span></span>
<span><span class="co">## 2           1     FALSE         1         2</span></span>
<span><span class="co">## 3           1     FALSE         1         3</span></span>
<span><span class="co">## 4           1     FALSE         1         3</span></span>
<span><span class="co">## 5           1     FALSE         1         4</span></span>
<span><span class="co">## ...       ...       ...       ...       ...</span></span>
<span><span class="co">## 8           1     FALSE         1         4</span></span>
<span><span class="co">## 9           1     FALSE         1         5</span></span>
<span><span class="co">## 10          1     FALSE         1         6</span></span>
<span><span class="co">## 11          1     FALSE         1         7</span></span>
<span><span class="co">## 12          1     FALSE         1         8</span></span></code></pre>
<p>Next we plot the identified chromatographic peaks in the extracted
ion chromatogram. We use the <code>col</code> parameter to color the
individual chromatogram lines. Colors can also be specified for the
identified peaks, <code>peakCol</code> for the foreground/border color,
<code>peakBg</code> for the background/fill color. One color has to be
provided for each chromatographic peak listed by
<code>chromPeaks</code>. Below we define a color to indicate the sample
group from which the sample is and use the sample information in the
peaks’ <code>"sample"</code> column to assign the correct color to each
chromatographic peak. More peak highlighting options are described
further below.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_colors</span> <span class="op">&lt;-</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">xchr</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">xchr</span>, col <span class="op">=</span> <span class="va">sample_colors</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"column"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-eic-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. Peak area of identified chromatographic peaks are highlighted in the sample group color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. Peak area of identified chromatographic
peaks are highlighted in the sample group color.
</p>
</div>
<p>Finally we perform the chromatographic peak detection on the full
data set. Note that we set the argument <code>prefilter</code> to
<code>c(6, 5000)</code> and <code>noise</code> to <code>5000</code> to
reduce the run time of this vignette. With this setting we consider only
signals with a value larger than 5000 in the peak detection step.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">80</span><span class="op">)</span>, noise <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                     prefilter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatographic-peak-detection.html">findChromPeaks</a></span><span class="op">(</span><span class="va">raw_data</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span></code></pre></div>
<p>The results are returned as an <code>XCMSnExp</code> object which
extends the <code>OnDiskMSnExp</code> object by storing also LC/GC-MS
preprocessing results. This means also that all methods to sub-set and
filter the data or to access the (raw) data are inherited from the
<code>OnDiskMSnExp</code> object and can thus be re-used. Note also that
it is possible to perform additional rounds of peak detection (e.g. on
MS level &gt; 1 data) on the <code>xdata</code> object by calling
<code>findChromPeaks</code> with the parameter
<code>add = TRUE</code>.</p>
<p>The results from the chromatographic peak detection can be accessed
with the <code>chromPeaks</code> method.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz mzmin mzmax       rt    rtmin    rtmax      into      intb  maxo</span></span>
<span><span class="co">## CP0001 453.2 453.2 453.2 2509.203 2501.378 2527.982 1007409.0 1007380.8 38152</span></span>
<span><span class="co">## CP0002 236.1 236.1 236.1 2518.593 2501.378 2537.372  253501.0  226896.3 12957</span></span>
<span><span class="co">## CP0003 594.0 594.0 594.0 2601.535 2581.191 2637.529  161042.2  149297.3  7850</span></span>
<span><span class="co">## CP0004 577.0 577.0 577.0 2604.665 2581.191 2626.574  136105.2  129195.5  6215</span></span>
<span><span class="co">## CP0005 369.2 369.2 369.2 2587.451 2556.151 2631.269  483852.3  483777.1  7215</span></span>
<span><span class="co">## CP0006 369.2 369.2 369.2 2568.671 2557.716 2578.061  144624.8  144602.9  7033</span></span>
<span><span class="co">##           sn sample</span></span>
<span><span class="co">## CP0001 38151      1</span></span>
<span><span class="co">## CP0002    11      1</span></span>
<span><span class="co">## CP0003    13      1</span></span>
<span><span class="co">## CP0004    13      1</span></span>
<span><span class="co">## CP0005  7214      1</span></span>
<span><span class="co">## CP0006  7032      1</span></span></code></pre>
<p>The returned <code>matrix</code> provides the m/z and retention time
range for each identified chromatographic peak as well as the integrated
signal intensity (“into”) and the maximal peak intensitity (“maxo”).
Columns “sample” contains the index of the sample in the
object/experiment in which the peak was identified.</p>
<p>Annotations for each individual peak can be extracted with the
<code>chromPeakData</code> function. This data frame could also be used
to add/store arbitrary annotations for each detected peak.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 1707 rows and 2 columns</span></span>
<span><span class="co">##         ms_level is_filled</span></span>
<span><span class="co">##        &lt;integer&gt; &lt;logical&gt;</span></span>
<span><span class="co">## CP0001         1     FALSE</span></span>
<span><span class="co">## CP0002         1     FALSE</span></span>
<span><span class="co">## CP0003         1     FALSE</span></span>
<span><span class="co">## CP0004         1     FALSE</span></span>
<span><span class="co">## CP0005         1     FALSE</span></span>
<span><span class="co">## ...          ...       ...</span></span>
<span><span class="co">## CP1703         1     FALSE</span></span>
<span><span class="co">## CP1704         1     FALSE</span></span>
<span><span class="co">## CP1705         1     FALSE</span></span>
<span><span class="co">## CP1706         1     FALSE</span></span>
<span><span class="co">## CP1707         1     FALSE</span></span></code></pre>
<p>Peak detection will not always work perfectly leading to peak
detection artifacts, such as overlapping peaks or artificially split
peaks. The <code>refineChromPeaks</code> function allows to
<em>refine</em> peak detection results by either removing identified
peaks not passing a certain criteria or by merging artificially split
chromatographic peaks. With parameter objects
<code>CleanPeaksParam</code> and <code>FilterIntensityParam</code> it is
possible to remove peaks with a retention time range or intensities
below a threshold, respectively (see their respective help pages for
more details and examples). With <code>MergeNeighboringPeaksParam</code>
it is possible to merge chromatographic peaks. Below we post-process the
peak detection results merging peaks overlapping in a 4 second window
per file if the signal between in between them is lower than 75% of the
smaller peak’s maximal intensity. See the
<code>MergeNeighboringPeaksParam</code> help page for a detailed
description of the settings and the approach.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mpp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-merge.html">MergeNeighboringPeaksParam</a></span><span class="op">(</span>expandRt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">xdata_pp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-clean.html">refineChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, <span class="va">mpp</span><span class="op">)</span></span></code></pre></div>
<p>An example for a merged peak is given below.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzr_1</span> <span class="op">&lt;-</span> <span class="fl">305.1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata</span>, <span class="fl">1</span><span class="op">)</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="va">chr_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata_pp</span>, <span class="fl">1</span><span class="op">)</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The splitted peak was merged into one." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing,
right: after refinement. The splitted peak was merged into one.
</p>
</div>
<p>For the first trace in the chromatogram above centWave detected 3
peaks (1 for the full area and two smaller ones, see left panel in the
plot above). The peak refinement with
<code>MergeNeighboringPeaksParam</code> reduced them to a single peak
(right panel in the figure above). Note that this refinement does not
merge neighboring peaks for which the signal in between them is lower
than a certain proportion (see figure below).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzr_1</span> <span class="op">&lt;-</span> <span class="fl">496.2</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata</span>, <span class="fl">1</span><span class="op">)</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="va">chr_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata_pp</span>, <span class="fl">1</span><span class="op">)</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-not-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The peaks were not merged." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing,
right: after refinement. The peaks were not merged.
</p>
</div>
<p>Note also that it is possible to perform the peak refinement on
extracted ion chromatograms. This could e.g. be used to fine-tune the
settings for the parameter. To illustrate this we perform below a peak
refinement on the extracted ion chromatogram <code>chr_1</code> reducing
the <code>minProp</code> parameter to force joining the two peaks.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks-clean.html">refineChromPeaks</a></span><span class="op">(</span><span class="va">chr_1</span>, <span class="fu"><a href="../reference/refineChromPeaks-merge.html">MergeNeighboringPeaksParam</a></span><span class="op">(</span>minProp <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax     into intb    maxo   sn</span></span>
<span><span class="co">## CPM1 496.2 496.2 496.2 3384.012 3294.809 3412.181 45940118   NA 1128960 1255</span></span>
<span><span class="co">##      sample row column</span></span>
<span><span class="co">## CPM1      1   1      1</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms_files/figure-html/peak-postprocessing-chr-1.png" width="700"></p>
<p>Before proceeding we replace the <code>xdata</code> object with the
results from the peak refinement.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="va">xdata_pp</span></span></code></pre></div>
<p>Below we use the data from the <code>chromPeaks</code> matrix to
calculate some per-file summaries.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>peak_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">z</span><span class="op">[</span>, <span class="st">"rtmax"</span><span class="op">]</span> <span class="op">-</span> <span class="va">z</span><span class="op">[</span>, <span class="st">"rtmin"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="cn">T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split.data.frame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span>, f <span class="op">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    FUN <span class="op">=</span> <span class="va">summary_fun</span><span class="op">)</span></span>
<span><span class="cn">T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="cn">T</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">fileNames</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html" class="external-link">pandoc.table</a></span><span class="op">(</span></span>
<span>    <span class="cn">T</span>,</span>
<span>    caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Summary statistics on identified chromatographic"</span>,</span>
<span>                     <span class="st">" peaks. Shown are number of identified peaks per"</span>,</span>
<span>                     <span class="st">" sample and widths/duration of chromatographic "</span>,</span>
<span>                     <span class="st">"peaks."</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Summary statistics on identified chromatographic peaks. Shown
are number of identified peaks per sample and widths/duration of
chromatographic peaks.</caption>
<colgroup>
<col width="20%">
<col width="17%">
<col width="10%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">peak_count</th>
<th align="center">rt.0%</th>
<th align="center">rt.25%</th>
<th align="center">rt.50%</th>
<th align="center">rt.75%</th>
<th align="center">rt.100%</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>ko15.CDF</strong></td>
<td align="center">224</td>
<td align="center">10.95</td>
<td align="center">29.73</td>
<td align="center">40.69</td>
<td align="center">51.64</td>
<td align="center">319.2</td>
</tr>
<tr class="even">
<td align="center"><strong>ko16.CDF</strong></td>
<td align="center">280</td>
<td align="center">9.39</td>
<td align="center">32.86</td>
<td align="center">44.6</td>
<td align="center">54.77</td>
<td align="center">150.2</td>
</tr>
<tr class="odd">
<td align="center"><strong>ko21.CDF</strong></td>
<td align="center">132</td>
<td align="center">10.95</td>
<td align="center">42.25</td>
<td align="center">50.08</td>
<td align="center">64.55</td>
<td align="center">164.3</td>
</tr>
<tr class="even">
<td align="center"><strong>ko22.CDF</strong></td>
<td align="center">149</td>
<td align="center">10.95</td>
<td align="center">28.17</td>
<td align="center">45.38</td>
<td align="center">57.9</td>
<td align="center">118.9</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt15.CDF</strong></td>
<td align="center">260</td>
<td align="center">10.95</td>
<td align="center">27.78</td>
<td align="center">42.25</td>
<td align="center">51.64</td>
<td align="center">161.2</td>
</tr>
<tr class="even">
<td align="center"><strong>wt16.CDF</strong></td>
<td align="center">180</td>
<td align="center">10.95</td>
<td align="center">34.43</td>
<td align="center">45.38</td>
<td align="center">54.77</td>
<td align="center">142.4</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt21.CDF</strong></td>
<td align="center">147</td>
<td align="center">9.389</td>
<td align="center">32.08</td>
<td align="center">46.95</td>
<td align="center">63.38</td>
<td align="center">133</td>
</tr>
<tr class="even">
<td align="center"><strong>wt22.CDF</strong></td>
<td align="center">204</td>
<td align="center">10.95</td>
<td align="center">34.43</td>
<td align="center">43.82</td>
<td align="center">56.34</td>
<td align="center">137.7</td>
</tr>
</tbody>
</table>
<p>We can also plot the location of the identified chromatographic peaks
in the m/z - retention time space for one file using the
<code>plotChromPeaks</code> function. Below we plot the chromatographic
peaks for the 3rd sample.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, file <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peaks-plot-1.png" alt="Identified chromatographic peaks in the m/z by retention time space for one sample." width="768"><p class="caption">
Identified chromatographic peaks in the m/z by retention time space for
one sample.
</p>
</div>
<p>To get a global overview of the peak detection we can plot the
frequency of identified peaks per file along the retention time axis.
This allows to identify time periods along the MS run in which a higher
number of peaks was identified and evaluate whether this is consistent
across files.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeakImage</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-image-1.png" alt="Frequency of identified chromatographic peaks along the retention time axis. The frequency is color coded with higher frequency being represented by yellow-white. Each line shows the peak frequency for one file." width="960"><p class="caption">
Frequency of identified chromatographic peaks along the retention time
axis. The frequency is color coded with higher frequency being
represented by yellow-white. Each line shows the peak frequency for one
file.
</p>
</div>
<p>Next we highlight the identified chromatographic peaks for the
example peak from before. Evaluating such plots on a list of peaks
corresponding to known peaks or internal standards helps to ensure that
peak detection settings were appropriate and correctly identified the
expected peaks. We extract the ion chromatogram from the peak detection
result object, which contains then also the identified chromatographic
peaks for that ion that we can extract with the <code>chromPeaks</code>
function.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax      into      intb  maxo sn</span></span>
<span><span class="co">## CP0045 335   335   335 2781.505 2761.160 2809.674  412134.3  383167.4 16856 23</span></span>
<span><span class="co">## CP0309 335   335   335 2786.199 2764.290 2812.803 1496244.2 1461187.3 58736 72</span></span>
<span><span class="co">## CP0587 335   335   335 2797.154 2775.245 2815.933  159058.8  149229.6  6295 13</span></span>
<span><span class="co">## CP1194 335   335   335 2786.199 2764.290 2812.803  932645.2  915333.8 35856 66</span></span>
<span><span class="co">## CP1378 335   335   335 2792.461 2768.987 2823.760  876585.5  848569.1 27200 36</span></span>
<span><span class="co">##        sample row column</span></span>
<span><span class="co">## CP0045      1   1      1</span></span>
<span><span class="co">## CP0309      2   1      2</span></span>
<span><span class="co">## CP0587      3   1      3</span></span>
<span><span class="co">## CP1194      6   1      6</span></span>
<span><span class="co">## CP1378      7   1      7</span></span></code></pre>
<p>We can also plot the extracted ion chromatogram. Identified
chromatographic peaks will be automatically highlighted in the plot.
Below we highlight chromatographic peaks with a rectangle from the
peak’s minimal to maximal rt and from an intensity of 0 to the maximal
signal of the peak.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_colors</span> <span class="op">&lt;-</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_ex</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_ex</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, peakType <span class="op">=</span> <span class="st">"rectangle"</span>,</span>
<span>     peakCol <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The rectangles indicate the identified chromatographic peaks per sample." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. The rectangles indicate the identified
chromatographic peaks per sample.
</p>
</div>
<p>Alternatively to the rectangle visualization above, it is possible to
represent the apex position of each peak with a single point (passing
argument <code>type = "point"</code> to the function), or draw the
actually identified peak by specifying <code>type = "polygon"</code>. To
completely omit highlighting the identified peaks (e.g. to plot base
peak chromatograms or similar) <code>type = "none"</code> can be used.
Below we use <code>type = "polygon"</code> to fill the peak area for
each identified chromatographic peak in each sample. Whether individual
peaks can be still identified in such a plot depends however on the
number of samples from which peaks are drawn.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_ex</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-polygon-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The signal area of identified chromatographic peaks are filled with a color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. The signal area of identified
chromatographic peaks are filled with a color.
</p>
</div>
<p>Note that we can also specifically extract identified chromatographic
peaks for a selected region by providing the respective m/z and
retention time ranges with the <code>mz</code> and <code>rt</code>
arguments in the <code>chromPeaks</code> method.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pander.html" class="external-link">pander</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span>,</span>
<span>       caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Identified chromatographic peaks in a selected "</span>,</span>
<span>                       <span class="st">"m/z and retention time range."</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<caption>Identified chromatographic peaks in a selected m/z and
retention time range. (continued below)</caption>
<colgroup>
<col width="16%">
<col width="7%">
<col width="10%">
<col width="10%">
<col width="8%">
<col width="10%">
<col width="10%">
<col width="12%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">mz</th>
<th align="center">mzmin</th>
<th align="center">mzmax</th>
<th align="center">rt</th>
<th align="center">rtmin</th>
<th align="center">rtmax</th>
<th align="center">into</th>
<th align="center">intb</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>CP0045</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2782</td>
<td align="center">2761</td>
<td align="center">2810</td>
<td align="center">412134</td>
<td align="center">383167</td>
</tr>
<tr class="even">
<td align="center"><strong>CP0309</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2786</td>
<td align="center">2764</td>
<td align="center">2813</td>
<td align="center">1496244</td>
<td align="center">1461187</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP0587</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2797</td>
<td align="center">2775</td>
<td align="center">2816</td>
<td align="center">159059</td>
<td align="center">149230</td>
</tr>
<tr class="even">
<td align="center"><strong>CP1194</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2786</td>
<td align="center">2764</td>
<td align="center">2813</td>
<td align="center">932645</td>
<td align="center">915334</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1378</strong></td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">335</td>
<td align="center">2792</td>
<td align="center">2769</td>
<td align="center">2824</td>
<td align="center">876586</td>
<td align="center">848569</td>
</tr>
</tbody>
</table>
<table style="width:49%;" class="table">
<colgroup>
<col width="18%">
<col width="11%">
<col width="6%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">maxo</th>
<th align="center">sn</th>
<th align="center">sample</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>CP0045</strong></td>
<td align="center">16856</td>
<td align="center">23</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="center"><strong>CP0309</strong></td>
<td align="center">58736</td>
<td align="center">72</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP0587</strong></td>
<td align="center">6295</td>
<td align="center">13</td>
<td align="center">3</td>
</tr>
<tr class="even">
<td align="center"><strong>CP1194</strong></td>
<td align="center">35856</td>
<td align="center">66</td>
<td align="center">6</td>
</tr>
<tr class="odd">
<td align="center"><strong>CP1378</strong></td>
<td align="center">27200</td>
<td align="center">36</td>
<td align="center">7</td>
</tr>
</tbody>
</table>
<p>Finally we plot also the distribution of peak intensity per sample.
This allows to investigate whether systematic differences in peak
signals between samples are present.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract a list of per-sample peak intensities (in log2 scale)</span></span>
<span><span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"into"</span><span class="op">]</span><span class="op">)</span>,</span>
<span>              f <span class="op">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">ints</span>, varwidth <span class="op">=</span> <span class="cn">TRUE</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>,</span>
<span>        ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">log</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">~</span><span class="va">intensity</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Peak intensities"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-intensity-boxplot-1.png" alt="Peak intensity distribution per sample." width="960"><p class="caption">
Peak intensity distribution per sample.
</p>
</div>
<p>Note that in addition to the above described identification of
chromatographic peaks, it is also possible to <em>manually</em> define
and add chromatographic peaks with the <code>manualChromPeaks</code>
function (see <code><a href="../reference/manualChromPeaks.html">?manualChromPeaks</a></code> help page for more
information).</p>
</div>
<div class="section level2">
<h2 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h2>
<p>The time at which analytes elute in the chromatography can vary
between samples (and even compounds). Such a difference was already
observable in the extracted ion chromatogram plot shown as an example in
the previous section. The alignment step, also referred to as retention
time correction, aims at adjusting this by shifting signals along the
retention time axis to align the signals between different samples
within an experiment.</p>
<p>A plethora of alignment algorithms exist (see <span class="citation">[3]</span>), with some of them being implemented also
in <code>xcms</code>. The method to perform the alignment/retention time
correction in <code>xcms</code> is <code>adjustRtime</code> which uses
different alignment algorithms depending on the provided parameter
class.</p>
<p>In the example below we use the <em>obiwarp</em> method <span class="citation">[4]</span> to align the samples. We use a
<code>binSize = 0.6</code> which creates warping functions in mz bins of
0.6. Also here it is advisable to modify the settings for each
experiment and evaluate if retention time correction did align internal
controls or known compounds properly.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/adjustRtime-obiwarp.html">ObiwarpParam</a></span><span class="op">(</span>binSize <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><code>adjustRtime</code>, besides calculating adjusted retention
times for each spectrum, does also adjust the reported retention times
of the identified chromatographic peaks.</p>
<p>To extract the adjusted retention times we can use the
<code>adjustedRtime</code> method, or simply the <code>rtime</code>
method that, if present, returns by default adjusted retention times
from an <code>XCMSnExp</code> object.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract adjusted retention times</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">adjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0001 F1.S0002 F1.S0003 F1.S0004 F1.S0005 F1.S0006 </span></span>
<span><span class="co">## 2501.378 2502.958 2504.538 2506.118 2507.699 2509.280</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Or simply use the rtime method</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0001 F1.S0002 F1.S0003 F1.S0004 F1.S0005 F1.S0006 </span></span>
<span><span class="co">## 2501.378 2502.958 2504.538 2506.118 2507.699 2509.280</span></span></code></pre>
<p><em>Raw</em> retention times can be extracted from an
<code>XCMSnExp</code> containing aligned data with
<code>rtime(xdata, adjusted = FALSE)</code>.</p>
<p>To evaluate the impact of the alignment we plot the BPC on the
adjusted data. In addition we plot the differences of the adjusted- to
the raw retention times per sample using the
<code>plotAdjustedRtime</code> function. For a base peak chromatogram it
makes no sense to also extract identified chromatographic peaks from the
result object. We thus use parameter <code>include = "none"</code> in
the <code>chromatogram</code> call to not include chromatographic peaks
in the returned object. Note that alternatively it would also be
possible to simply avoid plotting them by setting
<code>peakType = "none"</code> in the <code>plot</code> call.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the base peak chromatograms.</span></span>
<span><span class="va">bpis_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, include <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4.5</span>, <span class="fl">4.2</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpis_adj</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">bpis_adj</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Plot also the difference of adjusted to raw retention time.</span></span>
<span><span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-obiwarp-plot-1.png" alt="Obiwarp aligned data. Base peak chromatogram after alignment (top) and difference between adjusted and raw retention times along the retention time axis (bottom)." width="960"><p class="caption">
Obiwarp aligned data. Base peak chromatogram after alignment (top) and
difference between adjusted and raw retention times along the retention
time axis (bottom).
</p>
</div>
<p>Too large differences between adjusted and raw retention times could
indicate poorly performing samples or alignment.</p>
<p><strong>Note</strong>: <code>XCMSnExp</code> objects hold the raw
along with the adjusted retention times and subsetting will in most
cases drop the adjusted retention times. Sometimes it might thus be
useful to <strong>replace</strong> the raw retention times with the
adjusted retention times. This can be done with the
<code>applyAdjustedRtime</code>.</p>
<p>At last we evaluate the impact of the alignment on the test peak.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Plot the raw data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_raw</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the chromatogram from the adjusted object</span></span>
<span><span class="va">chr_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, rt <span class="op">=</span> <span class="va">rtr</span>, mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_adj</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>, peakType <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-peak-groups-example-peak-1.png" alt="Example extracted ion chromatogram before (top) and after alignment (bottom)." width="960"><p class="caption">
Example extracted ion chromatogram before (top) and after alignment
(bottom).
</p>
</div>
<div class="section level3">
<h3 id="subset-based-alignment">Subset-based alignment<a class="anchor" aria-label="anchor" href="#subset-based-alignment"></a>
</h3>
<p>In some experiments it might be helpful to perform the alignment
based on only a subset of the samples, e.g. if QC samples were injected
at regular intervals or if the experiment contains blanks. Alignment
method in <code>xcms</code> allow to estimate retention time drifts on a
subset of samples (either all samples excluding blanks or QC samples
injected at regular intervals during a measurement run) and use these to
adjust the full data set.</p>
<p>Parameters <code>subset</code> (of the <code>PeakGroupsParam</code>
or <code>ObiwarpParam</code> object) can be used to define the subset of
samples on which the alignment of the full data set will be based
(e.g. <code>subset</code> being the index of QC samples), and parameter
<code>subsetAdjust</code> allows to specify the method by which the
<em>left-out</em> samples will be adjusted. There are currently two
options available:</p>
<ul>
<li><p><code>subsetAdjust = "previous"</code>: adjust the retention
times of a non-subset sample based on the alignment results of the
previous subset sample (e.g. a QC sample). If samples are e.g. in the
order <em>A1</em>, <em>B1</em>, <em>B2</em>, <em>A2</em>, <em>B3</em>,
<em>B4</em> with <em>A</em> representing QC samples and <em>B</em> study
samples, using <code>subset = c(1, 4)</code> and
<code>subsetAdjust = "previous"</code> would result in all <em>A</em>
samples to be aligned with each other and non-subset samples <em>B1</em>
and <em>B2</em> being adjusted based on the alignment result of subset
samples <em>A1</em> and <em>B3</em> and <em>B4</em> on those of
<em>A2</em>.</p></li>
<li><p><code>subsetAdjust = "average"</code>: adjust retention times of
non-subset samples based on an interpolation of the alignment results of
the previous and subsequent subset sample. In the example above,
<em>B1</em> would be adjusted based on the average of adjusted retention
times between subset (QC) samples <em>A1</em> and <em>A2</em>. Since
there is no subset sample after non-subset samples <em>B3</em> and
<em>B4</em> these will be adjusted based on the alignment results of
<em>A2</em> alone. Note that a weighted average is used to calculate the
adjusted retention time averages, which uses the inverse of the
difference of the index of the non-subset sample to the subset samples
as weights. Thus, if we have a setup like <em>A1</em>, <em>B1</em>,
<em>B2</em>, <em>A2</em> the adjusted retention times of <em>A1</em>
would get a larger weight than those of <em>A2</em> in the adjustment of
non-subset sample <em>B1</em> causing it’s adjusted retention times to
be closer to those of <em>A1</em> than to <em>A2</em>. See below for
examples.</p></li>
</ul>
<p>Both cases require a meaningful/correct ordering of the samples
within the object (e.g. ordering by injection index).</p>
<p>The examples below aim to illustrate the effect of these alignment
options. We assume that samples 1, 4 and 7 in the <em>faahKO</em> data
set are QC samples (sample pools). We thus want to perform the alignment
based on these samples and subsequently adjust the retention times of
the left-out samples (2, 3, 5, 6 and 8) based on interpolation of the
results from the neighboring <em>subset</em> (QC) samples. After initial
peak grouping we perform below the alignment with the <em>peak
groups</em> method passing the indices of the samples on which we want
the alignment to be based on with the <code>subset</code> argument and
specify <code>subsetAdjust = "average"</code> to adjust the study
samples based on interpolation of the alignment results from neighboring
subset/QC samples.</p>
<p>Note that for any subset-alignment all parameters such as
<code>minFraction</code> are relative to the <code>subset</code>, not
the full experiment!</p>
<p>To re-perform an alignment we can first remove previous alignment
results with the <code>dropAdjustedRtime</code> function.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">dropAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define the experimental layout</span></span>
<span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">&lt;-</span> <span class="st">"study"</span></span>
<span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"QC"</span></span></code></pre></div>
<p>We next have to perform an initial correspondence analysis because
the <em>peak groups</em> alignment method adjusts the retention time by
aligning previously identified <em>hook peaks</em> (chromatographic
peaks present in most/all samples; details about the algorithm used are
presented in the next section). We use here the default settings, but it
is strongly advised to adapt the parameters for each data set. The
definition of the sample groups (i.e. assignment of individual samples
to the sample groups in the experiment) is mandatory for the
<code>PeakDensityParam</code>. If there are no sample groups in the
experiment <code>sampleGroups</code> should be set to a single value for
each file (e.g. <code>rep(1, length(fileNames(xdata))</code>).</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Initial peak grouping. Use sample_type as grouping variable</span></span>
<span><span class="va">pdp_subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="va">xdata</span><span class="op">$</span><span class="va">sample_type</span>,</span>
<span>                             minFraction <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pdp_subs</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define subset-alignment options and perform the alignment</span></span>
<span><span class="va">pgp_subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime-peakGroups.html">PeakGroupsParam</a></span><span class="op">(</span>minFraction <span class="op">=</span> <span class="fl">0.85</span>,</span>
<span>                            subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">)</span>,</span>
<span>                            subsetAdjust <span class="op">=</span> <span class="st">"average"</span>, span <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pgp_subs</span><span class="op">)</span></span></code></pre></div>
<p>Below we plot the results of the alignment labeling the samples being
part of the <em>subset</em> in green and the others in grey. This nicely
shows how the interpolation of the <code>subsetAdjust = "average"</code>
works: retention times of sample 2 are adjusted based on those from
subset sample 1 and 4, giving however more weight to the closer subset
sample 1 which results in the adjusted retention times of 2 being more
similar to those of sample 1. Sample 3 on the other hand gets adjusted
giving more weight to the second subset sample (4).</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"#00000040"</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">clrs</span><span class="op">[</span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#00ce0080"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, aggregationFun <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span>,</span>
<span>     col <span class="op">=</span> <span class="va">clrs</span>, peakType <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span>, col <span class="op">=</span> <span class="va">clrs</span>, peakGroupsPch <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  peakGroupsCol <span class="op">=</span> <span class="st">"#00ce0040"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-subset-plot-2-1.png" alt="Subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey." width="960"><p class="caption">
Subset-alignment results with option average. Difference between
adjusted and raw retention times along the retention time axis. Samples
on which the alignment models were estimated are shown in green, study
samples in grey.
</p>
</div>
<p>Option <code>subsetAdjust = "previous"</code> adjusts the retention
times of a non-subset sample based on a single subset sample (the
previous), which results in most cases in the adjusted retention times
of the non-subset sample being highly similar to those of the subset
sample which was used for adjustment.</p>
</div>
</div>
<div class="section level2">
<h2 id="correspondence">Correspondence<a class="anchor" aria-label="anchor" href="#correspondence"></a>
</h2>
<p>The final step in the metabolomics preprocessing is the
correspondence that matches detected chromatographic peaks between
samples (and depending on the settings, also within samples if they are
adjacent). The method to perform the correspondence in <code>xcms</code>
is <code>groupChromPeaks</code>. We will use the <em>peak density</em>
method <span class="citation">[5]</span> to group chromatographic peaks.
The algorithm combines chromatographic peaks depending on the density of
peaks along the retention time axis within small slices along the mz
dimension. To illustrate this we plot below the chromatogram for an mz
slice with multiple chromatographic peaks within each sample. We use
below a value of 0.4 for the <code>minFraction</code> parameter hence
only chromatographic peaks present in at least 40% of the samples per
sample group are grouped into a feature. The sample group assignment is
specified with the <code>sampleGroups</code> argument.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define the mz slice.</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">305.05</span>, <span class="fl">305.15</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract and plot the chromatograms</span></span>
<span><span class="va">chr_mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span></span>
<span><span class="co">## Define the parameters for the peak density method</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">0.4</span>, bw <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotChromPeakDensity.html">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">chr_mzr</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, param <span class="op">=</span> <span class="va">pdp</span>,</span>
<span>                     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_mzr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                     peakCol <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_mzr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                     peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/correspondence-example-slice-1.png" alt="Example for peak density correspondence. Upper panel: chromatogram for an mz slice with multiple chromatographic peaks. lower panel: identified chromatographic peaks at their retention time (x-axis) and index within samples of the experiments (y-axis) for different values of the bw parameter. The black line represents the peak density estimate. Grouping of peaks (based on the provided settings) is indicated by grey rectangles." width="960"><p class="caption">
Example for peak density correspondence. Upper panel: chromatogram for
an mz slice with multiple chromatographic peaks. lower panel: identified
chromatographic peaks at their retention time (x-axis) and index within
samples of the experiments (y-axis) for different values of the bw
parameter. The black line represents the peak density estimate. Grouping
of peaks (based on the provided settings) is indicated by grey
rectangles.
</p>
</div>
<p>The upper panel in the plot above shows the extracted ion
chromatogram for each sample with the detected peaks highlighted. The
middle and lower plot shows the retention time for each detected peak
within the different samples. The black solid line represents the
density distribution of detected peaks along the retention times. Peaks
combined into <em>features</em> (peak groups) are indicated with grey
rectangles. This type of visualization is ideal to test correspondence
settings on example m/z slices before applying them to the full data
set.</p>
<p>Below we perform the correspondence analysis with the defined
settings on the full data set.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform the correspondence</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks-density.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">0.4</span>, bw <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<p>Results from the xcms-based preprocessing can be summarized into a
<code>SummarizedExperiment</code> object from the <em><a href="https://bioconductor.org/packages/3.17/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package with the <code>quantify</code> method. This object will contain
the feature abundances as the <em>assay matrix</em>, the feature
definition (their m/z, retention time and other metadata) as
<code>rowData</code> (i.e. row annotations) and the sample/phenotype
information as <code>colData</code> (i.e. column annotations). All the
processing history will be put into the object’s <code>metadata</code>.
This object can then be used for any further
(<code>xcms</code>-independent) processing and analysis.</p>
<p>Below we use <code>quantify</code> to generate the result object for
the present analysis. The parameters <code>value</code> and any other
additional parameters are passed along to the <code>featureValues</code>
method that is used internally to create the feature abundance
matrix.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify</a></span><span class="op">(</span><span class="va">xdata</span>, value <span class="op">=</span> <span class="st">"into"</span><span class="op">)</span></span></code></pre></div>
<p>Sample annotations can be accessed with the <code>colData</code>
method.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 3 columns</span></span>
<span><span class="co">##          sample_name sample_group sample_type</span></span>
<span><span class="co">##          &lt;character&gt;  &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## ko15.CDF        ko15           KO          QC</span></span>
<span><span class="co">## ko16.CDF        ko16           KO       study</span></span>
<span><span class="co">## ko21.CDF        ko21           KO       study</span></span>
<span><span class="co">## ko22.CDF        ko22           KO          QC</span></span>
<span><span class="co">## wt15.CDF        wt15           WT       study</span></span>
<span><span class="co">## wt16.CDF        wt16           WT       study</span></span>
<span><span class="co">## wt21.CDF        wt21           WT          QC</span></span>
<span><span class="co">## wt22.CDF        wt22           WT       study</span></span></code></pre>
<p>Feature annotations with <code>rowData</code>:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 225 rows and 11 columns</span></span>
<span><span class="co">##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FT001     200.1     200.1     200.1   2901.63   2880.73   2922.53         2</span></span>
<span><span class="co">## FT002     205.0     205.0     205.0   2789.39   2782.30   2795.36         8</span></span>
<span><span class="co">## FT003     206.0     206.0     206.0   2788.73   2780.73   2792.86         7</span></span>
<span><span class="co">## FT004     207.1     207.1     207.1   2718.12   2713.21   2726.70         7</span></span>
<span><span class="co">## FT005     219.1     219.1     219.1   2518.82   2517.40   2520.81         3</span></span>
<span><span class="co">## ...         ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## FT221    591.30     591.3     591.3   3005.03   2992.87   3006.05         5</span></span>
<span><span class="co">## FT222    592.15     592.1     592.3   3022.11   2981.91   3107.59         6</span></span>
<span><span class="co">## FT223    594.20     594.2     594.2   3418.16   3359.10   3427.90         3</span></span>
<span><span class="co">## FT224    595.25     595.2     595.3   3010.15   2992.87   3013.77         6</span></span>
<span><span class="co">## FT225    596.20     596.2     596.2   2997.91   2992.87   3002.95         2</span></span>
<span><span class="co">##              KO        WT         peakidx  ms_level</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt;          &lt;list&gt; &lt;integer&gt;</span></span>
<span><span class="co">## FT001         2         0         287,679         1</span></span>
<span><span class="co">## FT002         4         4  47,272,542,...         1</span></span>
<span><span class="co">## FT003         3         4  32,259,663,...         1</span></span>
<span><span class="co">## FT004         4         3  19,249,525,...         1</span></span>
<span><span class="co">## FT005         1         2   639, 788,1376         1</span></span>
<span><span class="co">## ...         ...       ...             ...       ...</span></span>
<span><span class="co">## FT221         2         3 349,684,880,...         1</span></span>
<span><span class="co">## FT222         1         3  86,861,862,...         1</span></span>
<span><span class="co">## FT223         1         2   604, 985,1543         1</span></span>
<span><span class="co">## FT224         2         3  67,353,876,...         1</span></span>
<span><span class="co">## FT225         0         2        866,1447         1</span></span></code></pre>
<p>The feature abundances can be accessed with the <code>assay</code>
method. Note also that a <code>SummarizedExperiment</code> supports
multiple such assay matrices.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001        NA  506848.9        NA  169955.6        NA        NA        NA</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7        NA  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9        NA</span></span>
<span><span class="co">## FT005        NA        NA        NA  107348.5  223951.8        NA        NA</span></span>
<span><span class="co">## FT006  286221.4        NA  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">##         wt22.CDF</span></span>
<span><span class="co">## FT001         NA</span></span>
<span><span class="co">## FT002 1354004.93</span></span>
<span><span class="co">## FT003  185399.47</span></span>
<span><span class="co">## FT004  221937.53</span></span>
<span><span class="co">## FT005   84772.92</span></span>
<span><span class="co">## FT006  271128.02</span></span></code></pre>
<p>In addition it is possible to extract the results from the
correspondence analysis individually using the
<code>featureDefinitions</code> and <code>featureValues</code> methods,
the former returning a <code>DataFrame</code> with the definition of the
features (i.e. the mz and rt ranges and, in column <code>peakidx</code>,
the index of the chromatographic peaks in the <code>chromPeaks</code>
matrix for each feature), the latter the feature abundances.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract the feature definitions</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">featureDefinitions</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 225 rows and 11 columns</span></span>
<span><span class="co">##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FT001     200.1     200.1     200.1   2901.63   2880.73   2922.53         2</span></span>
<span><span class="co">## FT002     205.0     205.0     205.0   2789.39   2782.30   2795.36         8</span></span>
<span><span class="co">## FT003     206.0     206.0     206.0   2788.73   2780.73   2792.86         7</span></span>
<span><span class="co">## FT004     207.1     207.1     207.1   2718.12   2713.21   2726.70         7</span></span>
<span><span class="co">## FT005     219.1     219.1     219.1   2518.82   2517.40   2520.81         3</span></span>
<span><span class="co">## ...         ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## FT221    591.30     591.3     591.3   3005.03   2992.87   3006.05         5</span></span>
<span><span class="co">## FT222    592.15     592.1     592.3   3022.11   2981.91   3107.59         6</span></span>
<span><span class="co">## FT223    594.20     594.2     594.2   3418.16   3359.10   3427.90         3</span></span>
<span><span class="co">## FT224    595.25     595.2     595.3   3010.15   2992.87   3013.77         6</span></span>
<span><span class="co">## FT225    596.20     596.2     596.2   2997.91   2992.87   3002.95         2</span></span>
<span><span class="co">##              KO        WT         peakidx  ms_level</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt;          &lt;list&gt; &lt;integer&gt;</span></span>
<span><span class="co">## FT001         2         0         287,679         1</span></span>
<span><span class="co">## FT002         4         4  47,272,542,...         1</span></span>
<span><span class="co">## FT003         3         4  32,259,663,...         1</span></span>
<span><span class="co">## FT004         4         3  19,249,525,...         1</span></span>
<span><span class="co">## FT005         1         2   639, 788,1376         1</span></span>
<span><span class="co">## ...         ...       ...             ...       ...</span></span>
<span><span class="co">## FT221         2         3 349,684,880,...         1</span></span>
<span><span class="co">## FT222         1         3  86,861,862,...         1</span></span>
<span><span class="co">## FT223         1         2   604, 985,1543         1</span></span>
<span><span class="co">## FT224         2         3  67,353,876,...         1</span></span>
<span><span class="co">## FT225         0         2        866,1447         1</span></span></code></pre>
<p>The <code>featureValues</code> method returns a <code>matrix</code>
with rows being features and columns samples. The content of this matrix
can be defined using the <code>value</code> argument. The default
<code>value = "into"</code> returns a matrix with the integrated signal
of the peaks corresponding to a feature in a sample. Any column name of
the <code>chromPeaks</code> matrix can be passed to the argument
<code>value</code>. Below we extract the integrated peak intensity per
feature/sample.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract the into column for each feature.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, value <span class="op">=</span> <span class="st">"into"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001        NA  506848.9        NA  169955.6        NA        NA        NA</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7        NA  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9        NA</span></span>
<span><span class="co">## FT005        NA        NA        NA  107348.5  223951.8        NA        NA</span></span>
<span><span class="co">## FT006  286221.4        NA  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">##         wt22.CDF</span></span>
<span><span class="co">## FT001         NA</span></span>
<span><span class="co">## FT002 1354004.93</span></span>
<span><span class="co">## FT003  185399.47</span></span>
<span><span class="co">## FT004  221937.53</span></span>
<span><span class="co">## FT005   84772.92</span></span>
<span><span class="co">## FT006  271128.02</span></span></code></pre>
<p>This feature matrix contains <code>NA</code> for samples in which no
chromatographic peak was detected in the feature’s m/z-rt region. While
in many cases there might indeed be no peak signal in the respective
region, it might also be that there is signal, but the peak detection
algorithm failed to detect a chromatographic peak (e.g. because the
signal was too low or too noisy). <code>xcms</code> provides the
<code>fillChromPeaks</code> method to <em>fill in</em> intensity data
for such missing values from the original files. The <em>filled in</em>
peaks are added to the <code>chromPeaks</code> matrix and indicated with
a value <code>TRUE</code> in the <code>"is_filled"</code> column of the
<code>chromPeakData</code> data frame. Below we perform such a gap
filling.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fillChromPeaks.html">fillChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/fillChromPeaks.html">ChromPeakAreaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF   ko16.CDF   ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001  159738.1  506848.88  113441.08  169955.6  216096.6  145509.7  230477.9</span></span>
<span><span class="co">## FT002 1924712.0 1757150.96 1383416.72 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.67  162897.19  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.66  343897.76  208002.8  364609.8  360908.9  223322.5</span></span>
<span><span class="co">## FT005  135978.5   25524.79   71530.84  107348.5  223951.8  134398.9  190203.8</span></span>
<span><span class="co">## FT006  286221.4  289908.23  164008.97  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">##         wt22.CDF</span></span>
<span><span class="co">## FT001  140551.30</span></span>
<span><span class="co">## FT002 1354004.93</span></span>
<span><span class="co">## FT003  185399.47</span></span>
<span><span class="co">## FT004  221937.53</span></span>
<span><span class="co">## FT005   84772.92</span></span>
<span><span class="co">## FT006  271128.02</span></span></code></pre>
<p>For features without detected peaks in a sample, the method extracts
all intensities in the mz-rt region of the feature, integrates the
signal and adds a <em>filled-in</em> peak to the <code>chromPeaks</code>
matrix. No peak is added if no signal is measured/available for the
mz-rt region of the feature. For these, even after filling in missing
peak data, a <code>NA</code> is reported in the
<code>featureValues</code> matrix.</p>
<p>Different options to define the mz-rt region of the features are
available. With the <code><a href="../reference/fillChromPeaks.html">ChromPeakAreaParam()</a></code> parameter object
used above, the feature area is defined using the m/z and rt ranges of
all of its (detected) chromatographic peaks: the lower m/z value of the
area is defined as the lower quartile (25% quantile) of the
<code>"mzmin"</code> values of all peaks of the feature, the upper m/z
value as the upper quartile (75% quantile) of the <code>"mzmax"</code>
values, the lower rt value as the lower quartile (25% quantile) of the
<code>"rtmin"</code> and the upper rt value as the upper quartile (75%
quantile) of the <code>"rtmax"</code> values. This ensures that the
signal is integrated from a feature-specific area.</p>
<p>Alternatively, it is possible to use the
<code>FillChromPeaksParam</code> parameter object in the
<code>fillChromPeaks</code> call, which resembles the approach of the
original (old) <code>xcms</code> implementation.</p>
<p>Below we compare the number of missing values before and after
filling in missing values. We can use the parameter <code>filled</code>
of the <code>featureValues</code> method to define whether or not
filled-in peak values should be returned too.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Missing values before filling in peaks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, filled <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>,</span>
<span>      FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ko15.CDF ko16.CDF ko21.CDF ko22.CDF wt15.CDF wt16.CDF wt21.CDF wt22.CDF </span></span>
<span><span class="co">##       71       69      114      102       54       89      104       75</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Missing values after filling in peaks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>,</span>
<span>      FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ko15.CDF ko16.CDF ko21.CDF ko22.CDF wt15.CDF wt16.CDF wt21.CDF wt22.CDF </span></span>
<span><span class="co">##        4        4        6        6        2        6        6        3</span></span></code></pre>
<p>Next we use the <code>featureSummary</code> function to get a general
per-feature summary that includes the number of samples in which a peak
was found or the number of samples in which more than one peak was
assigned to the feature. Specifying also sample groups breaks down these
summary statistics for each individual sample group.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/featureSummary.html">featureSummary</a></span><span class="op">(</span><span class="va">xdata</span>, group <span class="op">=</span> <span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       count  perc multi_count multi_perc       rsd KO_count KO_perc</span></span>
<span><span class="co">## FT001     2  25.0           0          0 0.7039537        2      50</span></span>
<span><span class="co">## FT002     8 100.0           0          0 0.1936518        4     100</span></span>
<span><span class="co">## FT003     7  87.5           0          0 0.1717662        3      75</span></span>
<span><span class="co">## FT004     7  87.5           0          0 0.2609145        4     100</span></span>
<span><span class="co">## FT005     3  37.5           0          0 0.5385767        1      25</span></span>
<span><span class="co">## FT006     7  87.5           0          0 0.3016973        3      75</span></span>
<span><span class="co">##       KO_multi_count KO_multi_perc    KO_rsd WT_count WT_perc WT_multi_count</span></span>
<span><span class="co">## FT001              0             0 0.7039537        0       0              0</span></span>
<span><span class="co">## FT002              0             0 0.2178920        4     100              0</span></span>
<span><span class="co">## FT003              0             0 0.2501505        4     100              0</span></span>
<span><span class="co">## FT004              0             0 0.2957873        3      75              0</span></span>
<span><span class="co">## FT005              0             0        NA        2      50              0</span></span>
<span><span class="co">## FT006              0             0 0.3765933        4     100              0</span></span>
<span><span class="co">##       WT_multi_perc    WT_rsd</span></span>
<span><span class="co">## FT001             0        NA</span></span>
<span><span class="co">## FT002             0 0.1918936</span></span>
<span><span class="co">## FT003             0 0.1327983</span></span>
<span><span class="co">## FT004             0 0.2575039</span></span>
<span><span class="co">## FT005             0 0.6375539</span></span>
<span><span class="co">## FT006             0 0.1641781</span></span></code></pre>
<p>We can add the feature value matrix with the filled-in data for
missing peaks also to our <code>SummarizedExperiment</code> object
<code>res</code> as an additional <em>assay</em>:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">raw_filled</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, filled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We have now two matrices (assays) available, the matrix with the
detected and the matrix with the detected and filled-in values, each can
be accessed by their name.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "raw"        "raw_filled"</span></span></code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001        NA  506848.9        NA  169955.6        NA        NA        NA</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7        NA  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9        NA</span></span>
<span><span class="co">## FT005        NA        NA        NA  107348.5  223951.8        NA        NA</span></span>
<span><span class="co">## FT006  286221.4        NA  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">##         wt22.CDF</span></span>
<span><span class="co">## FT001         NA</span></span>
<span><span class="co">## FT002 1354004.93</span></span>
<span><span class="co">## FT003  185399.47</span></span>
<span><span class="co">## FT004  221937.53</span></span>
<span><span class="co">## FT005   84772.92</span></span>
<span><span class="co">## FT006  271128.02</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF   ko16.CDF   ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001  159738.1  506848.88  113441.08  169955.6  216096.6  145509.7  230477.9</span></span>
<span><span class="co">## FT002 1924712.0 1757150.96 1383416.72 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.67  162897.19  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.66  343897.76  208002.8  364609.8  360908.9  223322.5</span></span>
<span><span class="co">## FT005  135978.5   25524.79   71530.84  107348.5  223951.8  134398.9  190203.8</span></span>
<span><span class="co">## FT006  286221.4  289908.23  164008.97  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">##         wt22.CDF</span></span>
<span><span class="co">## FT001  140551.30</span></span>
<span><span class="co">## FT002 1354004.93</span></span>
<span><span class="co">## FT003  185399.47</span></span>
<span><span class="co">## FT004  221937.53</span></span>
<span><span class="co">## FT005   84772.92</span></span>
<span><span class="co">## FT006  271128.02</span></span></code></pre>
<p>The performance of peak detection, alignment and correspondence
should always be evaluated by inspecting extracted ion chromatograms
e.g. of known compounds, internal standards or identified features in
general. The <code>featureChromatograms</code> function allows to
extract chromatograms for each feature present in
<code>featureDefinitions</code>. The returned
<code>MChromatograms</code> object contains an ion chromatogram for each
feature (each row containing the data for one feature) and sample (each
column representing containing data for one sample). Below we extract
the chromatograms for the first 4 features.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_chroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureChromatograms.html">featureChromatograms</a></span><span class="op">(</span><span class="va">xdata</span>, features <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_chroms</span></span></code></pre></div>
<pre><code><span><span class="co">## XChromatograms with 4 rows and 8 columns</span></span>
<span><span class="co">##                    1               2               3               4</span></span>
<span><span class="co">##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;</span></span>
<span><span class="co">## [1,]        peaks: 0        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [3,]        peaks: 1        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## [4,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">##                    5               6               7               8</span></span>
<span><span class="co">##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;</span></span>
<span><span class="co">## [1,]        peaks: 0        peaks: 0        peaks: 0        peaks: 0</span></span>
<span><span class="co">## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [3,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [4,]        peaks: 1        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## phenoData with 3 variables</span></span>
<span><span class="co">## featureData with 5 variables</span></span>
<span><span class="co">## - - - xcms preprocessing - - -</span></span>
<span><span class="co">## Chromatographic peak detection:</span></span>
<span><span class="co">##  method: centWave </span></span>
<span><span class="co">## Correspondence:</span></span>
<span><span class="co">##  method: chromatographic peak density </span></span>
<span><span class="co">##  4 feature(s) identified.</span></span></code></pre>
<p>And plot the extracted ion chromatograms. We again use the group
color for each identified peak to fill the area.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">feature_chroms</span>, col <span class="op">=</span> <span class="va">sample_colors</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">feature_chroms</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/feature-eic-1.png" alt="Extracted ion chromatograms for features 1 to 4." width="768"><p class="caption">
Extracted ion chromatograms for features 1 to 4.
</p>
</div>
<p>To access the EICs of the second feature we can simply subset the
<code>feature_chroms</code> object.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eic_2</span> <span class="op">&lt;-</span> <span class="va">feature_chroms</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">eic_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax    into    intb  maxo sn</span></span>
<span><span class="co">## CP0055 205   205   205 2790.427 2770.441 2813.596 1924712 1850331 84280 64</span></span>
<span><span class="co">## CP0310 205   205   205 2794.406 2772.731 2819.327 1757151 1711473 68384 69</span></span>
<span><span class="co">## CP0595 205   205   205 2795.358 2773.524 2820.417 1383417 1334570 47384 54</span></span>
<span><span class="co">## CP0736 205   205   205 2788.495 2768.076 2812.080 1180288 1126958 48336 32</span></span>
<span><span class="co">## CP0921 205   205   205 2782.296 2761.887 2805.849 2129885 2054677 93312 44</span></span>
<span><span class="co">## CP1197 205   205   205 2787.083 2766.688 2812.188 1634342 1566379 67984 53</span></span>
<span><span class="co">## CP1379 205   205   205 2790.294 2763.639 2821.635 1623589 1531573 49208 28</span></span>
<span><span class="co">## CP1542 205   205   205 2787.159 2766.777 2812.235 1354005 1299188 55712 35</span></span>
<span><span class="co">##        sample row column</span></span>
<span><span class="co">## CP0055      1   1      1</span></span>
<span><span class="co">## CP0310      2   1      2</span></span>
<span><span class="co">## CP0595      3   1      3</span></span>
<span><span class="co">## CP0736      4   1      4</span></span>
<span><span class="co">## CP0921      5   1      5</span></span>
<span><span class="co">## CP1197      6   1      6</span></span>
<span><span class="co">## CP1379      7   1      7</span></span>
<span><span class="co">## CP1542      8   1      8</span></span></code></pre>
<p>At last we perform a principal component analysis to evaluate the
grouping of the samples in this experiment. Note that we did not perform
any data normalization hence the grouping might (and will) also be
influenced by technical biases.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract the features and log2 transform them</span></span>
<span><span class="va">ft_ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_filled"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Perform the PCA omitting all features with an NA in any of the</span></span>
<span><span class="co">## samples. Also, the intensities are mean centered.</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">ft_ints</span><span class="op">)</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot the PCA</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">xdata</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span></span>
<span><span class="va">pcSummary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">21</span>, main <span class="op">=</span> <span class="st">""</span>,</span>
<span>     xlab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">pcSummary</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>                                   digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, <span class="st">" % variance"</span><span class="op">)</span>,</span>
<span>     ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">pcSummary</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>                                   digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, <span class="st">" % variance"</span><span class="op">)</span>,</span>
<span>     col <span class="op">=</span> <span class="st">"darkgrey"</span>, bg <span class="op">=</span> <span class="va">cols</span>, cex <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">xdata</span><span class="op">$</span><span class="va">sample_name</span>, col <span class="op">=</span> <span class="st">"darkgrey"</span>,</span>
<span>     pos <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/final-pca-1.png" alt="PCA for the faahKO data set, un-normalized intensities." width="768"><p class="caption">
PCA for the faahKO data set, un-normalized intensities.
</p>
</div>
<p>We can see the expected separation between the KO and WT samples on
PC2. On PC1 samples separate based on their ID, samples with an ID &lt;=
18 from samples with an ID &gt; 18. This separation might be caused by a
technical bias (e.g. measurements performed on different days/weeks) or
due to biological properties of the mice analyzed (sex, age, litter
mates etc).</p>
</div>
<div class="section level2">
<h2 id="further-data-processing-and-analysis">Further data processing and analysis<a class="anchor" aria-label="anchor" href="#further-data-processing-and-analysis"></a>
</h2>
<p>Normalizing features’ signal intensities is required, but at present
not (yet) supported in <code>xcms</code> (some methods might be added in
near future). It is advised to use the <code>SummarizedExperiment</code>
returned by the <code>quantify</code> method for any further data
processing, as this type of object stores feature definitions, sample
annotations as well as feature abundances in the same object. For the
identification of e.g. features with significant different
intensities/abundances it is suggested to use functionality provided in
other R packages, such as Bioconductor’s excellent <code>limma</code>
package. To enable support also for other packages that rely on the
<em>old</em> <code>xcmsSet</code> result object, it is possible to
coerce the new <code>XCMSnExp</code> object to an <code>xcmsSet</code>
object using <code>xset &lt;- as(x, "xcmsSet")</code>, with
<code>x</code> being an <code>XCMSnExp</code> object.</p>
</div>
<div class="section level2">
<h2 id="additional-details-and-notes">Additional details and notes<a class="anchor" aria-label="anchor" href="#additional-details-and-notes"></a>
</h2>
<p>For a detailed description of the new data objects and
changes/improvements compared to the original user interface see the
<em>new_functionality</em> vignette.</p>
<div class="section level3">
<h3 id="evaluating-the-process-history">Evaluating the process history<a class="anchor" aria-label="anchor" href="#evaluating-the-process-history"></a>
</h3>
<p><code>XCMSnExp</code> objects allow to capture all performed
pre-processing steps along with the used parameter class within the
<code>@processHistory</code> slot. Storing also the parameter class
ensures the highest possible degree of analysis documentation and in
future might enable to <em>replay</em> analyses or parts of it. The list
of all performed preprocessings can be extracted using the
<code>processHistory</code> method.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak detection </span></span>
<span><span class="co">##  date: Mon Jan 16 12:17:26 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: CentWaveParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak refinement </span></span>
<span><span class="co">##  date: Mon Jan 16 12:17:46 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: MergeNeighboringPeaksParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Mon Jan 16 12:18:01 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Retention time correction </span></span>
<span><span class="co">##  date: Mon Jan 16 12:18:02 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakGroupsParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Mon Jan 16 12:18:09 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Missing peak filling </span></span>
<span><span class="co">##  date: Mon Jan 16 12:18:11 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: ChromPeakAreaParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>It is also possible to extract specific processing steps by
specifying its type. Available <em>types</em> can be listed with the
<code>processHistoryTypes</code> function. Below we extract the
parameter class for the alignment/retention time adjustment step.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span><span class="op">(</span><span class="va">xdata</span>, type <span class="op">=</span> <span class="st">"Retention time correction"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ph</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Retention time correction </span></span>
<span><span class="co">##  date: Mon Jan 16 12:18:02 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakGroupsParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>And we can also extract the parameter class used in this
preprocessing step.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Access the parameter</span></span>
<span><span class="fu"><a href="../reference/ProcessHistory-class.html">processParam</a></span><span class="op">(</span><span class="va">ph</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class:  PeakGroupsParam </span></span>
<span><span class="co">##  Parameters:</span></span>
<span><span class="co">##  - minFraction: [1] 0.85</span></span>
<span><span class="co">##  - extraPeaks: [1] 1</span></span>
<span><span class="co">##  - smooth: [1] "loess"</span></span>
<span><span class="co">##  - span: [1] 0.4</span></span>
<span><span class="co">##  - family: [1] "gaussian"</span></span>
<span><span class="co">##  - peakGroupsMatrix:      ko15.CDF ko22.CDF wt21.CDF</span></span>
<span><span class="co">## FT12 2617.185 2623.444       NA</span></span>
<span><span class="co">## FT26 2679.783 2686.043       NA</span></span>
<span><span class="co">## FT25 2679.783 2686.043 2690.739</span></span>
<span><span class="co">## FT30 2676.653 2687.608 2695.434</span></span>
<span><span class="co">## FT31 2679.783 2687.608 2693.869</span></span>
<span><span class="co">## FT32 2679.783 2689.172 2692.304</span></span>
<span><span class="co">## FT42 2681.348 2689.172 2696.999</span></span>
<span><span class="co">## FT43 2678.218 2689.172 2700.129</span></span>
<span><span class="co">## FT45 2678.218 2689.172 2692.304</span></span>
<span><span class="co">## FT03 2711.082 2723.601       NA</span></span>
<span><span class="co">## FT07 2712.647 2720.472 2723.603</span></span>
<span><span class="co">## FT20 2784.635       NA       NA</span></span>
<span><span class="co">## FT13 2784.635 2790.894       NA</span></span>
<span><span class="co">## FT06 2784.635 2794.024       NA</span></span>
<span><span class="co">## FT01 2784.635 2789.330 2795.591</span></span>
<span><span class="co">## FT02 2783.070 2789.330 2794.026</span></span>
<span><span class="co">## FT18 2783.070 2790.894 2792.461</span></span>
<span><span class="co">## FT79 2784.635 2790.894 2792.461</span></span>
<span><span class="co">## FT09 2792.459 2801.849 2809.676</span></span>
<span><span class="co">## FT75 2903.571 2914.526 2922.352</span></span>
<span><span class="co">## FT14 2923.916 2923.916 2928.612</span></span>
<span><span class="co">## FT16 2923.916 2922.351 2928.612</span></span>
<span><span class="co">## FT15 2923.916 2927.045 2933.307</span></span>
<span><span class="co">## FT17 2995.903 2995.903 3002.165</span></span>
<span><span class="co">## FT80 2992.773 3008.423 3011.555</span></span>
<span><span class="co">## FT22 3019.378 3020.943 3031.899</span></span>
<span><span class="co">## FT04 3024.073 3014.683 3033.464</span></span>
<span><span class="co">## FT41 3005.293 3027.202 3028.769</span></span>
<span><span class="co">## FT40 3053.807 3069.456 3063.198</span></span>
<span><span class="co">## FT46 3064.761 3085.106 3089.802</span></span>
<span><span class="co">## FT51 3128.924 3146.139 3153.966</span></span>
<span><span class="co">## FT35 3194.652 3158.659 3163.355</span></span>
<span><span class="co">## FT60 3155.529 3171.178 3177.440</span></span>
<span><span class="co">## FT69 3166.483 3179.003 3204.044</span></span>
<span><span class="co">## FT70 3168.048       NA 3208.739</span></span>
<span><span class="co">## FT73 3185.263 3224.387 3225.953</span></span>
<span><span class="co">## FT74 3185.263 3227.516 3227.518</span></span>
<span><span class="co">## FT76 3189.958 3230.646 3233.778</span></span>
<span><span class="co">## FT77 3191.523 3230.646 3233.778</span></span>
<span><span class="co">## FT36 3258.816 3260.381       NA</span></span>
<span><span class="co">## FT08 3255.686 3260.381 3269.772</span></span>
<span><span class="co">## FT19 3276.030 3297.939 3304.201</span></span>
<span><span class="co">## FT47 3276.030 3297.939 3304.201</span></span>
<span><span class="co">## FT53 3318.284 3316.719 3333.935</span></span>
<span><span class="co">## FT52 3302.634 3324.544 3326.110</span></span>
<span><span class="co">## FT54 3297.939 3324.544 3330.805</span></span>
<span><span class="co">## FT39 3324.544 3327.674 3337.065</span></span>
<span><span class="co">## FT23 3299.504 3333.933 3332.370</span></span>
<span><span class="co">## FT71 3341.758 3357.408 3362.104</span></span>
<span><span class="co">## FT72 3341.758 3358.973 3362.104</span></span>
<span><span class="co">## FT37       NA 3355.843 3362.104</span></span>
<span><span class="co">## FT38 3341.758 3360.538 3362.104</span></span>
<span><span class="co">## FT10 3379.317 3398.096 3412.183</span></span>
<span><span class="co">## FT63       NA 3402.791 3410.618</span></span>
<span><span class="co">## FT21 3415.311 3421.571 3426.267</span></span>
<span><span class="co">## FT78 3374.622 3427.831 3427.832</span></span>
<span><span class="co">## FT55 3416.876 3429.395 3432.527</span></span>
<span><span class="co">## FT64 3377.752 3446.610 3446.612</span></span>
<span><span class="co">## FT65 3377.752 3448.175 3449.742</span></span>
<span><span class="co">## FT58 3384.012 3449.740 3449.742</span></span>
<span><span class="co">## FT49 3454.435 3470.084 3474.781</span></span>
<span><span class="co">## FT27 3490.429 3498.253 3498.255</span></span>
<span><span class="co">##  - subset: [1] 1 4 7</span></span>
<span><span class="co">##  - subsetAdjust: [1] "average"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="subsetting-and-filtering">Subsetting and filtering<a class="anchor" aria-label="anchor" href="#subsetting-and-filtering"></a>
</h3>
<p><code>XCMSnEx</code> objects can be subsetted/filtered using the
<code>[</code> method, or one of the many <code>filter*</code> methods.
All these methods aim to ensure that the data in the returned object is
consistent. This means for example that if the object is subsetted by
selecting specific spectra (by using the <code>[</code> method) all
identified chromatographic peaks are removed. Correspondence results
(i.e. identified features) are removed if the object is subsetted to
contain only data from selected files (using the <code>filterFile</code>
method). This is because the correspondence results depend on the files
on which the analysis was performed - running a correspondence on a
subset of the files would lead to different results. Note that with
<code>keepFeatures = TRUE</code> it would be possible to overwrite this
and keep also correspondence results for the specified files.</p>
<p>As an exception, it is possible to force keeping adjusted retention
times in the subsetted object setting the <code>keepAdjustedRtime</code>
argument to <code>TRUE</code> in any of the subsetting methods.</p>
<p>Below we subset our results object the data for the files 2 and
4.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Do we have identified chromatographic peaks?</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasChromPeaks</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Peak detection is performed separately on each file, thus the
subsetted object contains all identified chromatographic peaks from the
two files. However, we used a retention time adjustment (alignment) that
was based on available features. All features have however been removed
and also the adjusted retention times (since the alignment based on
features that were identified on chromatographic peaks on all
files).</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Do we still have features?</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasFeatures</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Do we still have adjusted retention times?</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>We can however use the <code>keepAdjustedRtime</code> argument to
force keeping the adjusted retention times, <code>keepFeatures</code>
would even keep correspondence results.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata</span>, keepAdjustedRtime <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>The <code>filterRt</code> method can be used to subset the object to
spectra within a certain retention time range.</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">xdata</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">3500</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3000.186 3499.956</span></span></code></pre>
<p>Filtering by retention time does not change/affect adjusted retention
times (also, if adjusted retention times are present, the filtering is
performed <strong>on</strong> the adjusted retention times).</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Also, we have all identified chromatographic peaks within the
specified retention time range:</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasChromPeaks</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span><span class="op">[</span>, <span class="st">"rt"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3000.712 3499.399</span></span></code></pre>
<p>The most natural way to subset any object in R is with
<code>[</code>. Using <code>[</code> on an <code>XCMSnExp</code> object
subsets it keeping only the selected spectra. The index <code>i</code>
used in <code>[</code> has thus to be an integer between 1 and the total
number of spectra (across all files). Below we subset <code>xdata</code>
using both <code>[</code> and <code>filterFile</code> to keep all
spectra from one file.</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract all data from the 3rd file.</span></span>
<span><span class="va">one_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSnExp-class.html" class="external-link">filterFile</a></span><span class="op">(</span><span class="va">xdata</span>, file <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">one_file_2</span> <span class="op">&lt;-</span> <span class="va">xdata</span><span class="op">[</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Is the content the same?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">one_file</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">one_file_2</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Attributes: &lt; Component \"rt\": Mean relative difference: 0.0007411294 &gt;"</span></span></code></pre>
<p>While the spectra-content is the same in both objects,
<code>one_file</code> contains also the identified chromatographic peaks
while <code>one_file_2</code> does not. Thus, in most situations
subsetting using one of the filter functions is preferred over the use
of <code>[</code>.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subsetting with filterFile preserves chromatographic peaks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">one_file</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax      into      intb   maxo sn</span></span>
<span><span class="co">## CP0555 231   231   231 2509.494 2503.234 2515.754  184167.0  165087.9  15818 11</span></span>
<span><span class="co">## CP0556 334   334   334 2514.189 2504.799 2520.449  107281.7  104197.1   7976 14</span></span>
<span><span class="co">## CP0557 337   337   337 2515.754 2503.234 2553.313 4381724.8 4040088.7 123448 21</span></span>
<span><span class="co">## CP0558 333   333   333 2515.754 2503.234 2545.488  934386.2  877441.9  27944 15</span></span>
<span><span class="co">## CP0559 316   316   316 2520.449 2503.234 2556.443  863947.7  832855.2  24496 29</span></span>
<span><span class="co">## CP0560 332   332   332 2518.884 2503.234 2553.313 4835730.4 4644120.0 131520 41</span></span>
<span><span class="co">##        sample</span></span>
<span><span class="co">## CP0555      1</span></span>
<span><span class="co">## CP0556      1</span></span>
<span><span class="co">## CP0557      1</span></span>
<span><span class="co">## CP0558      1</span></span>
<span><span class="co">## CP0559      1</span></span>
<span><span class="co">## CP0560      1</span></span></code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subsetting with [ not</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">one_file_2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<p>Note however that also <code>[</code> does support the
<code>keepAdjustedRtime</code> argument. Below we subset the object to
spectra 20:30.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subs</span> <span class="op">&lt;-</span> <span class="va">xdata</span><span class="op">[</span><span class="fl">20</span><span class="op">:</span><span class="fl">30</span>, keepAdjustedRtime <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">hasAdjustedRtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Access adjusted retention times:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">subs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0020 F1.S0021 F1.S0022 F1.S0023 F1.S0024 F1.S0025 F1.S0026 F1.S0027 </span></span>
<span><span class="co">## 2539.356 2540.921 2542.486 2544.051 2545.616 2547.181 2548.746 2550.311 </span></span>
<span><span class="co">## F1.S0028 F1.S0029 F1.S0030 </span></span>
<span><span class="co">## 2551.876 2553.441 2555.006</span></span></code></pre>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Access raw retention times:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">subs</span>, adjusted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## F1.S0020 F1.S0021 F1.S0022 F1.S0023 F1.S0024 F1.S0025 F1.S0026 F1.S0027 </span></span>
<span><span class="co">## 2531.112 2532.677 2534.242 2535.807 2537.372 2538.937 2540.502 2542.067 </span></span>
<span><span class="co">## F1.S0028 F1.S0029 F1.S0030 </span></span>
<span><span class="co">## 2543.632 2545.197 2546.762</span></span></code></pre>
<p>As with <code>MSnExp</code> and <code>OnDiskMSnExp</code> objects,
<code>[[</code> can be used to extract a single spectrum object from an
<code>XCMSnExp</code> object. The retention time of the spectrum
corresponds to the adjusted retention time if present.</p>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract a single spectrum</span></span>
<span><span class="va">xdata</span><span class="op">[[</span><span class="fl">14</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "Spectrum1"</span></span>
<span><span class="co">##  Retention time: 42:10 </span></span>
<span><span class="co">##  MSn level: 1 </span></span>
<span><span class="co">##  Total ion count: 445 </span></span>
<span><span class="co">##  Polarity: -1</span></span></code></pre>
<p>At last we can also use the <code>split</code> method that allows to
split an <code>XCMSnExp</code> based on a provided factor
<code>f</code>. Below we split <code>xdata</code> per file. Using
<code>keepAdjustedRtime = TRUE</code> ensures that adjusted retention
times are not removed.</p>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">xdata</span>, f <span class="op">=</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span>, keepAdjustedRtime <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">x_list</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   1   2   3   4   5   6   7   8 </span></span>
<span><span class="co">## 639 639 639 639 639 639 639 639</span></span></code></pre>
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x_list</span>, <span class="va">hasAdjustedRtime</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`1`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`3`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`4`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`5`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`6`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`7`</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`8`</span></span>
<span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="parallel-processing">Parallel processing<a class="anchor" aria-label="anchor" href="#parallel-processing"></a>
</h3>
<p>Most methods in <code>xcms</code> support parallel processing.
Parallel processing is handled and configured by the
<code>BiocParallel</code> Bioconductor package and can be globally
defined for an R session.</p>
<p>Unix-based systems (Linux, macOS) support
<code>multicore</code>-based parallel processing. To configure it
globally we <code>register</code> the parameter class. Note also that
<code>bpstart</code> is used below to initialize the parallel
processes.</p>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Windows supports only socket-based parallel processing:</p>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code>multicore</code>-based parallel processing might be
buggy or failing on macOS. If so, the <code>DoparParam</code> could be
used instead (requiring the <code>foreach</code> package).</p>
<p>For other options and details see the vignettes from the
<code>BiocParallel</code> package.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Saghatelian04" class="csl-entry">
1. Saghatelian A, Trauger SA, Want EJ, Hawkins EG, Siuzdak G, Cravatt
BF: <strong><a href="http://dx.doi.org/10.1021/bi0480335" class="external-link">Assignment of
endogenous substrates to enzymes by global metabolite
profiling</a></strong>. <em>Biochemistry</em> 2004,
<strong>43</strong>:14332–9.
</div>
<div id="ref-Tautenhahn:2008fx" class="csl-entry">
2. Tautenhahn R, Böttcher C, Neumann S: <strong><span class="nocase">Highly sensitive feature detection for high resolution
LC/MS.</span></strong> <em>BMC Bioinformatics</em> 2008,
<strong>9</strong>:504.
</div>
<div id="ref-Smith:2013gr" class="csl-entry">
3. Smith R, Ventura D, Prince JT: <strong><span class="nocase">LC-MS
alignment in theory and practice: a comprehensive algorithmic
review.</span></strong> <em>Briefings in bioinformatics</em> 2013,
<strong>16</strong>:bbt080–117.
</div>
<div id="ref-Prince:2006jj" class="csl-entry">
4. Prince JT, Marcotte EM: <strong><span class="nocase">Chromatographic
alignment of ESI-LC-MS proteomics data sets by ordered bijective
interpolated warping.</span></strong> <em>Analytical chemistry</em>
2006, <strong>78</strong>:6140–6152.
</div>
<div id="ref-Smith:2006ic" class="csl-entry">
5. Smith CA, Want EJ, O’Maille G, Abagyan R, Siuzdak G: <strong><span class="nocase">XCMS: processing mass spectrometry data for metabolite
profiling using nonlinear peak alignment, matching, and
identification.</span></strong> <em>Analytical chemistry</em> 2006,
<strong>78</strong>:779–787.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Steffen Neumann.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
