<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="xcms">
<title>LC-MS data pre-processing and analysis with xcms • xcms</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LC-MS data pre-processing and analysis with xcms">
<meta property="og:description" content="xcms">
<meta property="og:image" content="https://sneumann.github.io/xcms/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xcms</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.99.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/xcms.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/LC-MS-feature-grouping.html">Compounding (grouping) of LC-MS features</a>
    <a class="dropdown-item" href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    <a class="dropdown-item" href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sneumann/xcms/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>LC-MS data pre-processing and analysis with xcms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/HEAD/vignettes/xcms.Rmd" class="external-link"><code>vignettes/xcms.Rmd</code></a></small>
      <div class="d-none name"><code>xcms.Rmd</code></div>
    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em><br><strong>Authors</strong>: Johannes Rainer<br><strong>Modified</strong>: 2023-10-18 12:21:55.86337<br><strong>Compiled</strong>: Wed Oct 18 13:41:38 2023</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em> package
provides the functionality to perform the pre-processing of LC-MS, GC-MS
or LC-MS/MS data in which raw signals from mzML, mzXML or CDF files are
processed into <em>feature</em> abundances. This pre-processing includes
chromatographic peak detection, sample alignment and correspondence
analysis.</p>
<p>The first version of the package was already published in 2006 <span class="citation">[1]</span> and has since been updated and modernized in
several rounds to better integrate it with other R-based packages for
the analysis of untargeted metabolomics data. This includes version 3 of
<em>xcms</em> that used the <em><a href="https://bioconductor.org/packages/3.18/MSnbase" class="external-link">MSnbase</a></em>
package for MS data representation <span class="citation">[2]</span>.
The most recent update (<em>xcms</em> version 4) enables in addition
pre-processing of MS data represented by the modern <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
packages which provides an even better integration with the <a href="https://rformassspectrometry.org" class="external-link">RforMassSpectrometry</a> R
package ecosystem simplifying e.g. also compound annotation <span class="citation">[3]</span>.</p>
<p>This document describes data import, exploration and pre-processing
of a simple test LC-MS data set with the <em>xcms</em> package version
&gt;= 4. The same functions can be applied to the older
<em>MSnbase</em>-based workflows (xcms version 3). Additional documents
and tutorials covering also other topics of untargeted metabolomics
analysis are listed at the end of this document.</p>
</div>
<div class="section level2">
<h2 id="pre-processing-of-lc-ms-data">Pre-processing of LC-MS data<a class="anchor" aria-label="anchor" href="#pre-processing-of-lc-ms-data"></a>
</h2>
<div class="section level3">
<h3 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h3>
<p><em>xcms</em> supports analysis of any LC-MS(/MS) data that can be
imported with the <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package. Such data will typically be provided in (AIA/ANDI) NetCDF,
mzXML and mzML format but can, through dedicated extensions to the
<em>Spectra</em> package, also be imported from other sources, e.g. also
directly from raw data files in manufacturer’s formats.</p>
<p>For demonstration purpose we will analyze in this document a small
subset of the data from <span class="citation">[4]</span> in which the
metabolic consequences of the knock-out of the fatty acid amide
hydrolase (FAAH) gene in mice was investigated. The raw data files (in
NetCDF format) are provided through the <em><a href="https://bioconductor.org/packages/3.18/faahKO" class="external-link">faahKO</a></em>
data package. The data set consists of samples from the spinal cords of
6 knock-out and 6 wild-type mice. Each file contains data in centroid
mode acquired in positive ion polarity from 200-600 m/z and 2500-4500
seconds. To speed-up processing of this vignette we will restrict the
analysis to only 8 files.</p>
<p>Below we load all required packages, locate the raw CDF files within
the <em>faahKO</em> package and build a <em>phenodata</em>
<code>data.frame</code> describing the experimental setup. Generally,
such data frames should contain all relevant experimental variables and
sample descriptions (including also the names of the raw data files) and
will be imported into R using either the <code>read.table</code>
function (if the file is in <em>csv</em> or tabulator delimited text
file format) or also using functions from the <em>readxl</em> R package
if it is in Excel file format.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/xcms" class="external-link">xcms</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.doi.org/10.1021/bi0480335" class="external-link">faahKO</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Get the full path to the CDF files</span></span>
<span><span class="va">cdfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"cdf"</span>, package <span class="op">=</span> <span class="st">"faahKO"</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">8</span>, <span class="fl">11</span>, <span class="fl">12</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Create a phenodata data.frame</span></span>
<span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">cdfs</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">".CDF"</span>,</span>
<span>                                   replacement <span class="op">=</span> <span class="st">""</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                 sample_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"KO"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                 stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We next load our data using the <code>readMsExperiment</code>
function from the <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/readMsExperiment.html" class="external-link">readMsExperiment</a></span><span class="op">(</span>spectraFiles <span class="op">=</span> <span class="va">cdfs</span>, sampleData <span class="op">=</span> <span class="va">pd</span><span class="op">)</span></span>
<span><span class="va">data</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class MsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (10224) </span></span>
<span><span class="co">##  Experiment data: 8 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 8 sample(s) to 10224 element(s).</span></span></code></pre>
<p>The MS spectra data from our experiment is now available as a
<code>Spectra</code> object within <code>data</code>. Note that this
<code>MsExperiment</code> container could in addition to spectra data
also contain other types of data or also references to other files. See
the vignette from the <em><a href="https://bioconductor.org/packages/3.18/MsExperiment" class="external-link">MsExperiment</a></em>
for more details. Also, when loading data from mzML, mzXML or CDF files,
by default only general spectra data is loaded into memory while the
actual <em>peaks data</em>, i.e. the m/z and intensity values are only
retrieved on-the-fly from the raw files when needed (this is similar to
the <em>MSnbase</em> <em>on-disk</em> mode described in <span class="citation">[2]</span>). This guarantees a low memory footprint
hence allowing to analyze also large experiments without the need of
high performance computing environments. Note that also different
alternative <em>backends</em> (and hence data representations) could be
used for the <code>Spectra</code> object within <code>data</code> with
eventually even lower memory footprint, or higher performance. See the
package vignette from the <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package or the <a href="https://jorainer.github.io/SpectraTutorials" class="external-link">SpectraTutorials</a>
tutorial for more details on <code>Spectra</code> backends and how to
change between them.</p>
</div>
<div class="section level3">
<h3 id="initial-data-inspection">Initial data inspection<a class="anchor" aria-label="anchor" href="#initial-data-inspection"></a>
</h3>
<p>The <code>MsExperiment</code> object is a simple and flexible
container for MS experiments. The <em>raw</em> MS data is stored as a
<code>Spectra</code> object that can be accessed through the
<code>spectra</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 10224 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##         msLevel     rtime scanIndex</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1             1   2501.38         1</span></span>
<span><span class="co">## 2             1   2502.94         2</span></span>
<span><span class="co">## 3             1   2504.51         3</span></span>
<span><span class="co">## 4             1   2506.07         4</span></span>
<span><span class="co">## 5             1   2507.64         5</span></span>
<span><span class="co">## ...         ...       ...       ...</span></span>
<span><span class="co">## 10220         1   4493.56      1274</span></span>
<span><span class="co">## 10221         1   4495.13      1275</span></span>
<span><span class="co">## 10222         1   4496.69      1276</span></span>
<span><span class="co">## 10223         1   4498.26      1277</span></span>
<span><span class="co">## 10224         1   4499.82      1278</span></span>
<span><span class="co">##  ... 33 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## ko15.CDF</span></span>
<span><span class="co">## ko16.CDF</span></span>
<span><span class="co">## ko21.CDF</span></span>
<span><span class="co">##  ... 5 more files</span></span></code></pre>
<p>All spectra are organized <em>sequentially</em> (i.e., not by file)
but the <code>fromFile</code> function can be used to get for each
spectrum the information to which of the data files it belongs. Below we
simply count the number of spectra per file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    1    2    3    4    5    6    7    8 </span></span>
<span><span class="co">## 1278 1278 1278 1278 1278 1278 1278 1278</span></span></code></pre>
<p>Information on samples can be retrieved through the
<code>sampleData</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 3 columns</span></span>
<span><span class="co">##   sample_name sample_group spectraOrigin</span></span>
<span><span class="co">##   &lt;character&gt;  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">## 1        ko15           KO /__w/_temp...</span></span>
<span><span class="co">## 2        ko16           KO /__w/_temp...</span></span>
<span><span class="co">## 3        ko21           KO /__w/_temp...</span></span>
<span><span class="co">## 4        ko22           KO /__w/_temp...</span></span>
<span><span class="co">## 5        wt15           WT /__w/_temp...</span></span>
<span><span class="co">## 6        wt16           WT /__w/_temp...</span></span>
<span><span class="co">## 7        wt21           WT /__w/_temp...</span></span>
<span><span class="co">## 8        wt22           WT /__w/_temp...</span></span></code></pre>
<p>Each row in this <code>DataFrame</code> represents one sample (input
file). Using <code>[</code> it is possible to subset a
<code>MsExperiment</code> object <strong>by sample</strong>. Below we
subset the <code>data</code> to the 3rd sample (file) and access its
spectra and sample data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_3</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data_3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 1278 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##        msLevel     rtime scanIndex</span></span>
<span><span class="co">##      &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1            1   2501.38         1</span></span>
<span><span class="co">## 2            1   2502.94         2</span></span>
<span><span class="co">## 3            1   2504.51         3</span></span>
<span><span class="co">## 4            1   2506.07         4</span></span>
<span><span class="co">## 5            1   2507.64         5</span></span>
<span><span class="co">## ...        ...       ...       ...</span></span>
<span><span class="co">## 1274         1   4493.56      1274</span></span>
<span><span class="co">## 1275         1   4495.13      1275</span></span>
<span><span class="co">## 1276         1   4496.69      1276</span></span>
<span><span class="co">## 1277         1   4498.26      1277</span></span>
<span><span class="co">## 1278         1   4499.82      1278</span></span>
<span><span class="co">##  ... 33 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## ko21.CDF</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data_3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 1 row and 3 columns</span></span>
<span><span class="co">##   sample_name sample_group spectraOrigin</span></span>
<span><span class="co">##   &lt;character&gt;  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">## 1        ko21           KO /__w/_temp...</span></span></code></pre>
<p>As a first evaluation of the data we below plot the base peak
chromatogram (BPC) for each file in our experiment. We use the
<code>chromatogram</code> method and set the <code>aggregationFun</code>
to <code>"max"</code> to return for each spectrum the maximal intensity
and hence create the BPC from the raw data. To create a total ion
chromatogram we could set <code>aggregationFun</code> to
<code>"sum"</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the base peak chromatograms. This reads data from the files.</span></span>
<span><span class="va">bpis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="co">## Define colors for the two groups</span></span>
<span><span class="va">group_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"Set1"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"60"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">group_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot all chromatograms.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpis</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-bpc-1.png" alt="Base peak chromatogram." width="1152"><p class="caption">
Base peak chromatogram.
</p>
</div>
<p>The <code>chromatogram</code> method returned a
<code>MChromatograms</code> object that organizes individual
<code>Chromatogram</code> objects (which in fact contain the
chromatographic data) in a two-dimensional array: columns represent
samples and rows (optionally) m/z and/or retention time ranges. Below we
extract the chromatogram of the first sample and access its retention
time and intensity values.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bpi_1</span> <span class="op">&lt;-</span> <span class="va">bpis</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">bpi_1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2501.378 2502.943 2504.508 2506.073 2507.638 2509.203</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">bpi_1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 43888 43960 43392 42632 42200 42288</span></span></code></pre>
<p>From the BPC above it seems that after around 4200 seconds no signal
is measured anymore. Thus, we filter below the full data set to a
retention time range from 2550 to 4250 seconds using the
<code>filterRt</code> function. Note that at present this will only
subset the spectra within the <code>MsExperiment</code>. Subsequently we
re-create also the BPC.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">data</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2550</span>, <span class="fl">4250</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Filter spectra</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## creating the BPC on the subsetted data</span></span>
<span><span class="va">bpis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
<p>We next create boxplots representing the distribution of the total
ion currents per data file. Such plots can be very useful to spot
potentially problematic MS runs. To extract this information, we use the
<code>tic</code> function on the <code>Spectra</code> object within
<code>data</code> and split the values by file using
<code>fromFile</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the total ion current by file</span></span>
<span><span class="va">tc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span>f <span class="op">=</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/Spectrum-class.html" class="external-link">fromFile</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">tc</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"intensity"</span>, main <span class="op">=</span> <span class="st">"Total ion current"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-tic-boxplot-1.png" alt="Distribution of total ion currents per file." width="768"><p class="caption">
Distribution of total ion currents per file.
</p>
</div>
<p>In addition, we can also cluster the samples based on similarity of
their base peak chromatograms. Samples would thus be grouped based on
similarity of their LC runs. For that we need however to <em>bin</em>
the data along the retention time axis, since retention times will
generally differ between samples. Below we use the <code>bin</code>
function on the BPC to bin intensities into 2 second wide retention time
bins. The clustering is then performed using complete linkage
hierarchical clustering on the pairwise correlations of the binned base
peak chromatograms.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Bin the BPC</span></span>
<span><span class="va">bpis_bin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">bin</a></span><span class="op">(</span><span class="va">bpis</span>, binSize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate correlation on the log2 transformed base peak intensities</span></span>
<span><span class="va">cormat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bpis_bin</span>, <span class="va">intensity</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cormat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cormat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bpis_bin</span><span class="op">$</span><span class="va">sample_name</span></span>
<span></span>
<span><span class="co">## Define which phenodata columns should be highlighted in the plot</span></span>
<span><span class="va">ann</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">bpis_bin</span><span class="op">$</span><span class="va">sample_group</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ann</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bpis_bin</span><span class="op">$</span><span class="va">sample_name</span></span>
<span></span>
<span><span class="co">## Perform the cluster analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">cormat</span>, annotation <span class="op">=</span> <span class="va">ann</span>,</span>
<span>         annotation_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">group_colors</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/data-inspection-bpc-heatmap-1.png" alt="Grouping of samples based on similarity of their base peak chromatogram." width="672"><p class="caption">
Grouping of samples based on similarity of their base peak chromatogram.
</p>
</div>
<p>The samples cluster in a pairwise manner, with the KO and WT samples
for the same sample index having the most similar BPC.</p>
</div>
<div class="section level3">
<h3 id="chromatographic-peak-detection">Chromatographic peak detection<a class="anchor" aria-label="anchor" href="#chromatographic-peak-detection"></a>
</h3>
<p>Chromatographic peak detection aims at identifying all signal in each
sample created from ions of the same originating compound species.
Chromatographic peak detection can be performed in <em>xcms</em> with
the <code>findChromPeaks</code> function and a <em>parameter</em> object
which defines and configures the algorithm that should be used (see
<code><a href="../reference/findChromPeaks.html">?findChromPeaks</a></code> for a list of supported algorithms). Before
running any peak detection it is however strongly suggested to first
visually inspect the extracted ion chromatogram of e.g. internal
standards or compounds known to be present in the samples in order to
evaluate and adapt the settings of the peak detection algorithm since
the default settings will not be appropriate for most LC-MS setups.</p>
<p>Below we extract the EIC for one compound using the
<code>chromatogram</code> function by specifying in addition the m/z and
retention time range where we would expect the signal for that
compound.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define the rt and m/z range of the peak area</span></span>
<span><span class="va">rtr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2700</span>, <span class="fl">2900</span><span class="op">)</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">334.9</span>, <span class="fl">335.1</span><span class="op">)</span></span>
<span><span class="co">## extract the chromatogram</span></span>
<span><span class="va">chr_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">data</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_raw</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-plot-eic-1.png" alt="Extracted ion chromatogram for one peak." width="768"><p class="caption">
Extracted ion chromatogram for one peak.
</p>
</div>
<p>Note that <code>Chromatogram</code> objects extracted by the
<code>chromatogram</code> method contain an <code>NA</code> value if in
a certain scan (i.e. in a spectrum for a specific retention time) no
signal was measured in the respective m/z range. This is reflected by
the lines not being drawn as continuous lines in the plot above.</p>
<p>The peak above has thus a width of about 50 seconds. We can use this
information to define the <code>peakwidth</code> parameter of the
<em>centWave</em> peak detection method <span class="citation">[5]</span> that we will use for chromatographic peak
detection on our data. The <code>peakwidth</code> parameter allows to
define the expected lower and upper width (in retention time dimension)
of chromatographic peaks. For our data we set it thus to
<code>peakwidth = c(20, 80)</code>. The second important parameter for
<em>centWave</em> is <code>ppm</code> which defines the expected maximum
deviation of m/z values of the centroids that should be included into
one chromatographic peak (note that this is <strong>not</strong> the
precision of m/z values provided by the MS instrument manufacturer).</p>
<p>For the <code>ppm</code> parameter we extract the full MS data
(intensity, retention time and m/z values) corresponding to the above
peak. To this end we first filter the raw object by retention time, then
by m/z and finally plot the object with <code>type = "XIC"</code> to
produce the plot below. We use the <em>pipe</em> (<code>|&gt;</code>)
operator to better illustrate the corresponding workflow.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span>rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMz</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"XIC"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-detection-plot-ms-data-1.png" alt="Visualization of the raw MS data for one peak. For each plot: upper panel: chromatogram plotting the intensity values against the retention time, lower panel m/z against retention time plot. The individual data points are colored according to the intensity." width="1344"><p class="caption">
Visualization of the raw MS data for one peak. For each plot: upper
panel: chromatogram plotting the intensity values against the retention
time, lower panel m/z against retention time plot. The individual data
points are colored according to the intensity.
</p>
</div>
<p>In the present data there is actually no variation in the m/z values.
Usually the m/z values of the individual centroids (lower panel) in the
plots above would scatter around the <em>real</em> m/z value of the
compound (in an intensity dependent manner).</p>
<p>The first step of the <em>centWave</em> algorithm defines <em>regions
of interest</em> (ROI) that are subsequently screened for the presence
of chromatographic peaks. These ROIs are defined based on the difference
of m/z values of centroids from consecutive scans (spectra). In detail,
centroids from consecutive scans are included into a ROI if the
difference between their m/z and the mean m/z of the ROI is smaller than
the user defined <code>ppm</code> parameter. A reasonable choice for the
<code>ppm</code> could thus be the maximal m/z difference of data points
from neighboring scans/spectra that are part of a chromatographic peak
for an internal standard of known compound. It is suggested to inspect
the ranges of m/z values for several compounds (either internal
standards or compounds known to be present in the sample) and define the
<code>ppm</code> parameter for <em>centWave</em> according to these. See
also this <a href="https://jorainer.github.io/metabolomics2018" class="external-link">tutorial</a> for
additional information and examples on choosing and testing peak
detection settings.</p>
<p>Chromatographic peak detection can also be performed on extracted ion
chromatograms, which helps testing and tuning peak detection settings.
Note however that peak detection on EICs does not involve the first step
of <em>centWave</em> described above and will thus <strong>not</strong>
consider the <code>ppm</code> parameter. Also, since less data is
available to the algorithms, background signal estimation is performed
differently and different settings for <code>snthresh</code> will need
to be used (generally a lower <code>snthresh</code> will be used for
EICs since the estimated background signal tends to be higher for data
subsets than for the full data). Below we perform the peak detection
with the <code>findChromPeaks</code> function on the EIC generated
above. The submitted <em>parameter</em> object defines which algorithm
will be used and allows to define the settings for this algorithm. We
use a <code>CentWaveParam</code> parameter object to use and configure
the <em>centWave</em> algorithm with default settings, except for
<code>snthresh</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xchr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks.html">findChromPeaks</a></span><span class="op">(</span><span class="va">chr_raw</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>snthresh <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can access the identified chromatographic peaks with the
<code>chromPeaks</code> function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             rt    rtmin    rtmax        into        intb  maxo   sn row column</span></span>
<span><span class="co">##  [1,] 2781.505 2761.160 2809.674  412134.255  355516.374 16856   13   1      1</span></span>
<span><span class="co">##  [2,] 2786.199 2764.290 2812.803 1496244.206 1391821.332 58736   20   1      2</span></span>
<span><span class="co">##  [3,] 2734.556 2714.211 2765.855   21579.367   18449.428   899    4   1      3</span></span>
<span><span class="co">##  [4,] 2797.154 2775.245 2815.933  159058.782  150289.314  6295   12   1      3</span></span>
<span><span class="co">##  [5,] 2784.635 2761.160 2808.109   54947.545   37923.532  2715    2   1      4</span></span>
<span><span class="co">##  [6,] 2859.752 2845.668 2878.532   13895.212   13874.868   905  904   1      4</span></span>
<span><span class="co">##  [7,] 2897.311 2891.051 2898.876    5457.155    5450.895   995  994   1      4</span></span>
<span><span class="co">##  [8,] 2819.064 2808.109 2834.713   19456.914   19438.134  1347 1576   1      4</span></span>
<span><span class="co">##  [9,] 2789.329 2762.725 2808.109  174473.311   91114.975  8325    3   1      5</span></span>
<span><span class="co">## [10,] 2786.199 2764.290 2812.803  932645.211  569236.246 35856    2   1      6</span></span>
<span><span class="co">## [11,] 2792.461 2768.987 2823.760  876585.527  511324.134 27200    2   1      7</span></span>
<span><span class="co">## [12,] 2789.329 2773.680 2806.544   89582.569   73871.386  5427    6   1      8</span></span></code></pre>
<p>Parallel to the <code>chromPeaks</code> matrix there is also a
<code>chromPeakData</code> data frame that allows to add arbitrary
annotations to each chromatographic peak, such as e.g. the MS level in
which the peak was detected:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 12 rows and 4 columns</span></span>
<span><span class="co">##      ms_level is_filled       row    column</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;logical&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           1     FALSE         1         1</span></span>
<span><span class="co">## 2           1     FALSE         1         2</span></span>
<span><span class="co">## 3           1     FALSE         1         3</span></span>
<span><span class="co">## 4           1     FALSE         1         3</span></span>
<span><span class="co">## 5           1     FALSE         1         4</span></span>
<span><span class="co">## ...       ...       ...       ...       ...</span></span>
<span><span class="co">## 8           1     FALSE         1         4</span></span>
<span><span class="co">## 9           1     FALSE         1         5</span></span>
<span><span class="co">## 10          1     FALSE         1         6</span></span>
<span><span class="co">## 11          1     FALSE         1         7</span></span>
<span><span class="co">## 12          1     FALSE         1         8</span></span></code></pre>
<p>Below we plot the EIC along with all identified chromatographic peaks
using the <code>plot</code> function on the result object from above.
Additional parameters <code>peakCol</code> and <code>peakBg</code> allow
to define a foreground and background (fill) color for each identified
chromatographic peak in the provided result object (i.e., we need to
define one color for each row of <code>chromPeaks(xchr)</code> - column
<code>"column"</code> (or <code>"sample"</code> if present) in that peak
matrix specifies the sample in which the peak was identified).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define a color for each sample</span></span>
<span><span class="va">sample_colors</span> <span class="op">&lt;-</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">xchr</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span></span>
<span><span class="co">## Define the background color for each chromatographic peak</span></span>
<span><span class="va">bg</span> <span class="op">&lt;-</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xchr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"column"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## Parameter `col` defines the color of each sample/line, `peakBg` of each</span></span>
<span><span class="co">## chromatographic peak.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">xchr</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, peakBg <span class="op">=</span> <span class="va">bg</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-eic-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. Peak area of identified chromatographic peaks are highlighted in the sample group color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. Peak area of identified chromatographic
peaks are highlighted in the sample group color.
</p>
</div>
<p>If we are happy with the settings we can use them for the peak
detection on the full data set (i.e. on the <code>MsExperiment</code>
object with the data for the whole experiment). Note that below we set
the argument <code>prefilter</code> to <code>c(6, 5000)</code> and
<code>noise</code> to <code>5000</code> to reduce the run time of this
vignette. With this setting we consider only ROIs with at least 6
centroids with an intensity larger than 5000 for the <em>centWave</em>
chromatographic peak detection.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">80</span><span class="op">)</span>, noise <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                     prefilter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">5000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks.html">findChromPeaks</a></span><span class="op">(</span><span class="va">data</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span></code></pre></div>
<p>The results of <code>findChromPeaks</code> on a
<code>MsExperiment</code> object are returned as an
<code>XcmsExperiment</code> object. This object extends
<code>MsExperiment</code> directly (hence providing the same access to
all raw data) and contains all <em>xcms</em> pre-processing results.
Note also that additional rounds of chromatographic peak detections
could be performed and their results being added to existing peak
detection results by additional calls to <code>findChromPeaks</code> on
the result object and using parameter <code>add = TRUE</code>.</p>
<p>The <code>chromPeaks</code> function can also here be used to access
the results from the chromatographic peak detection. Below we show the
first 6 identified chromatographic peaks.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           mz mzmin mzmax       rt    rtmin    rtmax     into     intb  maxo sn</span></span>
<span><span class="co">## CP0001 594.0 594.0 594.0 2601.535 2581.191 2637.529 161042.2 146073.3  7850 11</span></span>
<span><span class="co">## CP0002 577.0 577.0 577.0 2604.665 2581.191 2626.574 136105.2 128067.9  6215 11</span></span>
<span><span class="co">## CP0003 307.0 307.0 307.0 2618.750 2592.145 2645.354 284782.4 264907.0 16872 20</span></span>
<span><span class="co">## CP0004 302.0 302.0 302.0 2617.185 2595.275 2640.659 687146.6 669778.1 30552 43</span></span>
<span><span class="co">## CP0005 370.1 370.1 370.1 2673.523 2643.789 2700.127 449284.6 417225.3 25672 17</span></span>
<span><span class="co">## CP0006 427.0 427.0 427.0 2675.088 2643.789 2684.478 283334.7 263943.2 11025 13</span></span>
<span><span class="co">##        sample</span></span>
<span><span class="co">## CP0001      1</span></span>
<span><span class="co">## CP0002      1</span></span>
<span><span class="co">## CP0003      1</span></span>
<span><span class="co">## CP0004      1</span></span>
<span><span class="co">## CP0005      1</span></span>
<span><span class="co">## CP0006      1</span></span></code></pre>
<p>Columns of this <code>chromPeaks</code> matrix might differ depending
on the used peak detection algorithm. Columns that all algorithms have
to provide are: <code>"mz"</code>, <code>"mzmin"</code>,
<code>"mzmax"</code>, <code>"rt"</code>, <code>"rtmin"</code> and
<code>"rtmax"</code> that define the m/z and retention time range of the
chromatographic peak (i.e. all mass peaks within that area are used to
integrate the peak signal) as well as the m/z and retention time of the
peak apex. Other mandatory columns are <code>"into"</code> and
<code>"maxo"</code> with the absolute integrated peak signal and the
maximum peak intensity. Finally, <code>"sample"</code> provides the
index of the sample in which the peak was identified.</p>
<p>Additional annotations for each individual peak can be extracted with
the <code>chromPeakData</code> function. This data frame could also be
used to add/store arbitrary annotations for each detected peak (that
don’t necessarily need to be numeric).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2908 rows and 2 columns</span></span>
<span><span class="co">##         ms_level is_filled</span></span>
<span><span class="co">##        &lt;integer&gt; &lt;logical&gt;</span></span>
<span><span class="co">## CP0001         1     FALSE</span></span>
<span><span class="co">## CP0002         1     FALSE</span></span>
<span><span class="co">## CP0003         1     FALSE</span></span>
<span><span class="co">## CP0004         1     FALSE</span></span>
<span><span class="co">## CP0005         1     FALSE</span></span>
<span><span class="co">## ...          ...       ...</span></span>
<span><span class="co">## CP2904         1     FALSE</span></span>
<span><span class="co">## CP2905         1     FALSE</span></span>
<span><span class="co">## CP2906         1     FALSE</span></span>
<span><span class="co">## CP2907         1     FALSE</span></span>
<span><span class="co">## CP2908         1     FALSE</span></span></code></pre>
<p>Peak detection will not always work perfectly for all types of peak
shapes present in the data set leading to peak detection artifacts, such
as (partially or completely) overlapping peaks or artificially split
peaks (common issues especially for <em>centWave</em>). <em>xcms</em>
provides the <code>refineChromPeaks</code> function that can be called
on peak detection results in order to <em>refine</em> (or clean) peak
detection results by either removing identified peaks not passing a
certain criteria or by merging artificially split or partially or
completely overlapping chromatographic peaks. Different algorithms are
available that can again be configured with their respective parameter
objects: <code>CleanPeaksParam</code> and
<code>FilterIntensityParam</code> allow to remove peaks with their
retention time range or intensity being below a specified threshold,
respectively. With <code>MergeNeighboringPeaksParam</code> it is
possible to merge chromatographic peaks and hence remove many of the
above mentioned (<em>centWave</em>) peak detection artifacts. See also
<code><a href="../reference/refineChromPeaks.html">?refineChromPeaks</a></code> for more information and help on the
different methods.</p>
<p>Below we post-process the peak detection results merging peaks that
overlap in a 4 second window per file and for which the signal between
them is lower than 75% of the smaller peak’s maximal intensity. See the
<code><a href="../reference/refineChromPeaks.html">?MergeNeighboringPeaksParam</a></code> help page for a detailed
description of the settings and the approach.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mpp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks.html">MergeNeighboringPeaksParam</a></span><span class="op">(</span>expandRt <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">xdata_pp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks.html">refineChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, <span class="va">mpp</span><span class="op">)</span></span></code></pre></div>
<p>An example for a merged peak is given below.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzr_1</span> <span class="op">&lt;-</span> <span class="fl">305.1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="va">chr_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata_pp</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The splitted peak was merged into one." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing,
right: after refinement. The splitted peak was merged into one.
</p>
</div>
<p><em>centWave</em> thus detected originally 3 chromatographic peaks in
the m/z slice (left panel in the plot above) and peak refinement with
<code>MergeNeighboringPeaksParam</code> and the specified settings
merged the first two peaks into a single one (right panel in the figure
above). Other close peaks, with a lower intensity between them, were
however not merged (see below).</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mzr_1</span> <span class="op">&lt;-</span> <span class="fl">496.2</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.01</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">chr_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="va">chr_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata_pp</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, mz <span class="op">=</span> <span class="va">mzr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms_files/figure-html/peak-postprocessing-not-merged-1.png" alt="Result from the peak refinement step. Left: data before processing, right: after refinement. The peaks were not merged." width="700"><p class="caption">
Result from the peak refinement step. Left: data before processing,
right: after refinement. The peaks were not merged.
</p>
</div>
<p>It is also possible to perform the peak refinement on extracted ion
chromatograms. This could again be used to test and fine-tune the
settings for the parameter. To illustrate this we perform below a peak
refinement on the extracted ion chromatogram <code>chr_1</code> reducing
the <code>minProp</code> parameter to force joining the two peaks.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/refineChromPeaks.html">refineChromPeaks</a></span><span class="op">(</span><span class="va">chr_1</span>, <span class="fu"><a href="../reference/refineChromPeaks.html">MergeNeighboringPeaksParam</a></span><span class="op">(</span>minProp <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz  mzmin  mzmax       rt    rtmin    rtmax     into intb    maxo  sn</span></span>
<span><span class="co">## CPM1 496.2 496.19 496.21 3384.012 3294.809 3412.181 45940118   NA 1128960 177</span></span>
<span><span class="co">##      sample row column</span></span>
<span><span class="co">## CPM1      1   1      1</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms_files/figure-html/peak-postprocessing-chr-1.png" width="700"></p>
<p>Before proceeding we next replace the <code>xdata</code> object with
the results from the peak refinement step.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="va">xdata_pp</span></span></code></pre></div>
<p>Below we use the data from the <code>chromPeaks</code> matrix to
calculate per-file summaries of the peak detection results, such as the
number of peaks per file as well as the distribution of the retention
time widths.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>peak_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">z</span><span class="op">[</span>, <span class="st">"rtmax"</span><span class="op">]</span> <span class="op">-</span> <span class="va">z</span><span class="op">[</span>, <span class="st">"rtmin"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="cn">T</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split.data.frame</a></span><span class="op">(</span>f <span class="op">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>FUN <span class="op">=</span> <span class="va">summary_fun</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span>what <span class="op">=</span> <span class="va">rbind</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="cn">T</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">fileNames</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html" class="external-link">pandoc.table</a></span><span class="op">(</span></span>
<span>    <span class="cn">T</span>,</span>
<span>    caption <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Summary statistics on identified chromatographic"</span>,</span>
<span>                     <span class="st">" peaks. Shown are number of identified peaks per"</span>,</span>
<span>                     <span class="st">" sample and widths/duration of chromatographic "</span>,</span>
<span>                     <span class="st">"peaks."</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Summary statistics on identified chromatographic peaks. Shown
are number of identified peaks per sample and widths/duration of
chromatographic peaks.</caption>
<colgroup>
<col width="20%">
<col width="17%">
<col width="10%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">peak_count</th>
<th align="center">rt.0%</th>
<th align="center">rt.25%</th>
<th align="center">rt.50%</th>
<th align="center">rt.75%</th>
<th align="center">rt.100%</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>ko15.CDF</strong></td>
<td align="center">397</td>
<td align="center">10.95</td>
<td align="center">34.43</td>
<td align="center">42.25</td>
<td align="center">53.21</td>
<td align="center">319.2</td>
</tr>
<tr class="even">
<td align="center"><strong>ko16.CDF</strong></td>
<td align="center">520</td>
<td align="center">10.95</td>
<td align="center">32.86</td>
<td align="center">42.25</td>
<td align="center">53.21</td>
<td align="center">151.8</td>
</tr>
<tr class="odd">
<td align="center"><strong>ko21.CDF</strong></td>
<td align="center">207</td>
<td align="center">10.95</td>
<td align="center">42.25</td>
<td align="center">50.08</td>
<td align="center">65.73</td>
<td align="center">164.3</td>
</tr>
<tr class="even">
<td align="center"><strong>ko22.CDF</strong></td>
<td align="center">233</td>
<td align="center">10.95</td>
<td align="center">37.56</td>
<td align="center">46.95</td>
<td align="center">59.47</td>
<td align="center">147.1</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt15.CDF</strong></td>
<td align="center">403</td>
<td align="center">10.95</td>
<td align="center">32.86</td>
<td align="center">42.25</td>
<td align="center">53.21</td>
<td align="center">161.2</td>
</tr>
<tr class="even">
<td align="center"><strong>wt16.CDF</strong></td>
<td align="center">361</td>
<td align="center">10.95</td>
<td align="center">35.99</td>
<td align="center">45.38</td>
<td align="center">57.9</td>
<td align="center">162.8</td>
</tr>
<tr class="odd">
<td align="center"><strong>wt21.CDF</strong></td>
<td align="center">227</td>
<td align="center">10.95</td>
<td align="center">35.21</td>
<td align="center">48.51</td>
<td align="center">64.16</td>
<td align="center">172.1</td>
</tr>
<tr class="even">
<td align="center"><strong>wt22.CDF</strong></td>
<td align="center">328</td>
<td align="center">10.95</td>
<td align="center">35.99</td>
<td align="center">45.38</td>
<td align="center">57.9</td>
<td align="center">228.5</td>
</tr>
</tbody>
</table>
<p>While by default <code>chromPeaks</code> will return all identified
chromatographic peaks in a result object it is also possible to extract
only chromatographic peaks for a specified m/z and/or rt range:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">334.9</span>, <span class="fl">335.1</span><span class="op">)</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2700</span>, <span class="fl">2900</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax      into      intb  maxo sn</span></span>
<span><span class="co">## CP0038 335   335   335 2781.505 2761.160 2809.674  412134.3  383167.4 16856 23</span></span>
<span><span class="co">## CP0494 335   335   335 2786.199 2764.290 2812.803 1496244.2 1461187.3 58736 72</span></span>
<span><span class="co">## CP1025 335   335   335 2797.154 2775.245 2815.933  159058.8  149229.6  6295 13</span></span>
<span><span class="co">## CP1964 335   335   335 2786.199 2764.290 2812.803  932645.2  915333.8 35856 66</span></span>
<span><span class="co">## CP2349 335   335   335 2792.461 2768.987 2823.760  876585.5  848569.1 27200 36</span></span>
<span><span class="co">##        sample</span></span>
<span><span class="co">## CP0038      1</span></span>
<span><span class="co">## CP0494      2</span></span>
<span><span class="co">## CP1025      3</span></span>
<span><span class="co">## CP1964      6</span></span>
<span><span class="co">## CP2349      7</span></span></code></pre>
<p>We can also plot the location of the identified chromatographic peaks
in the m/z - retention time space for one file using the
<code>plotChromPeaks</code> function. Below we plot this information for
the third sample.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, file <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peaks-plot-1.png" alt="Identified chromatographic peaks in the m/z by retention time space for one sample." width="768"><p class="caption">
Identified chromatographic peaks in the m/z by retention time space for
one sample.
</p>
</div>
<p>As a general overview of the peak detection results we can in
addition visualize the number of identified chromatographic peaks per
file along the retention time axis. Parameter <code>binSize</code>
allows to define the width of the bins in rt dimension in which peaks
should be counted. This number of chromatographic peaks within each bin
is then shown color-coded in the resulting plot.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotChromPeaks.html">plotChromPeakImage</a></span><span class="op">(</span><span class="va">xdata</span>, binSize <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-image-1.png" alt="Frequency of identified chromatographic peaks along the retention time axis. The frequency is color coded with higher frequency being represented by yellow-white. Each line shows the peak frequency for one file." width="960"><p class="caption">
Frequency of identified chromatographic peaks along the retention time
axis. The frequency is color coded with higher frequency being
represented by yellow-white. Each line shows the peak frequency for one
file.
</p>
</div>
<p>Note that extracting ion chromatorams from an <em>xcms</em> result
object will in addition to the chromatographic data also extract any
identified chromatographic peaks within that region. This can thus also
be used to validate and verify that the used peak detection settings
identified e.g. peaks for known compounds or internal standards
properly. Below we extract the ion chromatogram for the m/z - rt region
above and access the detected peaks in that region using the
<code>chromPeaks</code> function.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr_ex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="va">mzr</span>, rt <span class="op">=</span> <span class="va">rtr</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax      into      intb  maxo sn</span></span>
<span><span class="co">## CP0038 335   335   335 2781.505 2761.160 2809.674  412134.3  383167.4 16856 23</span></span>
<span><span class="co">## CP0494 335   335   335 2786.199 2764.290 2812.803 1496244.2 1461187.3 58736 72</span></span>
<span><span class="co">## CP1025 335   335   335 2797.154 2775.245 2815.933  159058.8  149229.6  6295 13</span></span>
<span><span class="co">## CP1964 335   335   335 2786.199 2764.290 2812.803  932645.2  915333.8 35856 66</span></span>
<span><span class="co">## CP2349 335   335   335 2792.461 2768.987 2823.760  876585.5  848569.1 27200 36</span></span>
<span><span class="co">##        sample row column</span></span>
<span><span class="co">## CP0038      1   1      1</span></span>
<span><span class="co">## CP0494      2   1      2</span></span>
<span><span class="co">## CP1025      3   1      3</span></span>
<span><span class="co">## CP1964      6   1      6</span></span>
<span><span class="co">## CP2349      7   1      7</span></span></code></pre>
<p>We can also plot this extracted ion chromatogram which will also
visualize all identified chromatographic peaks in that region.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_colors</span> <span class="op">&lt;-</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_ex</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_ex</span>, col <span class="op">=</span> <span class="va">group_colors</span><span class="op">[</span><span class="va">chr_raw</span><span class="op">$</span><span class="va">sample_group</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-polygon-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The signal area of identified chromatographic peaks are filled with a color." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. The signal area of identified
chromatographic peaks are filled with a color.
</p>
</div>
<p>Chromatographic peaks can be visualized also in other ways: by
setting <code>peakType = "rectangle"</code> the they are indicated with
a rectangle instead of the default highlighting option
(<code>peakType = "polygon"</code>) used above. As a third alternative
it would also possible to just indicate the apex position for each
identified chromatographic peak with a single point
(<code>peakType = "point"</code>). Below we plot the data again using
<code>peakType = "rectangle"</code>.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_ex</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, peakType <span class="op">=</span> <span class="st">"rectangle"</span>,</span>
<span>     peakCol <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_ex</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-highlight-chrom-peaks-plot-1.png" alt="Signal for an example peak. Red and blue colors represent KO and wild type samples, respectively. The rectangles indicate the identified chromatographic peaks per sample." width="960"><p class="caption">
Signal for an example peak. Red and blue colors represent KO and wild
type samples, respectively. The rectangles indicate the identified
chromatographic peaks per sample.
</p>
</div>
<p>Finally we plot also the distribution of peak intensity per sample.
This allows to investigate whether systematic differences in peak
signals between samples are present.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract a list of per-sample peak intensities (in log2 scale)</span></span>
<span><span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"into"</span><span class="op">]</span><span class="op">)</span>,</span>
<span>              f <span class="op">=</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">ints</span>, varwidth <span class="op">=</span> <span class="cn">TRUE</span>, col <span class="op">=</span> <span class="va">sample_colors</span>,</span>
<span>        ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">log</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">~</span><span class="va">intensity</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Peak intensities"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span>nx <span class="op">=</span> <span class="cn">NA</span>, ny <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/peak-detection-chrom-peak-intensity-boxplot-1.png" alt="Peak intensity distribution per sample." width="960"><p class="caption">
Peak intensity distribution per sample.
</p>
</div>
<p>Over and above the signal of the identified chromatographic peaks is
comparable across samples, with the exception of samples 3, 4 and, to
some degree, also 7 that show slightly higher average intensities, but
also a lower number of detected peaks (indicated by the smaller width of
the boxes).</p>
<p>Note that in addition to the above described identification of
chromatographic peaks, it is also possible to <em>manually</em> define
and add chromatographic peaks with the <code>manualChromPeaks</code>
function (see <code><a href="../reference/manualChromPeaks.html">?manualChromPeaks</a></code> help page for more
information).</p>
</div>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>The time at which analytes elute in the chromatography can vary
between samples (and even compounds). Such differences were already
visible in the BPC and even the extracted ion chromatogram plot in the
previous section. The alignment step, also referred to as retention time
correction, aims to adjust these differences by shifting signals along
the retention time axis and aligning them between different samples
within an experiment.</p>
<p>A plethora of alignment algorithms exist (see <span class="citation">[6]</span>), with some of them being also implemented
in <em>xcms</em>. Alignment of LC-MS data can be performed in
<em>xcms</em> using the <code>adjustRtime</code> method and an
algorithm-specific parameter class (see <code><a href="../reference/adjustRtime.html">?adjustRtime</a></code> for an
overview of available methods in <em>xcms</em>).</p>
<p>In the example below we use the <em>obiwarp</em> method <span class="citation">[7]</span> to align the samples. We use a
<code>binSize = 0.6</code> which creates warping functions in m/z bins
of 0.6. Also here it is advisable to modify and adapt the settings for
each experiment.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/adjustRtime.html">ObiwarpParam</a></span><span class="op">(</span>binSize <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that <code>adjustRtime</code>, besides calculating adjusted
retention times for each spectrum, adjusts also the retention times of
the identified chromatographic peaks in the <em>xcms</em> result object.
Adjusted retention times of individual spectra can be extracted from the
result object using either the <code>adjustedRtime</code> function or
using <code>rtime</code> with parameter <code>adjusted = TRUE</code>
(the default):</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract adjusted retention times</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">adjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2551.457 2553.089 2554.720 2556.352 2557.983 2559.615</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Or simply use the rtime method</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2551.457 2553.089 2554.720 2556.352 2557.983 2559.615</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get raw (unadjusted) retention times</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">xdata</span>, adjusted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2551.457 2553.022 2554.586 2556.151 2557.716 2559.281</span></span></code></pre>
<p>To evaluate the impact of the alignment we plot the BPC on the
adjusted data. In addition we plot also the differences between the
adjusted and the raw retention times per sample using the
<code>plotAdjustedRtime</code> function. To disable the automatic
extraction of all identified chromatographic peaks by the
<code>chromatogram</code> function (which would not make much sense for
a BPC) we use <code>chromPeaks = "none"</code> below.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the base peak chromatograms.</span></span>
<span><span class="va">bpis_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4.5</span>, <span class="fl">4.2</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpis</span>, col <span class="op">=</span> <span class="va">sample_colors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bpis_adj</span>, col <span class="op">=</span> <span class="va">sample_colors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Plot also the difference of adjusted to raw retention time.</span></span>
<span><span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span>, col <span class="op">=</span> <span class="va">sample_colors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-obiwarp-plot-1.png" alt="Obiwarp aligned data. Base peak chromatogram before (top) and after alignment (middle) and difference between adjusted and raw retention times along the retention time axis (bottom)." width="1152"><p class="caption">
Obiwarp aligned data. Base peak chromatogram before (top) and after
alignment (middle) and difference between adjusted and raw retention
times along the retention time axis (bottom).
</p>
</div>
<p>Too large differences between adjusted and raw retention times could
indicate poorly performing samples or alignment.</p>
<p>At last we evaluate also the impact of the alignment on the test
peak.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Plot the raw data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_raw</span>, col <span class="op">=</span> <span class="va">sample_colors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## Extract the chromatogram from the adjusted object</span></span>
<span><span class="va">chr_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, rt <span class="op">=</span> <span class="va">rtr</span>, mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr_adj</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, peakType <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-peak-groups-example-peak-1.png" alt="Example extracted ion chromatogram before (top) and after alignment (bottom)." width="960"><p class="caption">
Example extracted ion chromatogram before (top) and after alignment
(bottom).
</p>
</div>
<p><strong>Note</strong>: <em>xcms</em> result objects
(<code>XcmsExperiment</code> as well as <code>XCMSnExp</code>) contain
both the raw and the adjusted retention times for all spectra and subset
operation will in many cases drop adjusted retention times. Thus it
might sometimes be useful to immediately <strong>replace</strong> the
raw retention times in the data using the
<code>applyAdjustedRtime</code> function.</p>
<div class="section level4">
<h4 id="subset-based-alignment">Subset-based alignment<a class="anchor" aria-label="anchor" href="#subset-based-alignment"></a>
</h4>
<p>For some experiments it might be better to perform the alignment
based on only a subset of the available samples, e.g. if pooled QC
samples were injected at regular intervals or if the experiment contains
blanks. All alignment methods in <em>xcms</em> support such a
subset-based alignment in which retention time shifts are estimated on
only a specified subset of samples followed by an alignment of the whole
data set based on the aligned subset.</p>
<p>The subset of samples for such an alignment can be specified with the
parameter <code>subset</code> of the <code>PeakGroupsParam</code> or
<code>ObiwarpParam</code> object. Parameter <code>subsetAdjust</code>
allows to specify the <em>method</em> by which the <em>left-out</em>
samples will be adjusted. There are currently two options available:</p>
<ul>
<li><p><code>subsetAdjust = "previous"</code>: adjust the retention
times of a non-subset sample based on the alignment results of the
previous subset sample (e.g. a QC sample). If samples are e.g. in the
order <em>A1</em>, <em>B1</em>, <em>B2</em>, <em>A2</em>, <em>B3</em>,
<em>B4</em> with <em>A</em> representing QC samples and <em>B</em> study
samples, using <code>subset = c(1, 4)</code> and
<code>subsetAdjust = "previous"</code> would result in all <em>A</em>
samples to be aligned with each other and non-subset samples <em>B1</em>
and <em>B2</em> being adjusted based on the alignment result of subset
samples <em>A1</em> and <em>B3</em> and <em>B4</em> on those of
<em>A2</em>.</p></li>
<li><p><code>subsetAdjust = "average"</code>: adjust retention times of
non-subset samples based on an interpolation of the alignment results of
the previous and subsequent subset sample. In the example above,
<em>B1</em> would be adjusted based on the average of adjusted retention
times between subset (QC) samples <em>A1</em> and <em>A2</em>. Since
there is no subset sample after non-subset samples <em>B3</em> and
<em>B4</em> these will be adjusted based on the alignment results of
<em>A2</em> alone. Note that a weighted average is used to calculate the
adjusted retention time averages, which uses the inverse of the
difference of the index of the non-subset sample to the subset samples
as weights. Thus, if we have a setup like <em>A1</em>, <em>B1</em>,
<em>B2</em>, <em>A2</em> the adjusted retention times of <em>A1</em>
would get a larger weight than those of <em>A2</em> in the adjustment of
non-subset sample <em>B1</em> causing it’s adjusted retention times to
be closer to those of <em>A1</em> than to <em>A2</em>. See below for
examples.</p></li>
</ul>
<p><strong>Important</strong>: both cases require a meaningful/correct
ordering of the samples within the object (i.e., samples should be
ordered by injection index).</p>
<p>The examples below aim to illustrate the effect of these alignment
options. We assume samples 1, 4 and 7 in the <em>faahKO</em> data set to
be QC pool samples. We thus want to perform the alignment based on these
samples and subsequently adjust the retention times of the left-out
samples (2, 3, 5, 6 and 8) based on interpolation of the results from
the neighboring <em>subset</em> (QC) samples. After initial peak
grouping we perform the subset-based alignment with the <em>peak
groups</em> method by passing the indices of the QC samples with the
<code>subset</code> parameter to the <code>PeakGroupsParam</code>
function and set <code>subsetAdjust = "average"</code> to adjust the
study samples based on interpolation of the alignment results from
neighboring subset/QC samples.</p>
<p>Note that for any subset-alignment all parameters such as
<code>minFraction</code> are relative to the <code>subset</code>, not
the full experiment!</p>
<p>Below we first remove any previous alignment results with the
<code>dropAdjustedRtime</code> function to allow a fresh alignment using
the subset-based option outlined above. In addition to removing adjusted
retention times for all spectra, this function will also
<em>restore</em> the original retention times for identified
chromatographic peaks.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">dropAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define the experimental layout</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">&lt;-</span> <span class="st">"study"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"QC"</span></span></code></pre></div>
<p>For an alignment with the <em>peak groups</em> method an initial peak
grouping (correspondence) analysis is required, because the algorithm
estimates retention times shifts between samples using the retention
times of <em>hook peaks</em> (i.e. chromatographic peaks present in
most/all samples). Here we use the default settings for an <em>peak
density</em> method-based correspondence, but it is strongly advised to
adapt the parameters for each data set (details in the next section).
The definition of the sample groups (i.e. assignment of individual
samples to the sample groups in the experiment) is mandatory for the
<code>PeakDensityParam</code>. If there are no sample groups in the
experiment, <code>sampleGroups</code> should be set to a single value
for each file (e.g. <code>rep(1, length(fileNames(xdata))</code>).</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Initial peak grouping. Use sample_type as grouping variable</span></span>
<span><span class="va">pdp_subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_type</span>,</span>
<span>                             minFraction <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pdp_subs</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define subset-alignment options and perform the alignment</span></span>
<span><span class="va">pgp_subs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">PeakGroupsParam</a></span><span class="op">(</span></span>
<span>    minFraction <span class="op">=</span> <span class="fl">0.85</span>,</span>
<span>    subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">)</span>,</span>
<span>    subsetAdjust <span class="op">=</span> <span class="st">"average"</span>, span <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRtime.html">adjustRtime</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pgp_subs</span><span class="op">)</span></span></code></pre></div>
<p>Below we plot the results of the alignment highlighting the subset
samples in green. This nicely shows how the interpolation of the
<code>subsetAdjust = "average"</code> works: retention times of sample 2
are adjusted based on those from subset sample 1 and 4, giving however
more weight to the closer subset sample 1 which results in the adjusted
retention times of 2 being more similar to those of sample 1. Sample 3
on the other hand gets adjusted giving more weight to the second subset
sample (4).</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"#00000040"</span>, <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">clrs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_type</span> <span class="op">==</span> <span class="st">"QC"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#00ce0080"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4.5</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, aggregationFun <span class="op">=</span> <span class="st">"max"</span>, chromPeaks <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>,</span>
<span>     col <span class="op">=</span> <span class="va">clrs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotAdjustedRtime.html">plotAdjustedRtime</a></span><span class="op">(</span><span class="va">xdata</span>, col <span class="op">=</span> <span class="va">clrs</span>, peakGroupsPch <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  peakGroupsCol <span class="op">=</span> <span class="st">"#00ce0040"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/alignment-subset-plot-2-1.png" alt="Subset-alignment results with option average. Difference between adjusted and raw retention times along the retention time axis. Samples on which the alignment models were estimated are shown in green, study samples in grey." width="960"><p class="caption">
Subset-alignment results with option average. Difference between
adjusted and raw retention times along the retention time axis. Samples
on which the alignment models were estimated are shown in green, study
samples in grey.
</p>
</div>
<p>Option <code>subsetAdjust = "previous"</code> would adjust the
retention times of a non-subset sample based on a single subset sample
(the previous), which results in most cases in the adjusted retention
times of the non-subset sample being highly similar to those of the
subset sample which was used for adjustment.</p>
</div>
</div>
<div class="section level3">
<h3 id="correspondence">Correspondence<a class="anchor" aria-label="anchor" href="#correspondence"></a>
</h3>
<p>Correspondence is usually the final step in LC-MS data pre-processing
in which data, presumably representing signal from the same originating
ions, is matched across samples. As a result, chromatographic peaks from
different samples with similar m/z and retention times get grouped into
LC-MS <em>features</em>. The function to perform the correspondence in
<em>xcms</em> is called <code>groupChromPeaks</code> that again supports
different algorithms which can be selected and configured with a
specific parameter object (see <code><a href="../reference/groupChromPeaks.html">?groupChromPeaks</a></code> for an
overview). For our example we will use the <em>peak density</em> method
<span class="citation">[1]</span> that, within small slices along the
m/z dimension, combines chromatographic peaks depending on the density
of these peaks along the retention time axis. To illustrate this, we
<em>simulate</em> below the peak grouping for an m/z slice containing
multiple chromatoghaphic peaks within each sample using the
<code>plotChromPeakDensity</code> function and a
<code>PeakDensityParam</code> object with parameter
<code>minFraction = 0.4</code> (features are only defined if in at least
40% of samples a chromatographic peak was present) - parameter
<code>sampleGroups</code> is used to define to which sample group each
sample belongs.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define the mz slice.</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">305.05</span>, <span class="fl">305.15</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract and plot the chromatograms</span></span>
<span><span class="va">chr_mzr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">xdata</span>, mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span></span>
<span><span class="co">## Define the parameters for the peak density method</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_group</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">0.4</span>, bw <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotChromPeakDensity.html">plotChromPeakDensity</a></span><span class="op">(</span><span class="va">chr_mzr</span>, col <span class="op">=</span> <span class="va">sample_colors</span>, param <span class="op">=</span> <span class="va">pdp</span>,</span>
<span>                     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_mzr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                     peakCol <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">chr_mzr</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                     peakPch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/correspondence-example-slice-1.png" alt="Example for peak density correspondence. Upper panel: chromatogram for an mz slice with multiple chromatographic peaks. lower panel: identified chromatographic peaks at their retention time (x-axis) and index within samples of the experiments (y-axis) for different values of the bw parameter. The black line represents the peak density estimate. Grouping of peaks (based on the provided settings) is indicated by grey rectangles." width="960"><p class="caption">
Example for peak density correspondence. Upper panel: chromatogram for
an mz slice with multiple chromatographic peaks. lower panel: identified
chromatographic peaks at their retention time (x-axis) and index within
samples of the experiments (y-axis) for different values of the bw
parameter. The black line represents the peak density estimate. Grouping
of peaks (based on the provided settings) is indicated by grey
rectangles.
</p>
</div>
<p>The upper panel in the plot shows the extracted ion chromatogram for
each sample with the detected peaks highlighted. The retention times for
the individual chromatographic peaks in each sample (y-axis being the
index of the sample in the data set) are shown in the lower panel with
the solid black line representing the density estimate for the
distribution of detected peaks along the retention time. Parameter
<code>bw</code> defines the <em>smoothness</em> of this estimation. The
grey rectangles indicate which chromatographic peaks would be grouped
into a feature (each grey rectangle thus representing one feature). This
type of visualization is thus ideal to test, validate and optimize
correspondence settings on manually defined m/z slices before applying
them to the full data set. For the tested m/z slice the settings seemed
to be OK and we are thus applying them to the full data set below.
Especially the parameter <code>bw</code> will be very data set dependent
(or more specifically LC-dependent) and should be adapted to each data
set. See the <a href="https://jorainer.github.io/metabolomics2018" class="external-link">Metabolomics
pre-processing with <code>xcms</code></a> tutorial for examples and more
details.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform the correspondence</span></span>
<span><span class="va">pdp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">PeakDensityParam</a></span><span class="op">(</span>sampleGroups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/MsExperiment.html" class="external-link">sampleData</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">$</span><span class="va">sample_group</span>,</span>
<span>                        minFraction <span class="op">=</span> <span class="fl">0.4</span>, bw <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/groupChromPeaks.html">groupChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="va">pdp</span><span class="op">)</span></span></code></pre></div>
<p>Results from the correspondence analysis can be accessed with the
<code>featureDefinitions</code> and <code>featureValues</code> function.
The former returns a data frame with general information on each of the
defined features, with each row being one feature and columns providing
information on the median m/z and retention time as well as the indices
of the chromatographic peaks assigned to the feature in column
<code>"peakidx"</code>. Below we show the information on the first 6
features.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">featureDefinitions</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       mzmed mzmin mzmax    rtmed    rtmin    rtmax npeaks KO WT      peakidx</span></span>
<span><span class="co">## FT001 200.1 200.1 200.1 2902.177 2882.133 2922.222      2  2  0    458, 1161</span></span>
<span><span class="co">## FT002 205.0 205.0 205.0 2789.457 2782.376 2795.823      8  4  4 44, 443,....</span></span>
<span><span class="co">## FT003 206.0 206.0 206.0 2788.770 2780.807 2793.606      7  3  4 29, 430,....</span></span>
<span><span class="co">## FT004 207.1 207.1 207.1 2718.476 2713.314 2726.487      7  4  3 16, 420,....</span></span>
<span><span class="co">## FT005 233.0 233.0 233.1 3023.353 3014.971 3043.942      7  3  4 69, 959,....</span></span>
<span><span class="co">## FT006 241.1 241.1 241.2 3680.200 3661.067 3695.533      8  3  4 276, 284....</span></span>
<span><span class="co">##       ms_level</span></span>
<span><span class="co">## FT001        1</span></span>
<span><span class="co">## FT002        1</span></span>
<span><span class="co">## FT003        1</span></span>
<span><span class="co">## FT004        1</span></span>
<span><span class="co">## FT005        1</span></span>
<span><span class="co">## FT006        1</span></span></code></pre>
<p>The <code>featureValues</code> function returns a <code>matrix</code>
with rows being features and columns samples. The content of this matrix
can be defined using the <code>value</code> argument which can be any
column name in the <code>chromPeaks</code> matrix. With the default
<code>value = "into"</code> a matrix with the integrated signal of the
peaks corresponding to a feature in a sample are returned. This is then
generally used as the intensity matrix for downstream analysis. Below we
extract the intensities for the first 6 features.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, value <span class="op">=</span> <span class="st">"into"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001        NA  506848.9        NA  169955.6        NA        NA        NA</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7        NA  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9        NA</span></span>
<span><span class="co">## FT005  286221.4        NA  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">## FT006 1160580.5        NA  380970.3  588986.4 1286883.0 1739516.6  639755.3</span></span>
<span><span class="co">##        wt22.CDF</span></span>
<span><span class="co">## FT001        NA</span></span>
<span><span class="co">## FT002 1354004.9</span></span>
<span><span class="co">## FT003  185399.5</span></span>
<span><span class="co">## FT004  221937.5</span></span>
<span><span class="co">## FT005  271128.0</span></span>
<span><span class="co">## FT006  508546.4</span></span></code></pre>
<p>As we can see we have several missing values in this feature matrix.
Missing values are reported if in one sample no chromatographic peak was
detected in the m/z - rt region of the feature. This does however not
necessarily mean that there is no signal for that specific ion in that
sample. The chromatographic peak detection algorithm could also just
have failed to identify any peak in that region, e.g. because the signal
was too noisy or too low. Thus it is advisable to perform, after
correspondence, also a gap-filling (see next section).</p>
<p>The performance of peak detection, alignment and correspondence
should always be evaluated by inspecting extracted ion chromatograms
e.g. of known compounds, internal standards or identified features in
general. The <code>featureChromatograms</code> function allows to
extract chromatograms for each feature present in
<code>featureDefinitions</code>. The returned
<code>MChromatograms</code> object contains an ion chromatogram for each
feature (each row containing the data for one feature) and sample (each
column representing containing data for one sample). Parameter
<code>features</code> allows to define specific features for which the
EIC should be returned. These can be specified with their index or their
ID (i.e. their row name in the <code>featureDefinitions</code> data
frame. If <code>features</code> is not defined, EICs are returned for
<strong>all</strong> features in a data set, which can take also a
considerable amount of time. Below we extract the chromatograms for the
first 4 features.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_chroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/featureChromatograms.html">featureChromatograms</a></span><span class="op">(</span><span class="va">xdata</span>, features <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">feature_chroms</span></span></code></pre></div>
<pre><code><span><span class="co">## XChromatograms with 4 rows and 8 columns</span></span>
<span><span class="co">##             ko15.CDF        ko16.CDF        ko21.CDF        ko22.CDF</span></span>
<span><span class="co">##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;</span></span>
<span><span class="co">## [1,]        peaks: 0        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [3,]        peaks: 1        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## [4,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">##             wt15.CDF        wt16.CDF        wt21.CDF        wt22.CDF</span></span>
<span><span class="co">##      &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt; &lt;XChromatogram&gt;</span></span>
<span><span class="co">## [1,]        peaks: 0        peaks: 0        peaks: 0        peaks: 0</span></span>
<span><span class="co">## [2,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [3,]        peaks: 1        peaks: 1        peaks: 1        peaks: 1</span></span>
<span><span class="co">## [4,]        peaks: 1        peaks: 1        peaks: 0        peaks: 1</span></span>
<span><span class="co">## phenoData with 4 variables</span></span>
<span><span class="co">## featureData with 4 variables</span></span>
<span><span class="co">## - - - xcms preprocessing - - -</span></span>
<span><span class="co">## Chromatographic peak detection:</span></span>
<span><span class="co">##  method: centWave </span></span>
<span><span class="co">## Correspondence:</span></span>
<span><span class="co">##  method: chromatographic peak density </span></span>
<span><span class="co">##  4 feature(s) identified.</span></span></code></pre>
<p>And plot the extracted ion chromatograms. We again use the group
color for each identified peak to fill the area.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">feature_chroms</span>, col <span class="op">=</span> <span class="va">sample_colors</span>,</span>
<span>     peakBg <span class="op">=</span> <span class="va">sample_colors</span><span class="op">[</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">feature_chroms</span><span class="op">)</span><span class="op">[</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/feature-eic-1.png" alt="Extracted ion chromatograms for features 1 to 4." width="768"><p class="caption">
Extracted ion chromatograms for features 1 to 4.
</p>
</div>
<p>To access the EICs of the second feature we can simply subset the
<code>feature_chroms</code> object.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eic_2</span> <span class="op">&lt;-</span> <span class="va">feature_chroms</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">eic_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         mz mzmin mzmax       rt    rtmin    rtmax    into    intb  maxo sn</span></span>
<span><span class="co">## CP0048 205   205   205 2791.365 2770.790 2815.117 1924712 1850331 84280 64</span></span>
<span><span class="co">## CP0495 205   205   205 2795.185 2773.068 2820.462 1757151 1711473 68384 69</span></span>
<span><span class="co">## CP1033 205   205   205 2795.823 2773.753 2821.042 1383417 1334570 47384 54</span></span>
<span><span class="co">## CP1255 205   205   205 2788.583 2768.133 2812.172 1180288 1126958 48336 32</span></span>
<span><span class="co">## CP1535 205   205   205 2782.376 2761.974 2805.911 2129885 2054677 93312 44</span></span>
<span><span class="co">## CP1967 205   205   205 2787.149 2766.797 2812.193 1634342 1566379 67984 53</span></span>
<span><span class="co">## CP2350 205   205   205 2790.332 2763.780 2821.562 1623589 1531573 49208 28</span></span>
<span><span class="co">## CP2601 205   205   205 2787.209 2766.905 2812.192 1354005 1299188 55712 35</span></span>
<span><span class="co">##        sample row column</span></span>
<span><span class="co">## CP0048      1   1      1</span></span>
<span><span class="co">## CP0495      2   1      2</span></span>
<span><span class="co">## CP1033      3   1      3</span></span>
<span><span class="co">## CP1255      4   1      4</span></span>
<span><span class="co">## CP1535      5   1      5</span></span>
<span><span class="co">## CP1967      6   1      6</span></span>
<span><span class="co">## CP2350      7   1      7</span></span>
<span><span class="co">## CP2601      8   1      8</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="gap-filling">Gap filling<a class="anchor" aria-label="anchor" href="#gap-filling"></a>
</h3>
<p>Missing values in LC-MS data are in many cases the result of the
chromatographic peak detection algorithm failing to identify peaks
(because of noisy or low intensity signal). The aim of the gap filling
step is to reduce the number of such missing values by integrating
signals from the original data files for samples in which no
chromatographic peak was found from the m/z - rt region where signal
from the ion is expected. Gap filling can be performed in <em>xcms</em>
with the <code>fillChromPeaks</code> function and a parameter object
selecting and configuring the gap filling algorithm. The method of
choice is <code>ChromPeakAreaParam</code> that integrates the signal (in
samples in which no chromatographic peak was found for a feature) in the
m/z - rt region that is defined based on the m/z and retention time
ranges of all detected chromatographic peaks of that specific feature.
The lower m/z limit of the area is defined as the lower quartile (25%
quantile) of the <code>"mzmin"</code> values of all peaks of the
feature, the upper m/z value as the upper quartile (75% quantile) of the
<code>"mzmax"</code> values, the lower rt value as the lower quartile
(25% quantile) of the <code>"rtmin"</code> and the upper rt value as the
upper quartile (75% quantile) of the <code>"rtmax"</code> values. Below
we perform this gap filling on our test data and extract the feature
values for the first 6 features after gap filling. An <code>NA</code> is
reported if no signal is measured at all for a specific sample.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fillChromPeaks.html">fillChromPeaks</a></span><span class="op">(</span><span class="va">xdata</span>, param <span class="op">=</span> <span class="fu"><a href="../reference/fillChromPeaks.html">ChromPeakAreaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, value <span class="op">=</span> <span class="st">"into"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001  135422.4  506848.9  111872.0  169955.6  210333.1  141880.3  233271.3</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7  167516.7  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9  219142.3</span></span>
<span><span class="co">## FT005  286221.4  286594.4  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">## FT006 1160580.5 1145753.0  380970.3  588986.4 1286883.0 1739516.6  639755.3</span></span>
<span><span class="co">##        wt22.CDF</span></span>
<span><span class="co">## FT001  142254.8</span></span>
<span><span class="co">## FT002 1354004.9</span></span>
<span><span class="co">## FT003  185399.5</span></span>
<span><span class="co">## FT004  221937.5</span></span>
<span><span class="co">## FT005  271128.0</span></span>
<span><span class="co">## FT006  508546.4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="final-result">Final result<a class="anchor" aria-label="anchor" href="#final-result"></a>
</h3>
<p>While we can continue using the <em>xcms</em> result set for further
analysis (e.g. also for feature grouping with the <em><a href="https://bioconductor.org/packages/3.18/MsFeatures" class="external-link">MsFeatures</a></em>
package; see the LC-MS feature grouping vignette for details) we could
also extract all results as a <code>SummarizedExperiment</code> object.
This is the <em>standard</em> data container for Bioconductor defined in
the <em><a href="https://bioconductor.org/packages/3.18/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
package and integration with other Bioconductor packages might thus be
easier using that type of object. Below we use the <code>quantify</code>
function to extract the <em>xcms</em> pre-processing results as such a
<code>SummarizedExperiment</code> object. Internally, the
<code>featureValues</code> function is used to generate the feature
value matrix. We can pass any parameters from that function to the
<code>quantify</code> call. Below we use <code>value = "into"</code> and
<code>method = "sum"</code> to report the integrated peak signal as
intensity and to sum these values in samples in which more than one
chromatographic peak was assigned to a feature (for that option it is
important to run <code>refineChromPeaks</code> like described above to
merge overlapping peaks in each sample).</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'matrixStats'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'IRanges'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:xcms':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     distance</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">quantify</a></span><span class="op">(</span><span class="va">xdata</span>, value <span class="op">=</span> <span class="st">"into"</span>, method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span>
<span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 351 8 </span></span>
<span><span class="co">## metadata(6): '' '' ... '' ''</span></span>
<span><span class="co">## assays(1): raw</span></span>
<span><span class="co">## rownames(351): FT001 FT002 ... FT350 FT351</span></span>
<span><span class="co">## rowData names(10): mzmed mzmin ... WT ms_level</span></span>
<span><span class="co">## colnames(8): ko15.CDF ko16.CDF ... wt21.CDF wt22.CDF</span></span>
<span><span class="co">## colData names(4): sample_name sample_group spectraOrigin sample_type</span></span></code></pre>
<p>The information from <code>featureDefinitions</code> is now stored in
the <code>rowData</code> of this object. The <code>rowData</code>
provides annotations and information for each <strong>row</strong> in
the <code>SummarizedExperiment</code> (which in our case are the
<strong>features</strong>).</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 351 rows and 10 columns</span></span>
<span><span class="co">##           mzmed     mzmin     mzmax     rtmed     rtmin     rtmax    npeaks</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FT001     200.1     200.1     200.1   2902.18   2882.13   2922.22         2</span></span>
<span><span class="co">## FT002     205.0     205.0     205.0   2789.46   2782.38   2795.82         8</span></span>
<span><span class="co">## FT003     206.0     206.0     206.0   2788.77   2780.81   2793.61         7</span></span>
<span><span class="co">## FT004     207.1     207.1     207.1   2718.48   2713.31   2726.49         7</span></span>
<span><span class="co">## FT005     233.0     233.0     233.1   3023.35   3014.97   3043.94         7</span></span>
<span><span class="co">## ...         ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## FT347    595.25     595.2     595.3   3010.39   2992.58   3014.16         6</span></span>
<span><span class="co">## FT348    596.20     596.2     596.2   2997.60   2992.58   3002.62         2</span></span>
<span><span class="co">## FT349    596.30     596.3     596.3   3818.98   3811.68   3835.78         4</span></span>
<span><span class="co">## FT350    597.40     597.4     597.4   3821.10   3817.96   3825.14         3</span></span>
<span><span class="co">## FT351    599.30     599.3     599.3   4070.45   4042.10   4123.52         3</span></span>
<span><span class="co">##              KO        WT  ms_level</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## FT001         2         0         1</span></span>
<span><span class="co">## FT002         4         4         1</span></span>
<span><span class="co">## FT003         3         4         1</span></span>
<span><span class="co">## FT004         4         3         1</span></span>
<span><span class="co">## FT005         3         4         1</span></span>
<span><span class="co">## ...         ...       ...       ...</span></span>
<span><span class="co">## FT347         2         3         1</span></span>
<span><span class="co">## FT348         0         2         1</span></span>
<span><span class="co">## FT349         2         2         1</span></span>
<span><span class="co">## FT350         1         2         1</span></span>
<span><span class="co">## FT351         1         2         1</span></span></code></pre>
<p>Annotations for <strong>columns</strong> (in our case
<strong>samples</strong>) are stored as <code>colData</code>. In this
data frame each row contains annotations for one sample (and hence one
column in the feature values matrix).</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 4 columns</span></span>
<span><span class="co">##          sample_name sample_group spectraOrigin sample_type</span></span>
<span><span class="co">##          &lt;character&gt;  &lt;character&gt;   &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## ko15.CDF        ko15           KO /__w/_temp...          QC</span></span>
<span><span class="co">## ko16.CDF        ko16           KO /__w/_temp...       study</span></span>
<span><span class="co">## ko21.CDF        ko21           KO /__w/_temp...       study</span></span>
<span><span class="co">## ko22.CDF        ko22           KO /__w/_temp...          QC</span></span>
<span><span class="co">## wt15.CDF        wt15           WT /__w/_temp...       study</span></span>
<span><span class="co">## wt16.CDF        wt16           WT /__w/_temp...       study</span></span>
<span><span class="co">## wt21.CDF        wt21           WT /__w/_temp...          QC</span></span>
<span><span class="co">## wt22.CDF        wt22           WT /__w/_temp...       study</span></span></code></pre>
<p>Finally, the feature matrix is stored as an <code>assay</code> within
the object. Note that a <code>SummarizedExperiment</code> can have
multiple assays which have to be numeric matrices with the number of
rows and columns matching the number of features and samples,
respectively. Below we list the names of the available assays.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "raw"</span></span></code></pre>
<p>And we can access the actual data using the <code>assay</code>
function, optionally also providing the name of the assay we want to
access. Below we show the first 6 lines of that matrix.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001  135422.4  506848.9  111872.0  169955.6  210333.1  141880.3  233271.3</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7  167516.7  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9  219142.3</span></span>
<span><span class="co">## FT005  286221.4  286594.4  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">## FT006 1923307.8 1145753.0  380970.3  588986.4 1286883.0 1739516.6  639755.3</span></span>
<span><span class="co">##        wt22.CDF</span></span>
<span><span class="co">## FT001  142254.8</span></span>
<span><span class="co">## FT002 1354004.9</span></span>
<span><span class="co">## FT003  185399.5</span></span>
<span><span class="co">## FT004  221937.5</span></span>
<span><span class="co">## FT005  271128.0</span></span>
<span><span class="co">## FT006  508546.4</span></span></code></pre>
<p>Since a <code>SummarizedExperiment</code> supports multiple assays,
we in addition add also the feature value matrix
<strong>without</strong> filled-in values (i.e. feature intensities that
were added by the gap filling step).</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">$</span><span class="va">raw_nofill</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-peak-grouping-results.html">featureValues</a></span><span class="op">(</span><span class="va">xdata</span>, filled <span class="op">=</span> <span class="cn">FALSE</span>, method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span></code></pre></div>
<p>With that we have now two assays in our result object.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "raw"        "raw_nofill"</span></span></code></pre>
<p>And we can extract the feature values without gap-filling:</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw_nofill"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        ko15.CDF  ko16.CDF  ko21.CDF  ko22.CDF  wt15.CDF  wt16.CDF  wt21.CDF</span></span>
<span><span class="co">## FT001        NA  506848.9        NA  169955.6        NA        NA        NA</span></span>
<span><span class="co">## FT002 1924712.0 1757151.0 1383416.7 1180288.2 2129885.1 1634342.0 1623589.2</span></span>
<span><span class="co">## FT003  213659.3  289500.7        NA  178285.7  253825.6  241844.4  240606.0</span></span>
<span><span class="co">## FT004  349011.5  451863.7  343897.8  208002.8  364609.8  360908.9        NA</span></span>
<span><span class="co">## FT005  286221.4        NA  164009.0  149097.6  255697.7  311296.8  366441.5</span></span>
<span><span class="co">## FT006 1923307.8        NA  380970.3  588986.4 1286883.0 1739516.6  639755.3</span></span>
<span><span class="co">##        wt22.CDF</span></span>
<span><span class="co">## FT001        NA</span></span>
<span><span class="co">## FT002 1354004.9</span></span>
<span><span class="co">## FT003  185399.5</span></span>
<span><span class="co">## FT004  221937.5</span></span>
<span><span class="co">## FT005  271128.0</span></span>
<span><span class="co">## FT006  508546.4</span></span></code></pre>
<p>Finally, a history of the full processing with <em>xcms</em> is
available as <code>metadata</code> in the
<code>SummarizedExperiment</code>.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak detection </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:28 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: CentWaveParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak refinement </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:31 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: MergeNeighboringPeaksParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:47 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Retention time correction </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:48 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakGroupsParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak grouping </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:56 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: PeakDensityParam </span></span>
<span><span class="co">##  MS level(s) 1 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Missing peak filling </span></span>
<span><span class="co">##  date: Wed Oct 18 13:43:02 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: ChromPeakAreaParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>This same information can also be extracted from the <em>xcms</em>
result object using the <code>processHistory</code> function. Below we
extract the information for the first processing step.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class "XProcessHistory"</span></span>
<span><span class="co">##  type: Peak detection </span></span>
<span><span class="co">##  date: Wed Oct 18 13:42:28 2023 </span></span>
<span><span class="co">##  info:  </span></span>
<span><span class="co">##  fileIndex: 1,2,3,4,5,6,7,8 </span></span>
<span><span class="co">##  Parameter class: CentWaveParam </span></span>
<span><span class="co">##  MS level(s) 1</span></span></code></pre>
<p>These processing steps contain also the individual parameter objects
used for the analysis, hence allowing to exactly reproduce the
analysis.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">processHistory</a></span><span class="op">(</span><span class="va">xdata</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="../reference/ProcessHistory-class.html">processParam</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class:  CentWaveParam </span></span>
<span><span class="co">##  Parameters:</span></span>
<span><span class="co">##  - ppm: [1] 25</span></span>
<span><span class="co">##  - peakwidth: [1] 20 80</span></span>
<span><span class="co">##  - snthresh: [1] 10</span></span>
<span><span class="co">##  - prefilter: [1]    6 5000</span></span>
<span><span class="co">##  - mzCenterFun: [1] "wMean"</span></span>
<span><span class="co">##  - integrate: [1] 1</span></span>
<span><span class="co">##  - mzdiff: [1] -0.001</span></span>
<span><span class="co">##  - fitgauss: [1] FALSE</span></span>
<span><span class="co">##  - noise: [1] 5000</span></span>
<span><span class="co">##  - verboseColumns: [1] FALSE</span></span>
<span><span class="co">##  - roiList: list()</span></span>
<span><span class="co">##  - firstBaselineCheck: [1] TRUE</span></span>
<span><span class="co">##  - roiScales: numeric(0)</span></span>
<span><span class="co">##  - extendLengthMSW: [1] FALSE</span></span></code></pre>
<p>At last we perform also a principal component analysis to evaluate
the grouping of the samples in this experiment. Note that we did not
perform any data normalization hence the grouping might (and will) also
be influenced by technical biases.</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract the features and log2 transform them</span></span>
<span><span class="va">ft_ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">res</span>, <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Perform the PCA omitting all features with an NA in any of the</span></span>
<span><span class="co">## samples. Also, the intensities are mean centered.</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">ft_ints</span><span class="op">)</span><span class="op">)</span>, center <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot the PCA</span></span>
<span><span class="va">pcSummary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">pc</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">21</span>, main <span class="op">=</span> <span class="st">""</span>,</span>
<span>     xlab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">pcSummary</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>                                   digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, <span class="st">" % variance"</span><span class="op">)</span>,</span>
<span>     ylab <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="va">pcSummary</span><span class="op">$</span><span class="va">importance</span><span class="op">[</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>                                   digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, <span class="st">" % variance"</span><span class="op">)</span>,</span>
<span>     col <span class="op">=</span> <span class="st">"darkgrey"</span>, bg <span class="op">=</span> <span class="va">sample_colors</span>, cex <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="va">pc</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, labels <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">sample_name</span>, col <span class="op">=</span> <span class="st">"darkgrey"</span>,</span>
<span>     pos <span class="op">=</span> <span class="fl">3</span>, cex <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="xcms_files/figure-html/final-pca-1.png" alt="PCA for the faahKO data set, un-normalized intensities." width="768"><p class="caption">
PCA for the faahKO data set, un-normalized intensities.
</p>
</div>
<p>We can see the expected separation between the KO and WT samples on
PC2. On PC1 samples separate based on their ID, samples with an ID &lt;=
18 from samples with an ID &gt; 18. This separation might be caused by a
technical bias (e.g. measurements performed on different days/weeks) or
due to biological properties of the mice analyzed (sex, age, litter
mates etc).</p>
</div>
</div>
<div class="section level2">
<h2 id="further-data-processing-and-analysis">Further data processing and analysis<a class="anchor" aria-label="anchor" href="#further-data-processing-and-analysis"></a>
</h2>
<p>Normalizing features’ signal intensities is required, but at present
not (yet) supported in <code>xcms</code> (some methods might be added in
near future). It is advised to use the <code>SummarizedExperiment</code>
returned by the <code>quantify</code> method for any further data
processing, as this type of object stores feature definitions, sample
annotations as well as feature abundances in the same object. For the
identification of e.g. features with significant different
intensities/abundances it is suggested to use functionality provided in
other R packages, such as Bioconductor’s excellent <code>limma</code>
package.</p>
</div>
<div class="section level2">
<h2 id="additional-details-and-notes">Additional details and notes<a class="anchor" aria-label="anchor" href="#additional-details-and-notes"></a>
</h2>
<div class="section level3">
<h3 id="subsetting-and-filtering">Subsetting and filtering<a class="anchor" aria-label="anchor" href="#subsetting-and-filtering"></a>
</h3>
<p><em>xcms</em> result objects can be subset/filtered by sample using
the <code>[</code> method or one of the <code>filter*</code> functions
(although the <code>XcmsExperiment</code> supports at present only few
selected filter functions). In some cases filtering can remove
pre-processing results, but most filter functions support parameters
<code>keepFeatures</code> and <code>keepAdjustedRtime</code> that can be
set to <code>TRUE</code> to avoid their removal.</p>
</div>
<div class="section level3">
<h3 id="parallel-processing">Parallel processing<a class="anchor" aria-label="anchor" href="#parallel-processing"></a>
</h3>
<p>Most functions in <em>xcms</em> support parallel processing, which is
enabled by default and is performed, for most operations, on a per-file
basis. Parallel processing is handled and configured by the
<code>BiocParallel</code> Bioconductor package and can be globally
defined for an R session. Note that, while all data objects are designed
to have a low memory footprint by loading only peak data into memory
when needed, parallel processing will increase this demand: in general,
the full data for as many files as there are parallel processes are
loaded into memory. This needs also to be considered, when the number of
parallel processes is defined.</p>
<p>Unix-based systems (Linux, macOS) support
<code>multicore</code>-based parallel processing. To configure it
globally we <code>register</code> the parameter class. Note also that
<code>bpstart</code> is used below to initialize the parallel
processes.</p>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Windows supports only socket-based parallel processing:</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html" class="external-link">register</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">bpstart</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="main-differences-to-the-msnbase-based-xcms-version-3">Main differences to the <code>MSnbase</code>-based xcms version
3<a class="anchor" aria-label="anchor" href="#main-differences-to-the-msnbase-based-xcms-version-3"></a>
</h3>
<ul>
<li>
<code>Spectra</code> is used as the main container for the (raw) MS
data. Thus changing backends is supported at any stage of the analysis
(e.g. to load all data into memory or to load data from remote
databases).</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="additional-documentation-resources">Additional documentation resources<a class="anchor" aria-label="anchor" href="#additional-documentation-resources"></a>
</h2>
<p>Some of the documentations listed here are still based on xcms
version 3 but will be subsequently updated.</p>
<ul>
<li>
<a href="https://jorainer.github.io/metabolomics2018" class="external-link">Metabolomics
pre-processing with <code>xcms</code></a>: more detailed description of
the pre-processing of LC-MS data with <em>xcms</em>.</li>
<li>
<a href="https://jorainer.github.io/MetaboAnnotationTutorials" class="external-link">MetaboAnnotationTutorials</a>:
examples for annotation of metabolomics data from <span class="citation">[3]</span>.</li>
<li>
<span class="citation">[2]</span>: describes the concept of the
<em>on-disk</em> data modes used also in <em>xcms</em>.</li>
<li>
<a href="">SpectraTutorials</a>: tutorials describing the <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package and its functionality.</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>R packages used for this document are listed below.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SummarizedExperiment_1.31.1 GenomicRanges_1.53.2       </span></span>
<span><span class="co">##  [3] GenomeInfoDb_1.37.6         IRanges_2.35.3             </span></span>
<span><span class="co">##  [5] MatrixGenerics_1.13.1       matrixStats_1.0.0          </span></span>
<span><span class="co">##  [7] MsExperiment_1.3.0          pheatmap_1.0.12            </span></span>
<span><span class="co">##  [9] RColorBrewer_1.1-3          pander_0.6.5               </span></span>
<span><span class="co">## [11] faahKO_1.41.0               xcms_3.99.6                </span></span>
<span><span class="co">## [13] MSnbase_2.27.1              ProtGenerics_1.33.1        </span></span>
<span><span class="co">## [15] S4Vectors_0.39.3            mzR_2.35.1                 </span></span>
<span><span class="co">## [17] Rcpp_1.0.11                 Biobase_2.61.0             </span></span>
<span><span class="co">## [19] BiocGenerics_0.47.0         BiocParallel_1.35.4        </span></span>
<span><span class="co">## [21] BiocStyle_2.29.2           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-7                rlang_1.1.1                </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              clue_0.3-65                </span></span>
<span><span class="co">##   [5] MassSpecWavelet_1.67.0      compiler_4.3.1             </span></span>
<span><span class="co">##   [7] systemfonts_1.0.5           vctrs_0.6.4                </span></span>
<span><span class="co">##   [9] stringr_1.5.0               MetaboCoreUtils_1.9.4      </span></span>
<span><span class="co">##  [11] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">##  [13] fastmap_1.1.1               XVector_0.41.1             </span></span>
<span><span class="co">##  [15] utf8_1.2.3                  rmarkdown_2.25             </span></span>
<span><span class="co">##  [17] preprocessCore_1.63.1       ragg_1.2.6                 </span></span>
<span><span class="co">##  [19] purrr_1.0.2                 MultiAssayExperiment_1.27.5</span></span>
<span><span class="co">##  [21] xfun_0.40                   zlibbioc_1.47.0            </span></span>
<span><span class="co">##  [23] cachem_1.0.8                jsonlite_1.8.7             </span></span>
<span><span class="co">##  [25] progress_1.2.2              DelayedArray_0.27.10       </span></span>
<span><span class="co">##  [27] prettyunits_1.2.0           parallel_4.3.1             </span></span>
<span><span class="co">##  [29] cluster_2.1.4               R6_2.5.1                   </span></span>
<span><span class="co">##  [31] bslib_0.5.1                 stringi_1.7.12             </span></span>
<span><span class="co">##  [33] limma_3.57.10               jquerylib_0.1.4            </span></span>
<span><span class="co">##  [35] bookdown_0.36               iterators_1.0.14           </span></span>
<span><span class="co">##  [37] knitr_1.44                  igraph_1.5.1               </span></span>
<span><span class="co">##  [39] splines_4.3.1               Matrix_1.6-1.1             </span></span>
<span><span class="co">##  [41] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">##  [43] yaml_2.3.7                  doParallel_1.0.17          </span></span>
<span><span class="co">##  [45] codetools_0.2-19            affy_1.79.3                </span></span>
<span><span class="co">##  [47] lattice_0.21-9              tibble_3.2.1               </span></span>
<span><span class="co">##  [49] plyr_1.8.9                  evaluate_0.22              </span></span>
<span><span class="co">##  [51] survival_3.5-7              desc_1.4.2                 </span></span>
<span><span class="co">##  [53] Spectra_1.11.11             pillar_1.9.0               </span></span>
<span><span class="co">##  [55] affyio_1.71.0               BiocManager_1.30.22        </span></span>
<span><span class="co">##  [57] foreach_1.5.2               MALDIquant_1.22.1          </span></span>
<span><span class="co">##  [59] ncdf4_1.21                  generics_0.1.3             </span></span>
<span><span class="co">##  [61] rprojroot_2.0.3             RCurl_1.98-1.12            </span></span>
<span><span class="co">##  [63] hms_1.1.3                   ggplot2_3.4.4              </span></span>
<span><span class="co">##  [65] munsell_0.5.0               scales_1.2.1               </span></span>
<span><span class="co">##  [67] glue_1.6.2                  lazyeval_0.2.2             </span></span>
<span><span class="co">##  [69] MsFeatures_1.9.0            tools_4.3.1                </span></span>
<span><span class="co">##  [71] mzID_1.39.0                 robustbase_0.99-0          </span></span>
<span><span class="co">##  [73] QFeatures_1.11.2            vsn_3.69.0                 </span></span>
<span><span class="co">##  [75] RANN_2.6.1                  fs_1.6.3                   </span></span>
<span><span class="co">##  [77] XML_3.99-0.14               grid_4.3.1                 </span></span>
<span><span class="co">##  [79] impute_1.75.1               MsCoreUtils_1.13.1         </span></span>
<span><span class="co">##  [81] colorspace_2.1-0            GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">##  [83] cli_3.6.1                   textshaping_0.3.7          </span></span>
<span><span class="co">##  [85] fansi_1.0.5                 S4Arrays_1.1.6             </span></span>
<span><span class="co">##  [87] dplyr_1.1.3                 AnnotationFilter_1.25.0    </span></span>
<span><span class="co">##  [89] pcaMethods_1.93.0           gtable_0.3.4               </span></span>
<span><span class="co">##  [91] DEoptimR_1.1-3              sass_0.4.7                 </span></span>
<span><span class="co">##  [93] digest_0.6.33               SparseArray_1.1.12         </span></span>
<span><span class="co">##  [95] farver_2.1.1                multtest_2.57.0            </span></span>
<span><span class="co">##  [97] memoise_2.0.1               htmltools_0.5.6.1          </span></span>
<span><span class="co">##  [99] pkgdown_2.0.7.9000          lifecycle_1.0.3            </span></span>
<span><span class="co">## [101] statmod_1.5.0               MASS_7.3-60</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Smith:2006ic" class="csl-entry">
1. Smith CA, Want EJ, O’Maille G, Abagyan R, Siuzdak G: <strong><span class="nocase">XCMS: processing mass spectrometry data for metabolite
profiling using nonlinear peak alignment, matching, and
identification.</span></strong> <em>Analytical chemistry</em> 2006,
<strong>78</strong>:779–787.
</div>
<div id="ref-gattoMSnbaseEfficientElegant2020a" class="csl-entry">
2. Gatto L, Gibb S, Rainer J: <strong><a href="https://doi.org/10.1021/acs.jproteome.0c00313" class="external-link"><span>MSnbase</span>,
<span>Efficient</span> and <span>Elegant R</span>-<span>Based
Processing</span> and <span>Visualization</span> of <span>Raw Mass
Spectrometry Data</span></a></strong>. <em>Journal of Proteome
Research</em> 2020.
</div>
<div id="ref-rainer_modular_2022" class="csl-entry">
3. Rainer J, Vicini A, Salzer L, Stanstrup J, Badia JM, Neumann S,
Stravs MA, Verri Hernandes V, Gatto L, Gibb S, Witting M: <strong><a href="https://doi.org/10.3390/metabo12020173" class="external-link">A <span>Modular</span> and
<span>Expandable</span> <span>Ecosystem</span> for
<span>Metabolomics</span> <span>Data</span> <span>Annotation</span> in
<span>R</span></a></strong>. <em>Metabolites</em> 2022,
<strong>12</strong>:173.
</div>
<div id="ref-Saghatelian04" class="csl-entry">
4. Saghatelian A, Trauger SA, Want EJ, Hawkins EG, Siuzdak G, Cravatt
BF: <strong><a href="http://dx.doi.org/10.1021/bi0480335" class="external-link">Assignment of
endogenous substrates to enzymes by global metabolite
profiling</a></strong>. <em>Biochemistry</em> 2004,
<strong>43</strong>:14332–9.
</div>
<div id="ref-Tautenhahn:2008fx" class="csl-entry">
5. Tautenhahn R, Böttcher C, Neumann S: <strong><span class="nocase">Highly sensitive feature detection for high resolution
LC/MS.</span></strong> <em>BMC Bioinformatics</em> 2008,
<strong>9</strong>:504.
</div>
<div id="ref-Smith:2013gr" class="csl-entry">
6. Smith R, Ventura D, Prince JT: <strong><span class="nocase">LC-MS
alignment in theory and practice: a comprehensive algorithmic
review.</span></strong> <em>Briefings in bioinformatics</em> 2013,
<strong>16</strong>:bbt080–117.
</div>
<div id="ref-Prince:2006jj" class="csl-entry">
7. Prince JT, Marcotte EM: <strong><span class="nocase">Chromatographic
alignment of ESI-LC-MS proteomics data sets by ordered bijective
interpolated warping.</span></strong> <em>Analytical chemistry</em>
2006, <strong>78</strong>:6140–6152.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Colin A. Smith, Ralf Tautenhahn, Steffen Neumann, Paul Benton, Christopher Conley, Johannes Rainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
