<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="xcms">
<title>LC-MS/MS data analysis with xcms • xcms</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LC-MS/MS data analysis with xcms">
<meta property="og:description" content="xcms">
<meta property="og:image" content="https://sneumann.github.io/xcms/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">xcms</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.99.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/xcms.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/LC-MS-feature-grouping.html">Compounding (grouping) of LC-MS features</a>
    <a class="dropdown-item" href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    <a class="dropdown-item" href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/sneumann/xcms/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>LC-MS/MS data analysis with xcms</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/HEAD/vignettes/xcms-lcms-ms.Rmd" class="external-link"><code>vignettes/xcms-lcms-ms.Rmd</code></a></small>
      <div class="d-none name"><code>xcms-lcms-ms.Rmd</code></div>
    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em><br><strong>Authors</strong>: Johannes Rainer, Michael Witting<br><strong>Modified</strong>: 2023-10-18 12:21:55.86337<br><strong>Compiled</strong>: Wed Oct 18 13:40:43 2023</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Metabolite identification is an important step in non-targeted
metabolomics and requires different steps. One involves the use of
tandem mass spectrometry to generate fragmentation spectra of detected
metabolites (LC-MS/MS), which are then compared to fragmentation spectra
of known metabolites. Different approaches exist for the generation of
these fragmentation spectra, whereas the most used is data dependent
acquisition (DDA) also known as the top-n method. In this method the top
N most intense ions (m/z values) from a MS1 scan are selected for
fragmentation in the next N scans before the cycle starts again. This
method allows to generate clean MS2 fragmentation spectra on the fly
during acquisition without the need for further experiments, but suffers
from poor coverage of the detected metabolites (since only a limited
number of ions are fragmented) and less accurate quantification of the
compounds (since fewer MS1 scans are generated).</p>
<p>Data independent approaches (DIA) like Bruker bbCID, Agilent AllIons
or Waters MSe don’t use such a preselection, but rather fragment all
detected molecules at once. They are using alternating schemes with scan
of low and high collision energy to collect MS1 and MS2 data. Using this
approach, there is no problem in coverage, but the relation between the
precursor and fragment masses is lost leading to chimeric spectra.
Sequential Window Acquisition of all Theoretical Mass Spectra (or SWATH
<span class="citation">[1]</span>) combines both approaches through a
middle-way approach. There is no precursor selection and acquisition is
independent of acquired data, but rather than isolating all precusors at
once, defined windows (i.e. ranges of m/z values) are used and scanned.
This reduces the overlap of fragment spectra while still keeping a high
coverage.</p>
<p>This document showcases the analysis of two small LC-MS/MS data sets
using <em><a href="https://bioconductor.org/packages/3.18/xcms" class="external-link">xcms</a></em>. The
data files used are reversed-phase LC-MS/MS runs from the Agilent
Pesticide mix obtained from a Sciex 6600 Triple ToF operated in SWATH
acquisition mode. For comparison a DDA file from the same sample is
included.</p>
</div>
<div class="section level2">
<h2 id="analysis-of-dda-data">Analysis of DDA data<a class="anchor" aria-label="anchor" href="#analysis-of-dda-data"></a>
</h2>
<p>Below we load the example DDA data set and create a total ion
chromatogram of its MS1 data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sneumann/xcms" class="external-link">xcms</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsExperiment" class="external-link">MsExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dda_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"TripleTOF-SWATH"</span>, <span class="st">"PestMix1_DDA.mzML"</span>,</span>
<span>                        package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span></span>
<span><span class="va">dda_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/readMsExperiment.html" class="external-link">readMsExperiment</a></span><span class="op">(</span><span class="va">dda_file</span><span class="op">)</span></span>
<span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">dda_data</span>, aggregationFun <span class="op">=</span> <span class="st">"sum"</span>, msLevel <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></code></pre></div>
<p>According to the TIC most of the signal is measured between ~ 200 and
600 seconds (see plot below). We thus filter the DDA data to this
retention time range.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">230</span>, <span class="fl">610</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="xcms-lcms-ms_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## filter the data</span></span>
<span><span class="va">dda_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">dda_data</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">230</span>, <span class="fl">610</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Filter spectra</span></span></code></pre>
<p>The variable <code>dda_data</code> contains now all MS1 and MS2
spectra from the specified mzML file. The number of spectra for each MS
level is listed below. Note that we subset the experiment to the first
data file (using <code>[1]</code>) and then access directly the spectra
within this sample with the <code>spectra</code> function (which returns
a <code>Spectra</code> object from the <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package). Note that we use the pipe operator <code>|&gt;</code> for
better readability.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dda_data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">msLevel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    1    2 </span></span>
<span><span class="co">## 1389 2214</span></span></code></pre>
<p>For the MS2 spectra we can get the m/z of the precursor ion with the
<code>precursorMz</code> function. Below we first subset the data set
again to a single sample and filter to spectra from MS level 2
extracting then their precursor m/z values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dda_data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMsLevel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorMz</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 182.18777 182.18893  55.00579 182.19032 237.12296 237.11987</span></span></code></pre>
<p>With the <code>precursorIntensity</code> function it is also possible
to extract the intensity of the precursor ion.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dda_data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMsLevel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorIntensity</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0 0 0 0 0 0</span></span></code></pre>
<p>Some manufacturers (like Sciex) don’t define/export the precursor
intensity and thus either <code>NA</code> or <code>0</code> is reported.
We can however use the <code>estimatePrecursorIntensity</code> function
from the <em><a href="https://bioconductor.org/packages/3.18/Spectra" class="external-link">Spectra</a></em>
package to determine the precursor intensity for a MS 2 spectrum based
on the intensity of the respective ion in the previous MS1 scan (note
that with <code>method = "interpolation"</code> the precursor intensity
would be defined based on interpolation between the intensity in the
previous and subsequent MS1 scan). Below we estimate the precursor
intensities, on the full data (for MS1 spectra a <code>NA</code> value
is reported).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prec_int</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimatePrecursorIntensity.html">estimatePrecursorIntensity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">dda_data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We next set the precursor intensity in the spectrum metadata of
<code>dda_data</code>. So that it can be extracted later with the
<code>precursorIntensity</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">dda_data</span><span class="op">)</span><span class="op">$</span><span class="va">precursorIntensity</span> <span class="op">&lt;-</span> <span class="va">prec_int</span></span>
<span></span>
<span><span class="va">dda_data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterMsLevel</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorIntensity</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]        NA  9.198155  2.773988 27.590797  3.443145  7.621923</span></span></code></pre>
<p>Next we perform the chromatographic peak detection on the MS level 1
data with the <code>findChromPeaks</code> method. Below we define the
settings for a <em>centWave</em>-based peak detection and perform the
analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>snthresh <span class="op">=</span> <span class="fl">5</span>, noise <span class="op">=</span> <span class="fl">100</span>, ppm <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                     peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dda_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks.html">findChromPeaks</a></span><span class="op">(</span><span class="va">dda_data</span>, param <span class="op">=</span> <span class="va">cwp</span>, msLevel <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span></span></code></pre></div>
<p>In total 114 peaks were identified in the present data set.</p>
<p>The advantage of LC-MS/MS data is that (MS1) ions are fragmented and
the corresponding MS2 spectra measured. Thus, for some of the ions
(identified as MS1 chromatographic peaks) MS2 spectra are available.
These can facilitate the annotation of the respective MS1
chromatographic peaks (or MS1 features after a correspondence analysis).
Spectra for identified chromatographic peaks can be extracted with the
<code>chromPeakSpectra</code> method. MS2 spectra with their precursor
m/z and retention time within the rt and m/z range of the
chromatographic peak are returned.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/Spectra" class="external-link">Spectra</a></span><span class="op">)</span></span>
<span><span class="va">dda_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromPeakSpectra.html">chromPeakSpectra</a></span><span class="op">(</span><span class="va">dda_data</span>, msLevel <span class="op">=</span> <span class="fl">2L</span><span class="op">)</span></span>
<span><span class="va">dda_spectra</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 142 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##       msLevel     rtime scanIndex</span></span>
<span><span class="co">##     &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1           2   237.869      1812</span></span>
<span><span class="co">## 2           2   241.299      1846</span></span>
<span><span class="co">## 3           2   325.763      2439</span></span>
<span><span class="co">## 4           2   326.583      2446</span></span>
<span><span class="co">## 5           2   327.113      2450</span></span>
<span><span class="co">## ...       ...       ...       ...</span></span>
<span><span class="co">## 138         2   574.725      5110</span></span>
<span><span class="co">## 139         2   575.255      5115</span></span>
<span><span class="co">## 140         2   596.584      5272</span></span>
<span><span class="co">## 141         2   592.424      5236</span></span>
<span><span class="co">## 142         2   596.054      5266</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## PestMix1_DDA.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [230..610] on MS level(s) 1 2 [Wed Oct 18 13:41:01 2023]</span></span>
<span><span class="co">##  Filter: select MS level(s) 2 [Wed Oct 18 13:41:14 2023]</span></span>
<span><span class="co">##  Merge 1 Spectra into one [Wed Oct 18 13:41:14 2023]</span></span></code></pre>
<p>By default <code>chromPeakSpectra</code> returns all spectra
associated with a MS1 chromatographic peak, but parameter
<code>method</code> allows to choose and return only one spectrum per
peak (have a look at the <code><a href="../reference/chromPeakSpectra.html">?chromPeakSpectra</a></code> help page for
more details). Also, it would be possible to extract MS1 spectra for
each peak by specifying <code>msLevel = 1L</code> in the call above
(e.g. to evaluate the full MS1 signal at the peak’s apex position).</p>
<p>The returned <code>Spectra</code> contains also the reference to the
respective chromatographic peak as additional <em>spectra variable</em>
<code>"peak_id"</code> that contains the identifier for the
chromatographic peak (i.e. its row name in the <code>chromPeaks</code>
matrix).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dda_spectra</span><span class="op">$</span><span class="va">peak_id</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] "CP004" "CP004" "CP005" "CP005" "CP006" "CP006" "CP008" "CP008" "CP011"</span></span>
<span><span class="co">##  [10] "CP011" "CP012" "CP012" "CP013" "CP013" "CP013" "CP013" "CP014" "CP014"</span></span>
<span><span class="co">##  [19] "CP014" "CP014" "CP018" "CP022" "CP022" "CP022" "CP025" "CP025" "CP025"</span></span>
<span><span class="co">##  [28] "CP025" "CP026" "CP026" "CP026" "CP026" "CP033" "CP033" "CP034" "CP034"</span></span>
<span><span class="co">##  [37] "CP034" "CP034" "CP034" "CP035" "CP035" "CP035" "CP041" "CP041" "CP041"</span></span>
<span><span class="co">##  [46] "CP042" "CP042" "CP042" "CP043" "CP047" "CP047" "CP049" "CP049" "CP049"</span></span>
<span><span class="co">##  [55] "CP049" "CP050" "CP050" "CP050" "CP051" "CP051" "CP051" "CP054" "CP055"</span></span>
<span><span class="co">##  [64] "CP055" "CP055" "CP056" "CP056" "CP056" "CP056" "CP056" "CP060" "CP060"</span></span>
<span><span class="co">##  [73] "CP060" "CP060" "CP064" "CP064" "CP065" "CP065" "CP066" "CP066" "CP069"</span></span>
<span><span class="co">##  [82] "CP069" "CP069" "CP070" "CP070" "CP070" "CP072" "CP072" "CP072" "CP073"</span></span>
<span><span class="co">##  [91] "CP074" "CP074" "CP074" "CP074" "CP075" "CP075" "CP075" "CP077" "CP077"</span></span>
<span><span class="co">## [100] "CP077" "CP079" "CP079" "CP079" "CP079" "CP080" "CP080" "CP080" "CP081"</span></span>
<span><span class="co">## [109] "CP086" "CP086" "CP086" "CP086" "CP086" "CP088" "CP088" "CP088" "CP089"</span></span>
<span><span class="co">## [118] "CP089" "CP091" "CP091" "CP093" "CP093" "CP094" "CP094" "CP094" "CP095"</span></span>
<span><span class="co">## [127] "CP095" "CP095" "CP096" "CP096" "CP096" "CP098" "CP098" "CP098" "CP098"</span></span>
<span><span class="co">## [136] "CP098" "CP099" "CP099" "CP099" "CP100" "CP101" "CP101"</span></span></code></pre>
<p>Note also that with <code>return.type = "List"</code> a list parallel
to the <code>chromPeaks</code> matrix would be returned, i.e. each
element in that list would contain the spectra for the chromatographic
peak with the same index. Such data representation might eventually
simplify further processing.</p>
<p>We next use the MS2 information to aid in the annotation of a
chromatographic peak. As an example we use a chromatographic peak of an
ion with an m/z of 304.1131 which we extract in the code block
below.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex_mz</span> <span class="op">&lt;-</span> <span class="fl">304.1131</span></span>
<span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">dda_data</span>, mz <span class="op">=</span> <span class="va">ex_mz</span>, ppm <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             mz    mzmin    mzmax      rt   rtmin   rtmax    into     intb</span></span>
<span><span class="co">## CP056 304.1133 304.1126 304.1143 425.024 417.985 441.773 13040.8 13007.79</span></span>
<span><span class="co">##           maxo  sn sample</span></span>
<span><span class="co">## CP056 3978.987 232      1</span></span></code></pre>
<p>A search of potential ions with a similar m/z in a reference database
(e.g. <a href="https://metlin.scripps.edu" class="external-link">Metlin</a>) returned a large
list of potential hits, most with a very small ppm. For two of the hits,
<a href="https://en.wikipedia.org/wiki/Flumazenil" class="external-link">Flumazenil</a>
(Metlin ID 2724) and <a href="https://en.wikipedia.org/wiki/Fenamiphos" class="external-link">Fenamiphos</a> (Metlin
ID 72445) experimental MS2 spectra are available. Thus, we could match
the MS2 spectrum for the identified chromatographic peak against these
to annotate our ion. Below we extract all MS2 spectra that were
associated with the candidate chromatographic peak using the ID of the
peak in the present data set.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">dda_data</span>, mz <span class="op">=</span> <span class="va">ex_mz</span>, ppm <span class="op">=</span> <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ex_spectra</span> <span class="op">&lt;-</span> <span class="va">dda_spectra</span><span class="op">[</span><span class="va">dda_spectra</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">==</span> <span class="va">ex_id</span><span class="op">]</span></span>
<span><span class="va">ex_spectra</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 5 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##     msLevel     rtime scanIndex</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1         2   418.926      3505</span></span>
<span><span class="co">## 2         2   419.306      3510</span></span>
<span><span class="co">## 3         2   423.036      3582</span></span>
<span><span class="co">## 4         2   423.966      3603</span></span>
<span><span class="co">## 5         2   424.296      3609</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## PestMix1_DDA.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [230..610] on MS level(s) 1 2 [Wed Oct 18 13:41:01 2023]</span></span>
<span><span class="co">##  Filter: select MS level(s) 2 [Wed Oct 18 13:41:14 2023]</span></span>
<span><span class="co">##  Merge 1 Spectra into one [Wed Oct 18 13:41:14 2023]</span></span></code></pre>
<p>There are 5 MS2 spectra representing fragmentation of the ion(s)
measured in our candidate chromatographic peak. We next reduce this to a
single MS2 spectrum using the <code>combineSpectra</code> method
employing the <code>combinePeaks</code> function to determine which
peaks to keep in the resulting spectrum (have a look at the
<code><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">?combinePeaks</a></code> help page for details). Parameter
<code>f</code> allows to specify which spectra in the input object
should be combined into one. Note that this combination of multiple
fragment spectra into a single spectrum might not be generally the best
approach or suggested for all types of data.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">combineSpectra</a></span><span class="op">(</span><span class="va">ex_spectra</span>, FUN <span class="op">=</span> <span class="va">combinePeaks</span>, ppm <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                              peaks <span class="op">=</span> <span class="st">"intersect"</span>, minProp <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>                              intensityFun <span class="op">=</span> <span class="va">median</span>, mzFun <span class="op">=</span> <span class="va">median</span>,</span>
<span>                              f <span class="op">=</span> <span class="va">ex_spectra</span><span class="op">$</span><span class="va">peak_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: 'combinePeaks' for lists of peak matrices is deprecated; please use</span></span>
<span><span class="co">## 'combinePeaksData' instead.</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ex_spectrum</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 1 spectra in a MsBackendMemory backend:</span></span>
<span><span class="co">##     msLevel     rtime scanIndex</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1         2   418.926      3505</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [230..610] on MS level(s) 1 2 [Wed Oct 18 13:41:01 2023]</span></span>
<span><span class="co">##  Filter: select MS level(s) 2 [Wed Oct 18 13:41:14 2023]</span></span>
<span><span class="co">##  Merge 1 Spectra into one [Wed Oct 18 13:41:14 2023]</span></span>
<span><span class="co">##  ...1 more processings. Use 'processingLog' to list all.</span></span></code></pre>
<p>Mass peaks from all input spectra with a difference in m/z smaller 20
ppm (parameter <code>ppm</code>) were combined into one peak and the
median m/z and intensity is reported for these. Due to parameter
<code>minProp = 0.8</code>, the resulting MS2 spectrum contains only
peaks that were present in 80% of the input spectra.</p>
<p>A plot of this <em>consensus</em> spectrum is shown below.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">ex_spectrum</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/dda-ms2-consensus-plot-1.png" alt="Consensus MS2 spectrum created from all measured MS2 spectra for ions of chromatographic peak CP53." width="768"><p class="caption">
Consensus MS2 spectrum created from all measured MS2 spectra for ions of
chromatographic peak CP53.
</p>
</div>
<p>We could now match the consensus spectrum against a database of MS2
spectra. In our example we simply load MS2 spectra for the two compounds
with matching m/z exported from Metlin. For each of the compounds MS2
spectra created with collision energies of 0V, 10V, 20V and 40V are
available. Below we import the respective data and plot our candidate
spectrum against the MS2 spectra of Flumanezil and Fenamiphos (from a
collision energy of 20V). To import files in MGF format we have to load
the <em><a href="https://bioconductor.org/packages/3.18/MsBackendMgf" class="external-link">MsBackendMgf</a></em>
Bioconductor package which adds MGF file support to the
<code>Spectra</code> package.</p>
<p>Prior plotting we <em>scale</em> our experimental spectra to replace
all peak intensities with values relative to the maximum peak intensity
(which is set to a value of 100).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scale_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">z</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">z</span><span class="op">[</span>, <span class="st">"intensity"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z</span><span class="op">[</span>, <span class="st">"intensity"</span><span class="op">]</span> <span class="op">/</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">z</span><span class="op">[</span>, <span class="st">"intensity"</span><span class="op">]</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span>
<span>    <span class="va">z</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ex_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">ex_spectrum</span>, FUN <span class="op">=</span> <span class="va">scale_fun</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MsBackendMgf" class="external-link">MsBackendMgf</a></span><span class="op">)</span></span>
<span><span class="va">flumanezil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mgf"</span>, <span class="st">"metlin-2724.mgf"</span>, package <span class="op">=</span> <span class="st">"xcms"</span><span class="op">)</span>,</span>
<span>    source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsBackendMgf/man/MsBackendMgf.html" class="external-link">MsBackendMgf</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Start data import from 1 files ... done</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fenamiphos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mgf"</span>, <span class="st">"metlin-72445.mgf"</span>, package <span class="op">=</span> <span class="st">"xcms"</span><span class="op">)</span>,</span>
<span>    source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsBackendMgf/man/MsBackendMgf.html" class="external-link">MsBackendMgf</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Start data import from 1 files ... done</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">ex_spectrum</span>, <span class="va">flumanezil</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"against Flumanezil"</span>,</span>
<span>                  ppm <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">ex_spectrum</span>, <span class="va">fenamiphos</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"against Fenamiphos"</span>,</span>
<span>                  ppm <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/dda-ms2-metlin-match-1.png" alt="Mirror plots for the candidate MS2 spectrum against Flumanezil (left) and Fenamiphos (right). The upper panel represents the candidate MS2 spectrum, the lower the target MS2 spectrum. Matching peaks are indicated with a dot." width="1152"><p class="caption">
Mirror plots for the candidate MS2 spectrum against Flumanezil (left)
and Fenamiphos (right). The upper panel represents the candidate MS2
spectrum, the lower the target MS2 spectrum. Matching peaks are
indicated with a dot.
</p>
</div>
<p>Our candidate spectrum matches Fenamiphos, thus, our example
chromatographic peak represents signal measured for this compound. In
addition to plotting the spectra, we can also calculate similarities
between them with the <code>compareSpectra</code> method (which uses by
default the normalized dot-product to calculate the similarity).</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compareSpectra</a></span><span class="op">(</span><span class="va">ex_spectrum</span>, <span class="va">flumanezil</span>, ppm <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4.520957e-02 3.283806e-02 2.049379e-03 3.374354e-05</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compareSpectra</a></span><span class="op">(</span><span class="va">ex_spectrum</span>, <span class="va">fenamiphos</span>, ppm <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.1326234432 0.4879399946 0.7198406271 0.3997922658 0.0004876129</span></span>
<span><span class="co">## [6] 0.0028408885 0.0071030051 0.0053809736</span></span></code></pre>
<p>Clearly, the candidate spectrum does not match Flumanezil, while it
has a high similarity to Fenamiphos. While we performed here the
MS2-based annotation on a single chromatographic peak, this could be
easily extended to the full list of MS2 spectra (returned by
<code>chromPeakSpectra</code>) for all chromatographic peaks in an
experiment. See also <a href="https://jorainer.github.io/SpectraTutorials/" class="external-link">here</a> or <a href="https://jorainer.github.io/MetaboAnnotationTutorials" class="external-link">here</a> for
alternative tutorials on matching experimental fragment spectra against
a reference.</p>
<p>In the present example we used only a single data file and we did
thus not need to perform a sample alignment and correspondence analysis.
These tasks could however be performed similarly to <em>plain</em> LC-MS
data, retention times of recorded MS2 spectra would however also be
adjusted during alignment based on the MS1 data. After correspondence
analysis (peak grouping) MS2 spectra for <em>features</em> can be
extracted with the <code>featureSpectra</code> function which returns
all MS2 spectra associated with any chromatographic peak of a
feature.</p>
<p>Note also that this workflow can be included into the
<em>Feature-Based Molecular Networking</em> <a href="https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking/" class="external-link">FBMN</a>
to match MS2 spectra against <a href="https://gnps.ucsd.edu/" class="external-link">GNPS</a>.
See <a href="https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking-with-xcms3/" class="external-link">here</a>
for more details and examples.</p>
</div>
<div class="section level2">
<h2 id="dia-swath-data-analysis">DIA (SWATH) data analysis<a class="anchor" aria-label="anchor" href="#dia-swath-data-analysis"></a>
</h2>
<p>In this section we analyze a small SWATH data set consisting of a
single mzML file with data from the same sample analyzed in the previous
section but recorded in SWATH mode. We again read the data with the
<code>readMsExperiment</code> function. The resulting object will
contain all recorded MS1 and MS2 spectra in the specified file. Similar
to the previous data file, we filter the file to signal between 230 and
610 seconds.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">swath_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"TripleTOF-SWATH"</span>,</span>
<span>                          <span class="st">"PestMix1_SWATH.mzML"</span>,</span>
<span>                          package <span class="op">=</span> <span class="st">"msdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MsExperiment/man/readMsExperiment.html" class="external-link">readMsExperiment</a></span><span class="op">(</span><span class="va">swath_file</span><span class="op">)</span></span>
<span><span class="va">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterRt</a></span><span class="op">(</span><span class="va">swath_data</span>, rt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">230</span>, <span class="fl">610</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Below we determine the number of MS level 1 and 2 spectra in the
present data set.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">msLevel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    1    2 </span></span>
<span><span class="co">##  422 3378</span></span></code></pre>
<p>As described in the introduction, in SWATH mode all ions within
pre-defined isolation windows are fragmented and MS2 spectra measured.
The definition of these isolation windows (SWATH pockets) is imported
from the mzML files and available as additional <em>spectra
variables</em>. Below we inspect the respective information for the
first few spectra. The upper and lower isolation window m/z is available
with spectra variables <code>"isolationWindowLowerMz"</code> and
<code>"isolationWindowUpperMz"</code> respectively and the
<em>target</em> m/z of the isolation window with
<code>"isolationWindowTargetMz"</code>. We can use the
<code>spectraData</code> function to extract this information from the
spectra within our <code>swath_data</code> object.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectraData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"isolationWindowTargetMz"</span>, <span class="st">"isolationWindowLowerMz"</span>,</span>
<span>              <span class="st">"isolationWindowUpperMz"</span>, <span class="st">"msLevel"</span>, <span class="st">"rtime"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 5 columns</span></span>
<span><span class="co">##   isolationWindowTargetMz isolationWindowLowerMz isolationWindowUpperMz</span></span>
<span><span class="co">##                 &lt;numeric&gt;              &lt;numeric&gt;              &lt;numeric&gt;</span></span>
<span><span class="co">## 1                  299.10                  283.5                  314.7</span></span>
<span><span class="co">## 2                  329.80                  313.7                  345.9</span></span>
<span><span class="co">## 3                  367.35                  344.9                  389.8</span></span>
<span><span class="co">## 4                  601.85                  388.8                  814.9</span></span>
<span><span class="co">## 5                      NA                     NA                     NA</span></span>
<span><span class="co">## 6                  163.75                  139.5                  188.0</span></span>
<span><span class="co">##     msLevel     rtime</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1         2   230.073</span></span>
<span><span class="co">## 2         2   230.170</span></span>
<span><span class="co">## 3         2   230.267</span></span>
<span><span class="co">## 4         2   230.364</span></span>
<span><span class="co">## 5         1   230.459</span></span>
<span><span class="co">## 6         2   230.585</span></span></code></pre>
<p>We could also access these variables directly with the dedicated
<code>isolationWindowLowerMz</code> and
<code>isolationWindowUpperMz</code> functions.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">isolationWindowLowerMz</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 283.5 313.7 344.9 388.8    NA 139.5</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">isolationWindowUpperMz</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 314.7 345.9 389.8 814.9    NA 188.0</span></span></code></pre>
<p>In the present data set we use the value of the <em>isolation window
target m/z</em> to define the individual SWATH pockets. Below we list
the number of spectra that are recorded in each pocket/isolation
window.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/isolationWindowTargetMz-OnDiskMSnExp-method.html">isolationWindowTargetMz</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## 163.75 208.95 244.05 270.85  299.1  329.8 367.35 601.85 </span></span>
<span><span class="co">##    422    422    422    422    423    423    422    422</span></span></code></pre>
<p>We have thus between 422 and 423 MS2 spectra measured in each
isolation window.</p>
<p>To inspect the data we can also extract chromatograms from both the
measured MS1 as well as MS2 data. For MS2 data we have to set parameter
<code>msLevel = 2L</code> and in addition also specify the isolation
window from which we want to extract the data. Below we extract the TIC
of the MS1 data and of one of the isolation windows (isolation window
target m/z of 270.85) and plot these.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tic_ms1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">swath_data</span>, msLevel <span class="op">=</span> <span class="fl">1L</span>, aggregationFun <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span></span>
<span><span class="va">tic_ms2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">swath_data</span>, msLevel <span class="op">=</span> <span class="fl">2L</span>, aggregationFun <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>                        isolationWindowTargetMz <span class="op">=</span> <span class="fl">270.85</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tic_ms1</span>, main <span class="op">=</span> <span class="st">"MS1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tic_ms2</span>, main <span class="op">=</span> <span class="st">"MS2, isolation window m/z 270.85"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/unnamed-chunk-4-1.png" alt="TIC for MS1 (upper panel) and MS2 data from the isolation window with target m/z 270.85 (lower panel)." width="700"><p class="caption">
TIC for MS1 (upper panel) and MS2 data from the isolation window with
target m/z 270.85 (lower panel).
</p>
</div>
<p>Note that also DIA data from other manufacturers (e.g. Waters MSe)
are supported as long as a spectra variable
<code>isolationWindowTargetMz</code> is available. If that variable is
not provided in the raw data files it can also be manually assigned to
each spectrum in the data set. In the example below the precursor m/z
from the individual spectra would be used as their
<code>"isolationWindowTargetMz"</code> (note: the line below is only
shown as an example but not evaluated).</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">$</span><span class="va">isolationWindowTargetMz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorMz</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="chromatographic-peak-detection-in-ms1-and-ms2-data">Chromatographic peak detection in MS1 and MS2 data<a class="anchor" aria-label="anchor" href="#chromatographic-peak-detection-in-ms1-and-ms2-data"></a>
</h3>
<p>Similar to a <em>conventional</em> LC-MS analysis, we perform first a
chromatographic peak detection (on the MS level 1 data) with the
<code>findChromPeaks</code> method. Below we define the settings for a
<em>centWave</em>-based peak detection and perform the analysis.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>snthresh <span class="op">=</span> <span class="fl">5</span>, noise <span class="op">=</span> <span class="fl">100</span>, ppm <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                     peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks.html">findChromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span>
<span><span class="va">swath_data</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class XcmsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (422) MS2 (3378) </span></span>
<span><span class="co">##  Experiment data: 1 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 1 sample(s) to 3800 element(s).</span></span>
<span><span class="co">##  xcms results:</span></span>
<span><span class="co">##   - chromatographic peaks: 62 in MS level(s): 1</span></span></code></pre>
<p>Next we perform a chromatographic peak detection in MS level 2 data
separately for each individual isolation window. We use the
<code>findChromPeaksIsolationWindow</code> function employing the same
peak detection algorithm reducing however the required signal-to-noise
ratio. The <code>isolationWindow</code> parameter allows to specify
which MS2 spectra belong to which isolation window and hence defines in
which set of MS2 spectra chromatographic peak detection should be
performed. As a default the <code>"isolationWindowTargetMz"</code>
variable of the object’s spectra is used.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cwp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaks-centWave.html">CentWaveParam</a></span><span class="op">(</span>snthresh <span class="op">=</span> <span class="fl">3</span>, noise <span class="op">=</span> <span class="fl">10</span>, ppm <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                     peakwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">swath_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findChromPeaksIsolationWindow.html">findChromPeaksIsolationWindow</a></span><span class="op">(</span><span class="va">swath_data</span>, param <span class="op">=</span> <span class="va">cwp</span><span class="op">)</span></span>
<span><span class="va">swath_data</span></span></code></pre></div>
<pre><code><span><span class="co">## Object of class XcmsExperiment </span></span>
<span><span class="co">##  Spectra: MS1 (422) MS2 (3378) </span></span>
<span><span class="co">##  Experiment data: 1 sample(s)</span></span>
<span><span class="co">##  Sample data links:</span></span>
<span><span class="co">##   - spectra: 1 sample(s) to 3800 element(s).</span></span>
<span><span class="co">##  xcms results:</span></span>
<span><span class="co">##   - chromatographic peaks: 370 in MS level(s): 1, 2</span></span></code></pre>
<p>The <code>findChromPeaksIsolationWindow</code> function added all
peaks identified in the individual isolation windows to the
<code>chromPeaks</code> matrix containing already the MS1
chromatographic peaks. These newly added peaks can be identified through
the <code>"isolationWindow"</code> column in the object’s
<code>chromPeakData</code>.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 370 rows and 6 columns</span></span>
<span><span class="co">##        ms_level is_filled isolationWindow isolationWindowTargetMZ</span></span>
<span><span class="co">##       &lt;integer&gt; &lt;logical&gt;       &lt;numeric&gt;               &lt;numeric&gt;</span></span>
<span><span class="co">## CP01          1     FALSE              NA                      NA</span></span>
<span><span class="co">## CP02          1     FALSE              NA                      NA</span></span>
<span><span class="co">## CP03          1     FALSE              NA                      NA</span></span>
<span><span class="co">## CP04          1     FALSE              NA                      NA</span></span>
<span><span class="co">## CP05          1     FALSE              NA                      NA</span></span>
<span><span class="co">## ...         ...       ...             ...                     ...</span></span>
<span><span class="co">## CP366         2     FALSE          601.85                  601.85</span></span>
<span><span class="co">## CP367         2     FALSE          601.85                  601.85</span></span>
<span><span class="co">## CP368         2     FALSE          601.85                  601.85</span></span>
<span><span class="co">## CP369         2     FALSE          601.85                  601.85</span></span>
<span><span class="co">## CP370         2     FALSE          601.85                  601.85</span></span>
<span><span class="co">##       isolationWindowLowerMz isolationWindowUpperMz</span></span>
<span><span class="co">##                    &lt;numeric&gt;              &lt;numeric&gt;</span></span>
<span><span class="co">## CP01                      NA                     NA</span></span>
<span><span class="co">## CP02                      NA                     NA</span></span>
<span><span class="co">## CP03                      NA                     NA</span></span>
<span><span class="co">## CP04                      NA                     NA</span></span>
<span><span class="co">## CP05                      NA                     NA</span></span>
<span><span class="co">## ...                      ...                    ...</span></span>
<span><span class="co">## CP366                  388.8                  814.9</span></span>
<span><span class="co">## CP367                  388.8                  814.9</span></span>
<span><span class="co">## CP368                  388.8                  814.9</span></span>
<span><span class="co">## CP369                  388.8                  814.9</span></span>
<span><span class="co">## CP370                  388.8                  814.9</span></span></code></pre>
<p>Below we count the number of chromatographic peaks identified within
each isolation window (the number of chromatographic peaks identified in
MS1 is 62).</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">$</span><span class="va">isolationWindow</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## 163.75 208.95 244.05 270.85  299.1  329.8 367.35 601.85 </span></span>
<span><span class="co">##      2     38     32     14    105     23     62     32</span></span></code></pre>
<p>We thus successfully identified chromatographic peaks in the
different MS levels and isolation windows. As a next step we have to
identify which of the measured signals represents data from the same
original compound to <em>reconstruct</em> fragment spectra for each MS1
signal (chromatographic peak).</p>
</div>
<div class="section level3">
<h3 id="reconstruction-of-ms2-spectra">Reconstruction of MS2 spectra<a class="anchor" aria-label="anchor" href="#reconstruction-of-ms2-spectra"></a>
</h3>
<p>Identifying the signal of the fragment ions for the precursor
measured by each MS1 chromatographic peak is a non-trivial task. The MS2
spectrum of the fragment ion for each MS1 chromatographic peak has to be
reconstructed from the available MS2 signal (i.e. the chromatographic
peaks identified in MS level 2). For SWATH data, fragment ion signal
should be present in the same isolation window that contains the m/z of
the precursor ion and the chromatographic peak shape of the MS2
chromatographic peaks of fragment ions of a specific precursor should
have a similar retention time and peak shape than the precursor’s MS1
chromatographic peak.</p>
<p>After detection of MS1 and MS2 chromatographic peaks has been
performed, we can reconstruct the MS2 spectra using the
<code>reconstructChromPeakSpectra</code> function. This function defines
an MS2 spectrum for each MS1 chromatographic peak based on the following
approach:</p>
<ul>
<li>Identify MS2 chromatographic peaks in the isolation window
containing the m/z of the ion (the MS1 chromatographic peak) that have
approximately the same retention time than the MS1 chromatographic peak
(the accepted difference in retention time can be defined with the
<code>diffRt</code> parameter).</li>
<li>Extract the MS1 chromatographic peak and all MS2 chromatographic
peaks identified by the previous step and correlate the peak shapes of
the candidate MS2 chromatographic peaks with the shape of the MS1 peak.
MS2 chromatographic peaks with a correlation coefficient larger than
<code>minCor</code> are retained.</li>
<li>Reconstruct the MS2 spectrum using the m/z of all above selected MS2
chromatographic peaks and their intensity; each MS2 chromatographic peak
selected for an MS1 peak will thus represent one <strong>mass
peak</strong> in the reconstructed spectrum.</li>
</ul>
<p>To illustrate this process we perform the individual steps on the
example of fenamiphos (exact mass 303.105800777 and m/z of [M+H]+ adduct
304.113077). As a first step we extract the chromatographic peak for
this ion.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fenamiphos_mz</span> <span class="op">&lt;-</span> <span class="fl">304.113077</span></span>
<span><span class="va">fenamiphos_ms1_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span>, mz <span class="op">=</span> <span class="va">fenamiphos_mz</span>, ppm <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">fenamiphos_ms1_peak</span></span></code></pre></div>
<pre><code><span><span class="co">##            mz    mzmin    mzmax      rt   rtmin   rtmax     into     intb</span></span>
<span><span class="co">## CP34 304.1124 304.1121 304.1126 423.945 419.445 428.444 10697.34 10688.34</span></span>
<span><span class="co">##          maxo  sn sample</span></span>
<span><span class="co">## CP34 2401.849 618      1</span></span></code></pre>
<p>Next we identify all MS2 chromatographic peaks that were identified
in the isolation window containing the m/z of fenamiphos. The
information on the isolation window in which a chromatographic peak was
identified is available in the <code>chromPeakData</code>.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">$</span><span class="va">isolationWindowLowerMz</span> <span class="op">&lt;</span> <span class="va">fenamiphos_mz</span> <span class="op">&amp;</span></span>
<span>        <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeakData</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">$</span><span class="va">isolationWindowUpperMz</span> <span class="op">&gt;</span> <span class="va">fenamiphos_mz</span></span></code></pre></div>
<p>We also require the retention time of the MS2 chromatographic peaks
to be similar to the retention time of the MS1 peak and extract the
corresponding peak information. We thus below select all chromatographic
peaks for which the retention time range contains the retention time of
the apex position of the MS1 chromatographic peak.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">keep</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">[</span>, <span class="st">"rtmin"</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">fenamiphos_ms1_peak</span><span class="op">[</span>, <span class="st">"rt"</span><span class="op">]</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">[</span>, <span class="st">"rtmax"</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">fenamiphos_ms1_peak</span><span class="op">[</span>, <span class="st">"rt"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">fenamiphos_ms2_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>In total 24 MS2 chromatographic peaks match all the above conditions.
Next we extract the ion chromatogram of the MS1 peak and of all selected
candidate MS2 signals. To ensure chromatograms are extracted from
spectra in the correct isolation window we need to specify the
respective isolation window by passing its isolation window target m/z
to the <code>chromatogram</code> function (in addition to setting
<code>msLevel = 2</code>). This can be done by either getting the
<code>isolationWindowTargetMz</code> of the spectra after the data was
subset using <code>filterIsolationWindow</code> (as done below) or by
selecting the <code>isolationWindowTargetMz</code> closest to the m/z of
the compound of interest.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rtr</span> <span class="op">&lt;-</span> <span class="va">fenamiphos_ms1_peak</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="va">fenamiphos_ms1_peak</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">fenamiphos_ms1_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span><span class="va">swath_data</span>, rt <span class="op">=</span> <span class="va">rtr</span>, mz <span class="op">=</span> <span class="va">mzr</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rtr</span> <span class="op">&lt;-</span> <span class="va">fenamiphos_ms2_peak</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mzr</span> <span class="op">&lt;-</span> <span class="va">fenamiphos_ms2_peak</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Get the isolationWindowTargetMz for spectra containing the m/z of the</span></span>
<span><span class="co">## compound of interest</span></span>
<span><span class="va">swath_data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterIsolationWindow</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="va">fenamiphos_mz</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="../reference/isolationWindowTargetMz-OnDiskMSnExp-method.html">isolationWindowTargetMz</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## 299.1 </span></span>
<span><span class="co">##   423</span></span></code></pre>
<p>The target m/z of the isolation window containing the m/z of interest
is thus 299.1 and we can use this in the <code>chromatogram</code> call
below to extract the data from the correct (MS2) spectra.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fenamiphos_ms2_chr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span></span>
<span>    <span class="va">swath_data</span>, rt <span class="op">=</span> <span class="va">rtr</span>, mz <span class="op">=</span> <span class="va">mzr</span>, msLevel <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    isolationWindowTargetMz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">299.1</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rtr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can now plot the extracted ion chromatogram of the MS1 and the
extracted MS2 data.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">fenamiphos_ms1_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">fenamiphos_ms1_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>     xlab <span class="op">=</span> <span class="st">"retention time [s]"</span>, ylab <span class="op">=</span> <span class="st">"intensity"</span>, pch <span class="op">=</span> <span class="fl">16</span>,</span>
<span>     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5000</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"blue"</span>, type <span class="op">=</span> <span class="st">"b"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#' Add data from all MS2 peaks</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fenamiphos_ms2_chr</span><span class="op">@</span><span class="va">.Data</span>,</span>
<span>              <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">rtime</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">intensity</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span>,</span>
<span>                                 col <span class="op">=</span> <span class="st">"#00000080"</span>,</span>
<span>                                 type <span class="op">=</span> <span class="st">"b"</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/fena-eic-plot-1.png" alt="Extracted ion chromatograms for Fenamiphos from MS1 (blue) and potentially related signal in MS2 (grey)." width="960"><p class="caption">
Extracted ion chromatograms for Fenamiphos from MS1 (blue) and
potentially related signal in MS2 (grey).
</p>
</div>
<p>Next we can calculate correlations between the peak shapes of each
MS2 chromatogram with the MS1 peak. We illustrate this process on the
example of one MS2 chromatographic peaks. Note that, because MS1 and MS2
spectra are recorded consecutively, the retention times of the
individual data points will differ between the MS2 and MS1
chromatographic data and data points have thus to be matched (aligned)
before performing the correlation analysis. This is done automatically
by the <code>correlate</code> function. See the help for the
<code>align</code> method for more information on alignment options.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compareChromatograms</a></span><span class="op">(</span><span class="va">fenamiphos_ms2_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>               <span class="va">fenamiphos_ms1_chr</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>               ALIGNFUNARGS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"approx"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.9997871</span></span></code></pre>
<p>After identifying the MS2 chromatographic peaks with shapes of enough
high similarity to the MS1 chromatographic peaks, an MS2 spectrum could
be <em>reconstructed</em> based on the m/z and intensities of the MS2
chromatographic peaks (i.e., using their <code>"mz"</code> and
<code>"maxo"</code> or <code>"into"</code> values).</p>
<p>Instead of performing this assignment of MS2 signal to MS1
chromatographic peaks manually as above, we can use the
<code>reconstructChromPeakSpectra</code> function that performs the
exact same steps for all MS1 chromatographic peaks in a DIA data set.
Below we use this function to reconstruct MS2 spectra for our example
data requiring a peak shape correlation higher than <code>0.9</code>
between the candidate MS2 chromatographic peak and the target MS1
chromatographic peak.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">swath_spectra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/reconstructChromPeakSpectra.html">reconstructChromPeakSpectra</a></span><span class="op">(</span><span class="va">swath_data</span>, minCor <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span></span>
<span><span class="va">swath_spectra</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 62 spectra in a MsBackendMemory backend:</span></span>
<span><span class="co">##        msLevel     rtime scanIndex</span></span>
<span><span class="co">##      &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## CP01         2   239.458        NA</span></span>
<span><span class="co">## CP02         2   240.358        NA</span></span>
<span><span class="co">## CP03         2   329.577        NA</span></span>
<span><span class="co">## CP04         2   329.771        NA</span></span>
<span><span class="co">## CP05         2   346.164        NA</span></span>
<span><span class="co">## ...        ...       ...       ...</span></span>
<span><span class="co">## CP58         2   551.735        NA</span></span>
<span><span class="co">## CP59         2   551.735        NA</span></span>
<span><span class="co">## CP60         2   575.134        NA</span></span>
<span><span class="co">## CP61         2   575.134        NA</span></span>
<span><span class="co">## CP62         2   574.942        NA</span></span>
<span><span class="co">##  ... 20 more variables/columns.</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Merge 1 Spectra into one [Wed Oct 18 13:41:27 2023]</span></span></code></pre>
<p>As a result we got a <code>Spectra</code> object of length equal to
the number of MS1 peaks in our data. The length of a spectrum represents
the number of peaks it contains. Thus, a length of 0 indicates that no
matching peak (MS2 signal) could be found for the respective MS1
chromatographic peak.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">swath_spectra</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1]  0  0  1  1  1  0  0  0  0  0  0  0  3  0  3  4  0  3  0  1  0  9 14  1  0</span></span>
<span><span class="co">## [26]  0 15  4  1  1  2  4  6 15 12 11  2  4 13  0  0  0  0  1  2  0  1  0  0  0</span></span>
<span><span class="co">## [51]  3  0  2  1  7  7  0  0  0  0  0  2</span></span></code></pre>
<p>For reconstructed spectra additional annotations are available such
as the IDs of the MS2 chromatographic peaks from which the spectrum was
reconstructed (<code>"ms2_peak_id"</code>) as well as the correlation
coefficient of their chromatographic peak shape with the precursor’s
shape (<code>"ms2_peak_cor"</code>). Metadata column
<code>"peak_id"</code> contains the ID of the MS1 chromatographic
peak:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectraData</a></span><span class="op">(</span><span class="va">swath_spectra</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"peak_id"</span>, <span class="st">"ms2_peak_id"</span>, <span class="st">"ms2_peak_cor"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 62 rows and 3 columns</span></span>
<span><span class="co">##          peak_id ms2_peak_id      ms2_peak_cor</span></span>
<span><span class="co">##      &lt;character&gt;      &lt;list&gt;            &lt;list&gt;</span></span>
<span><span class="co">## CP01        CP01                              </span></span>
<span><span class="co">## CP02        CP02                              </span></span>
<span><span class="co">## CP03        CP03       CP063          0.950582</span></span>
<span><span class="co">## CP04        CP04       CP105           0.95157</span></span>
<span><span class="co">## CP05        CP05       CP153          0.924545</span></span>
<span><span class="co">## ...          ...         ...               ...</span></span>
<span><span class="co">## CP58        CP58                              </span></span>
<span><span class="co">## CP59        CP59                              </span></span>
<span><span class="co">## CP60        CP60                              </span></span>
<span><span class="co">## CP61        CP61                              </span></span>
<span><span class="co">## CP62        CP62 CP334,CP329 0.918915,0.911944</span></span></code></pre>
<p>We next extract the MS2 spectrum for our example peak most likely
representing [M+H]+ ions of Fenamiphos using its chromatographic peak
ID:</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fenamiphos_swath_spectrum</span> <span class="op">&lt;-</span> <span class="va">swath_spectra</span><span class="op">[</span></span>
<span>    <span class="va">swath_spectra</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fenamiphos_ms1_peak</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>We can now compare the reconstructed spectrum to the example
consensus spectrum from the DDA experiment in the previous section
(variable <code>ex_spectrum</code>) as well as to the MS2 spectrum for
Fenamiphos from Metlin (with a collision energy of 10V). For better
visualization we <em>normalize</em> also the peak intensities of the
reconstructed SWATH spectrum with the same function we used for the
experimental DDA spectrum.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fenamiphos_swath_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">fenamiphos_swath_spectrum</span>,</span>
<span>                                           <span class="va">scale_fun</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">fenamiphos_swath_spectrum</span>, <span class="va">ex_spectrum</span>,</span>
<span>     ppm <span class="op">=</span> <span class="fl">50</span>, main <span class="op">=</span> <span class="st">"against DDA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">fenamiphos_swath_spectrum</span>, <span class="va">fenamiphos</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>     ppm <span class="op">=</span> <span class="fl">50</span>, main <span class="op">=</span> <span class="st">"against Metlin"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/fena-swath-plot-1.png" alt="Mirror plot comparing the reconstructed MS2 spectrum for Fenamiphos (upper panel) against the measured spectrum from the DDA data and the Fenamiphhos spectrum from Metlin." width="1152"><p class="caption">
Mirror plot comparing the reconstructed MS2 spectrum for Fenamiphos
(upper panel) against the measured spectrum from the DDA data and the
Fenamiphhos spectrum from Metlin.
</p>
</div>
<p>If we wanted to get the EICs for the MS2 chromatographic peaks used
to generate this MS2 spectrum we can use the IDs of these peaks which
are provided with <code>$ms2_peak_id</code> of the result spectrum.</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pk_ids</span> <span class="op">&lt;-</span> <span class="va">fenamiphos_swath_spectrum</span><span class="op">$</span><span class="va">ms2_peak_id</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">pk_ids</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "CP199" "CP201" "CP211" "CP208" "CP200" "CP202" "CP217" "CP215" "CP205"</span></span>
<span><span class="co">## [10] "CP212" "CP221" "CP223" "CP213" "CP207" "CP220"</span></span></code></pre>
<p>With these peak IDs available we can extract their retention time
window and m/z ranges from the <code>chromPeaks</code> matrix and use
the <code>chromatogram</code> function to extract their EIC. Note
however that for SWATH data we have MS2 signal from different isolation
windows. Thus we have to first filter the <code>swath_data</code> object
by the isolation window containing the precursor m/z with the
<code>filterIsolationWindow</code> to subset the data to MS2 spectra
related to the ion of interest. In addition, we have to use
<code>msLevel = 2L</code> in the <code>chromatogram</code> call because
<code>chromatogram</code> extracts by default only data from MS1 spectra
and we need to specify the target m/z of the isolation window containing
the fragment data from the compound of interest.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rt_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">[</span><span class="va">pk_ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rtmin"</span>, <span class="st">"rtmax"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mz_range</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span><span class="op">)</span><span class="op">[</span><span class="va">pk_ids</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mzmin"</span>, <span class="st">"mzmax"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">pmz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">precursorMz</a></span><span class="op">(</span><span class="va">fenamiphos_swath_spectrum</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">## Determine the isolation window target m/z</span></span>
<span><span class="va">tmz</span> <span class="op">&lt;-</span> <span class="va">swath_data</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">filterIsolationWindow</a></span><span class="op">(</span>mz <span class="op">=</span> <span class="va">pmz</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">spectra</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="../reference/isolationWindowTargetMz-OnDiskMSnExp-method.html">isolationWindowTargetMz</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ms2_eics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chromatogram-method.html">chromatogram</a></span><span class="op">(</span></span>
<span>    <span class="va">swath_data</span>, rt <span class="op">=</span> <span class="va">rt_range</span>, mz <span class="op">=</span> <span class="va">mz_range</span>, msLevel <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    isolationWindowTargetMz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">tmz</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rt_range</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Each row of this <code>ms2_eics</code> contains now the EIC of one of
the MS2 chromatographic peaks. We can also plot these in an <em>overlay
plot</em>.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotChromatogramsOverlay.html">plotChromatogramsOverlay</a></span><span class="op">(</span><span class="va">ms2_eics</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/unnamed-chunk-14-1.png" alt="Overlay of EICs of chromatographic peaks used to reconstruct the MS2 spectrum for fenamiphos." width="700"><p class="caption">
Overlay of EICs of chromatographic peaks used to reconstruct the MS2
spectrum for fenamiphos.
</p>
</div>
<p>As a second example we analyze the signal from an [M+H]+ ion with an
m/z of 376.0381 (which would match <a href="https://en.wikipedia.org/wiki/Prochloraz" class="external-link">Prochloraz</a>). We
first identify the MS1 chromatographic peak for that m/z and retrieve
the reconstructed MS2 spectrum for that peak.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_mz</span> <span class="op">&lt;-</span> <span class="fl">376.0381</span></span>
<span></span>
<span><span class="va">prochloraz_ms1_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">swath_data</span>, msLevel <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>                                  mz <span class="op">=</span> <span class="va">prochloraz_mz</span>, ppm <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">prochloraz_ms1_peak</span></span></code></pre></div>
<pre><code><span><span class="co">##            mz   mzmin    mzmax      rt   rtmin   rtmax     into     intb</span></span>
<span><span class="co">## CP22 376.0373 376.037 376.0374 405.046 401.446 409.546 3664.051 3655.951</span></span>
<span><span class="co">##          maxo  sn sample</span></span>
<span><span class="co">## CP22 897.3923 278      1</span></span></code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_swath_spectrum</span> <span class="op">&lt;-</span> <span class="va">swath_spectra</span><span class="op">[</span></span>
<span>    <span class="va">swath_spectra</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">prochloraz_ms1_peak</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9</span></span></code></pre>
<p>The MS2 spectrum for the (tentative) MS1 signal for prochloraz
reconstructed from the SWATH MS2 data has thus 9 peaks.</p>
<p>In addition we identify the corresponding MS1 peak in the DDA data
set, extract all measured MS2 chromatographic peaks and build the
consensus spectrum from these.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_dda_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/XCMSnExp-class.html">chromPeaks</a></span><span class="op">(</span><span class="va">dda_data</span>, msLevel <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>                                  mz <span class="op">=</span> <span class="va">prochloraz_mz</span>, ppm <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">prochloraz_dda_peak</span></span></code></pre></div>
<pre><code><span><span class="co">##             mz    mzmin    mzmax      rt   rtmin   rtmax     into    intb</span></span>
<span><span class="co">## CP034 376.0385 376.0378 376.0391 405.295 401.166 410.145 5082.157 5072.77</span></span>
<span><span class="co">##           maxo  sn sample</span></span>
<span><span class="co">## CP034 1350.633 310      1</span></span></code></pre>
<p>The retention times for the chromatographic peaks from the DDA and
SWATH data match almost perfectly. Next we get the MS2 spectra for this
peak.</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_dda_spectra</span> <span class="op">&lt;-</span> <span class="va">dda_spectra</span><span class="op">[</span></span>
<span>    <span class="va">dda_spectra</span><span class="op">$</span><span class="va">peak_id</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">prochloraz_dda_peak</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">prochloraz_dda_spectra</span></span></code></pre></div>
<pre><code><span><span class="co">## MSn data (Spectra) with 5 spectra in a MsBackendMzR backend:</span></span>
<span><span class="co">##     msLevel     rtime scanIndex</span></span>
<span><span class="co">##   &lt;integer&gt; &lt;numeric&gt; &lt;integer&gt;</span></span>
<span><span class="co">## 1         2   401.438      3253</span></span>
<span><span class="co">## 2         2   402.198      3259</span></span>
<span><span class="co">## 3         2   404.677      3306</span></span>
<span><span class="co">## 4         2   405.127      3316</span></span>
<span><span class="co">## 5         2   405.877      3325</span></span>
<span><span class="co">##  ... 34 more variables/columns.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## file(s):</span></span>
<span><span class="co">## PestMix1_DDA.mzML</span></span>
<span><span class="co">## Processing:</span></span>
<span><span class="co">##  Filter: select retention time [230..610] on MS level(s) 1 2 [Wed Oct 18 13:41:01 2023]</span></span>
<span><span class="co">##  Filter: select MS level(s) 2 [Wed Oct 18 13:41:14 2023]</span></span>
<span><span class="co">##  Merge 1 Spectra into one [Wed Oct 18 13:41:14 2023]</span></span></code></pre>
<p>In total 5 spectra were measured, some with a relatively high number
of peaks. Next we combine them into a consensus spectrum.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_dda_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">combineSpectra</a></span><span class="op">(</span></span>
<span>    <span class="va">prochloraz_dda_spectra</span>, FUN <span class="op">=</span> <span class="va">combinePeaks</span>, ppm <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    peaks <span class="op">=</span> <span class="st">"intersect"</span>, minProp <span class="op">=</span> <span class="fl">0.8</span>, intensityFun <span class="op">=</span> <span class="va">median</span>, mzFun <span class="op">=</span> <span class="va">median</span>,</span>
<span>    f <span class="op">=</span> <span class="va">prochloraz_dda_spectra</span><span class="op">$</span><span class="va">peak_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Backend of the input object is read-only, will change that to an 'MsBackendMemory'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: 'combinePeaks' for lists of peak matrices is deprecated; please use</span></span>
<span><span class="co">## 'combinePeaksData' instead.</span></span></code></pre>
<p>At last we load also the Prochloraz MS2 spectra (for different
collision energies) from Metlin.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/Spectra.html" class="external-link">Spectra</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"mgf"</span>, <span class="st">"metlin-68898.mgf"</span>, package <span class="op">=</span> <span class="st">"xcms"</span><span class="op">)</span>,</span>
<span>    source <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/MsBackendMgf/man/MsBackendMgf.html" class="external-link">MsBackendMgf</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Start data import from 1 files ... done</span></span></code></pre>
<p>To validate the reconstructed spectrum we plot it against the
corresponding DDA spectrum and the MS2 spectrum for Prochloraz (for a
collision energy of 10V) from Metlin.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prochloraz_swath_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span>, <span class="va">scale_fun</span><span class="op">)</span></span>
<span><span class="va">prochloraz_dda_spectrum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">prochloraz_dda_spectrum</span>, <span class="va">scale_fun</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span>, <span class="va">prochloraz_dda_spectrum</span>,</span>
<span>                  ppm <span class="op">=</span> <span class="fl">40</span>, main <span class="op">=</span> <span class="st">"against DDA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectraMirror</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span>, <span class="va">prochloraz</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                  ppm <span class="op">=</span> <span class="fl">40</span>, main <span class="op">=</span> <span class="st">"against Metlin"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/pro-swath-plot-1.png" alt="Mirror plot comparing the reconstructed MS2 spectrum for Prochloraz (upper panel) against the measured spectrum from the DDA data and the Prochloraz spectrum from Metlin." width="1152"><p class="caption">
Mirror plot comparing the reconstructed MS2 spectrum for Prochloraz
(upper panel) against the measured spectrum from the DDA data and the
Prochloraz spectrum from Metlin.
</p>
</div>
<p>The spectra fit relatively well. Interestingly, the peak representing
the precursor (the right-most peak) seems to have a slightly shifted m/z
value in the reconstructed spectrum. Also, by closer inspecting the
spectrum two groups of peaks with small differences in m/z can be
observed (see plot below).</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/unnamed-chunk-15-1.png" alt="SWATH-derived MS2 spectrum for prochloraz." width="700"><p class="caption">
SWATH-derived MS2 spectrum for prochloraz.
</p>
</div>
<p>These could represent fragments from isotopes of the original
compound. DIA MS2 data, since all ions at a given retention time are
fragmented, can contain fragments from isotopes. We thus below use the
<code>isotopologues</code> function from the <em><a href="https://bioconductor.org/packages/3.18/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></em>
package to check for presence of potential isotope peaks in the
reconstructed MS2 spectrum for prochloraz.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RforMassSpectrometry/MetaboCoreUtils" class="external-link">MetaboCoreUtils</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/isotopologues.html" class="external-link">isotopologues</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/hidden_aliases.html" class="external-link">peaksData</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] 3 4 5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] 6 7</span></span></code></pre>
<p>Indeed, peaks 3, 4 and 5 as well as 6 and 7 have been assigned to a
group of potential isotope peaks. While this is no proof that the peaks
are indeed fragment isotopes of prochloraz it is highly likely (given
their difference in m/z and relative intensity differences). Below we
thus define a function that keeps only the monoisotopic peak for each
isotope group in a spectrum.</p>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Function to keep only the first (monoisotopic) peak for potential</span></span>
<span><span class="co">## isotopologue peak groups.</span></span>
<span><span class="va">rem_iso</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MetaboCoreUtils/man/isotopologues.html" class="external-link">isotopologues</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">idx</span>, <span class="kw">function</span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="va">z</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, use.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">idx</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">x</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="kw">else</span> <span class="va">x</span></span>
<span><span class="op">}</span></span>
<span><span class="va">prochloraz_swath_spectrum2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">addProcessing</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span>,</span>
<span>                                            <span class="va">rem_iso</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Spectra/man/spectra-plotting.html" class="external-link">plotSpectra</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="xcms-lcms-ms_files/figure-html/unnamed-chunk-18-1.png" alt="SWATH MS2 spectrum for prochloraz before (left) and after deisotoping (right)." width="700"><p class="caption">
SWATH MS2 spectrum for prochloraz before (left) and after deisotoping
(right).
</p>
</div>
<p>Removing the isotope peaks from the SWATH MS2 spectrum increases also
the spectra similarity score (since reference spectra generally will
contain only fragments of the ion of interest, but not of any of its
isotopes).</p>
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compareSpectra</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum</span>, <span class="va">prochloraz_dda_spectrum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.4623719</span></span></code></pre>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ProtGenerics/man/protgenerics.html" class="external-link">compareSpectra</a></span><span class="op">(</span><span class="va">prochloraz_swath_spectrum2</span>, <span class="va">prochloraz_dda_spectrum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5932303</span></span></code></pre>
<p>Similar to the DDA data, the reconstructed MS2 spectra from SWATH
data could be used in the annotation of the MS1 chromatographic
peaks.</p>
<!-- Other compounds with some issues to check: -->
<!-- - 306.162857, Buprofezin, Metlin 68681, consensus matches OKish with -->
<!--   SWATH. Metlin matches the DDA nicely, but not the SWATH. -->
<!-- - 300.0301, Azaconazole, Metlin 72479. Intensities match nicely, but m/z are -->
<!--   shifted. Metlin 20V matches DDA perfectly. SWATH m/z shifted. -->
<!-- - 224.083201. Spectra match relatively well, some almost perfect matches, but -->
<!--   consensus does not match that nicely. Nothing on Metlin. -->
<!-- - 256.1095, Dimethachlor, Metlin 72494. Intensities OK, m/z shifted. Metlin 20V -->
<!--   matches DDA perfectly. SWATH m/z shifted. -->
</div>
</div>
<div class="section level2">
<h2 id="outlook">Outlook<a class="anchor" aria-label="anchor" href="#outlook"></a>
</h2>
<p>Currently, spectra data representation, handling and processing is
being re-implemented as part of the <a href="https://rformassspectrometry.org" class="external-link">RforMassSpectrometry</a>
initiative aiming at increasing the performance of methods and
simplifying their use. Thus, parts of the workflow described here will
be changed (improved) in future.</p>
<p>Along with these developments, improved matching strategies for
larger data sets will be implemented as well as functionality to compare
<code>Spectra</code> directly to reference MS2 spectra from public
annotation resources (e.g. Massbank or HMDB). See for example <a href="https://jorainer.github.io/SpectraTutorials" class="external-link">here</a> for more
information.</p>
<p>Regarding SWATH data analysis, future development will involve
improved selection of the correct MS2 chromatographic peaks considering
also correlation with intensity values across several samples.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] MetaboCoreUtils_1.9.4 MsBackendMgf_1.9.0    MsExperiment_1.3.0   </span></span>
<span><span class="co">##  [4] pander_0.6.5          Spectra_1.11.11       xcms_3.99.6          </span></span>
<span><span class="co">##  [7] MSnbase_2.27.1        ProtGenerics_1.33.1   S4Vectors_0.39.3     </span></span>
<span><span class="co">## [10] mzR_2.35.1            Rcpp_1.0.11           Biobase_2.61.0       </span></span>
<span><span class="co">## [13] BiocGenerics_0.47.0   BiocParallel_1.35.4   BiocStyle_2.29.2     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-7                rlang_1.1.1                </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              clue_0.3-65                </span></span>
<span><span class="co">##   [5] MassSpecWavelet_1.67.0      matrixStats_1.0.0          </span></span>
<span><span class="co">##   [7] compiler_4.3.1              systemfonts_1.0.5          </span></span>
<span><span class="co">##   [9] vctrs_0.6.4                 stringr_1.5.0              </span></span>
<span><span class="co">##  [11] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">##  [13] fastmap_1.1.1               XVector_0.41.1             </span></span>
<span><span class="co">##  [15] utf8_1.2.3                  rmarkdown_2.25             </span></span>
<span><span class="co">##  [17] preprocessCore_1.63.1       ragg_1.2.6                 </span></span>
<span><span class="co">##  [19] purrr_1.0.2                 MultiAssayExperiment_1.27.5</span></span>
<span><span class="co">##  [21] xfun_0.40                   zlibbioc_1.47.0            </span></span>
<span><span class="co">##  [23] cachem_1.0.8                GenomeInfoDb_1.37.6        </span></span>
<span><span class="co">##  [25] jsonlite_1.8.7              progress_1.2.2             </span></span>
<span><span class="co">##  [27] DelayedArray_0.27.10        prettyunits_1.2.0          </span></span>
<span><span class="co">##  [29] parallel_4.3.1              cluster_2.1.4              </span></span>
<span><span class="co">##  [31] R6_2.5.1                    bslib_0.5.1                </span></span>
<span><span class="co">##  [33] stringi_1.7.12              RColorBrewer_1.1-3         </span></span>
<span><span class="co">##  [35] limma_3.57.10               GenomicRanges_1.53.2       </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4             bookdown_0.36              </span></span>
<span><span class="co">##  [39] SummarizedExperiment_1.31.1 iterators_1.0.14           </span></span>
<span><span class="co">##  [41] knitr_1.44                  IRanges_2.35.3             </span></span>
<span><span class="co">##  [43] igraph_1.5.1                splines_4.3.1              </span></span>
<span><span class="co">##  [45] Matrix_1.6-1.1              tidyselect_1.2.0           </span></span>
<span><span class="co">##  [47] abind_1.4-5                 yaml_2.3.7                 </span></span>
<span><span class="co">##  [49] doParallel_1.0.17           codetools_0.2-19           </span></span>
<span><span class="co">##  [51] affy_1.79.3                 lattice_0.21-9             </span></span>
<span><span class="co">##  [53] tibble_3.2.1                plyr_1.8.9                 </span></span>
<span><span class="co">##  [55] evaluate_0.22               survival_3.5-7             </span></span>
<span><span class="co">##  [57] desc_1.4.2                  pillar_1.9.0               </span></span>
<span><span class="co">##  [59] affyio_1.71.0               BiocManager_1.30.22        </span></span>
<span><span class="co">##  [61] MatrixGenerics_1.13.1       foreach_1.5.2              </span></span>
<span><span class="co">##  [63] MALDIquant_1.22.1           ncdf4_1.21                 </span></span>
<span><span class="co">##  [65] generics_0.1.3              rprojroot_2.0.3            </span></span>
<span><span class="co">##  [67] RCurl_1.98-1.12             hms_1.1.3                  </span></span>
<span><span class="co">##  [69] ggplot2_3.4.4               munsell_0.5.0              </span></span>
<span><span class="co">##  [71] scales_1.2.1                glue_1.6.2                 </span></span>
<span><span class="co">##  [73] lazyeval_0.2.2              MsFeatures_1.9.0           </span></span>
<span><span class="co">##  [75] tools_4.3.1                 mzID_1.39.0                </span></span>
<span><span class="co">##  [77] robustbase_0.99-0           QFeatures_1.11.2           </span></span>
<span><span class="co">##  [79] vsn_3.69.0                  RANN_2.6.1                 </span></span>
<span><span class="co">##  [81] fs_1.6.3                    XML_3.99-0.14              </span></span>
<span><span class="co">##  [83] grid_4.3.1                  impute_1.75.1              </span></span>
<span><span class="co">##  [85] MsCoreUtils_1.13.1          colorspace_2.1-0           </span></span>
<span><span class="co">##  [87] GenomeInfoDbData_1.2.11     cli_3.6.1                  </span></span>
<span><span class="co">##  [89] textshaping_0.3.7           fansi_1.0.5                </span></span>
<span><span class="co">##  [91] S4Arrays_1.1.6              dplyr_1.1.3                </span></span>
<span><span class="co">##  [93] AnnotationFilter_1.25.0     pcaMethods_1.93.0          </span></span>
<span><span class="co">##  [95] gtable_0.3.4                DEoptimR_1.1-3             </span></span>
<span><span class="co">##  [97] sass_0.4.7                  digest_0.6.33              </span></span>
<span><span class="co">##  [99] SparseArray_1.1.12          multtest_2.57.0            </span></span>
<span><span class="co">## [101] memoise_2.0.1               htmltools_0.5.6.1          </span></span>
<span><span class="co">## [103] pkgdown_2.0.7.9000          lifecycle_1.0.3            </span></span>
<span><span class="co">## [105] statmod_1.5.0               MASS_7.3-60</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Ludwig:2018hv" class="csl-entry">
1. Ludwig C, Gillet L, Rosenberger G, Amon S, Collins BC, Aebersold R:
<strong><a href="https://doi.org/10.15252/msb.20178126" class="external-link"><span class="nocase">Data-independent acquisition-based SWATH-MS for
quantitative proteomics: a tutorial.</span></a></strong> <em>Molecular
systems biology</em> 2018, <strong>14</strong>:e8126.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Colin A. Smith, Ralf Tautenhahn, Steffen Neumann, Paul Benton, Christopher Conley, Johannes Rainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
